{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/lauraluebbert/delphy/blob/main/tutorials/delphy_workflow_example.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ej4vLF8GBtT_"
      },
      "source": [
        "<div style=\"text-align: center;\">\n",
        "    <h1 style=\"font-size: 2.5em; color: #2C3E50;\">Running <a href=\"https://github.com/broadinstitute/delphy\" target=\"_blank\">Delphy</a> is as simple as 1, 2, 3 â€“ Ebolavirus Example</h1>\n",
        "    <h2 style=\"font-size: 1.5em; color: #34495E;\">Phylogenetic Tree Generation with Delphy on Ebolavirus Sequences</h2>\n",
        "</div>\n",
        "\n",
        "This notebook demonstrates how to generate a phylogenetic tree using [Delphy](https://github.com/broadinstitute/delphy) on viral sequences obtained from the [NCBI Virus Database](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/). Optionally, you can upload your own sequences to be included in the analysis.\n",
        "\n",
        "We will utilize the following tools:\n",
        "- [**gget**](https://github.com/pachterlab/gget/) to download sequences from the [NCBI Virus Database](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/)\n",
        "- [**MAFFT**](https://mafft.cbrc.jp/alignment/server/index.html) for creating a Multiple-Sequence Alignment (MSA)\n",
        "- [**Delphy**](https://github.com/broadinstitute/delphy) to generate the phylogenetic tree\n",
        "\n",
        "In this example notebook, we already filled out the NCBI Virus filters and options below to download and analyze all **complete *Zaire ebolavirus* sequences from *Homo sapiens* hosts collected between January 01, 2014 and December 31, 2014**.  \n",
        "\n",
        "Simply **click `Runtime` -> `Run all`** at the top of this notebook.\n",
        "\n",
        "If you encounter any problems or questions while using this notebook, please [report them here](https://github.com/broadinstitute/delphy/issues).\n",
        "\n",
        "Total runtime: ~2 min\n",
        "___\n",
        "___"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Apply filters to download sequences from [NCBI Virus](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/)"
      ],
      "metadata": {
        "id": "sWOSpHI-iGVO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title NCBI Virus filtering options:\n",
        "#@markdown Set any filter to `None` to disable it.\n",
        "\n",
        "def arg_str_to_bool(arg):\n",
        "  if arg == \"True\":\n",
        "      return True\n",
        "  elif arg == \"False\":\n",
        "      return False\n",
        "  elif arg == \"None\" or arg == \"\":\n",
        "      return None\n",
        "  else:\n",
        "      return arg\n",
        "\n",
        "#@markdown ## **Virus**\n",
        "\n",
        "virus = 'Zaire ebolavirus'  #@param {type:\"string\"}\n",
        "#@markdown  - Examples: 'Mammarenavirus lassaense' or 'coronaviridae' or 'NC_045512.2' or '142786' (Norovirus taxid).\n",
        "virus = arg_str_to_bool(virus)\n",
        "\n",
        "accession = False   #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if `virus` argument above is an NCBI accession (starts with 'NC'), e.g. 'NC_045512.2'.\n",
        "\n",
        "#@markdown ## **Host**\n",
        "\n",
        "host = 'homo sapiens'  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'homo sapiens' (alternative: use the `host_taxid` filter below). Input `None` to disable filtering by host.\n",
        "host = arg_str_to_bool(host)\n",
        "\n",
        "host_taxid = None  #@param {type:\"raw\"}\n",
        "#@markdown  - NCBI Taxonomy ID of host (e.g., 9443 for primates).\n",
        "host_taxid = arg_str_to_bool(host_taxid)\n",
        "\n",
        "#@markdown ## **Sequence completeness**\n",
        "\n",
        "annotated = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Set to `True` to only return sequences marked as 'annotated'; set to `False` to only return sequences NOT marked as 'annotated'.\n",
        "annotated = arg_str_to_bool(annotated)\n",
        "\n",
        "nuc_completeness = \"complete\"  #@param [\"None\", \"complete\", \"partial\"]\n",
        "#@markdown  - Choose between 'partial' or 'complete' nucleotide completeness.\n",
        "nuc_completeness = arg_str_to_bool(nuc_completeness)\n",
        "\n",
        "min_seq_length = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum sequence length, e.g. 6252.\n",
        "min_seq_length = arg_str_to_bool(min_seq_length)\n",
        "\n",
        "max_seq_length = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum sequence length, e.g. 7815.\n",
        "max_seq_length = arg_str_to_bool(max_seq_length)\n",
        "\n",
        "has_proteins = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Require sequences to contain specific proteins (e.g. 'GPC' **- include the quotation marks**) or a list of proteins (e.g. ['GPC', 'L']). Also accepts names of genes or segments.\n",
        "has_proteins = arg_str_to_bool(has_proteins)\n",
        "\n",
        "proteins_complete = False    #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if the proteins/genes/segments in `has_proteins` should be marked as 'complete'.\n",
        "proteins_complete = arg_str_to_bool(proteins_complete)\n",
        "\n",
        "max_ambiguous_chars = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum number of 'N' characters allowed in each sequence, e.g. 10.\n",
        "max_ambiguous_chars = arg_str_to_bool(max_ambiguous_chars)\n",
        "\n",
        "#@markdown ## **Gene/peptide/protein counts**\n",
        "\n",
        "min_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum gene count, e.g. 1.\n",
        "min_gene_count = arg_str_to_bool(min_gene_count)\n",
        "\n",
        "max_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum gene count, e.g. 40.\n",
        "max_gene_count = arg_str_to_bool(max_gene_count)\n",
        "\n",
        "min_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum peptide count, e.g. 2.\n",
        "min_mature_peptide_count = arg_str_to_bool(min_mature_peptide_count)\n",
        "\n",
        "max_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum peptide count, e.g. 15.\n",
        "max_mature_peptide_count = arg_str_to_bool(max_mature_peptide_count)\n",
        "\n",
        "min_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum protein count, e.g. 2.\n",
        "min_protein_count = arg_str_to_bool(min_protein_count)\n",
        "\n",
        "max_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum protein count, e.g. 10.\n",
        "max_protein_count = arg_str_to_bool(max_protein_count)\n",
        "\n",
        "#@markdown ## **Geographic location**\n",
        "\n",
        "geographic_location = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic location of sample collection, e.g. 'South Africa' or 'Germany'.\n",
        "geographic_location = arg_str_to_bool(geographic_location)\n",
        "\n",
        "geographic_region = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic region of sample collection, e.g. 'Africa' or 'Europe'.\n",
        "geographic_region = arg_str_to_bool(geographic_region)\n",
        "\n",
        "#@markdown ## **Dates**\n",
        "#@markdown All dates should be supplied in YYYY-MM-DD format.\n",
        "min_collection_date = \"2014-01-01\"  #@param {type:\"string\"}\n",
        "#@markdown  - Minimum collection date, e.g. '2000-01-01'.\n",
        "min_collection_date = arg_str_to_bool(min_collection_date)\n",
        "\n",
        "max_collection_date = \"2014-12-31\"  #@param {type:\"string\"}\n",
        "#@markdown  - Maximum collection date, e.g. '2014-12-04'.\n",
        "max_collection_date = arg_str_to_bool(max_collection_date)\n",
        "\n",
        "min_release_date = \"None\"  #@param {type:\"string\"}\n",
        "#@markdown - Minimum release date of the sequences, e.g. '2000-01-01'.\n",
        "#@markdown - **Specifying a minimum release date is recommended** where possible, as it speeds up and reduces the disk space required by the data download from NCBI Virus.\n",
        "min_release_date = arg_str_to_bool(min_release_date)\n",
        "\n",
        "max_release_date = \"None\"  #@param {type:\"string\"}\n",
        "#@markdown  - Maximum release date of the sequences, e.g. '2014-12-04'.\n",
        "max_release_date = arg_str_to_bool(max_release_date)\n",
        "\n",
        "#@markdown ## **Source**\n",
        "\n",
        "submitter_country = None  #@param {type:\"string\"}\n",
        "#@markdown  - Country that submitted the sequence, e.g. 'South Africa' or 'Germany'.\n",
        "submitter_country = arg_str_to_bool(submitter_country)\n",
        "\n",
        "lab_passaged = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Set to `True` to only return sequences that have been passaged in a laboratory setting; set to `False` to only return sequences that have NOT been passaged in a laboratory setting.\n",
        "lab_passaged = arg_str_to_bool(lab_passaged)\n",
        "\n",
        "source_database = None  #@param {type:\"string\"}\n",
        "#@markdown  - Source database of the sequence, e.g. 'GenBank' or 'RefSeq'.\n",
        "source_database = arg_str_to_bool(source_database)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "6cazzIZfiJDU"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Optional: Upload a fasta file with your own sequences to add to the analysis\n",
        "  **1) Click on the folder icon on the left.  \n",
        "  2) Upload your file(s) to the Google Colab server by dragging in your file(s) (or use rightclick -> Upload).  \n",
        "  3) Specify the name of your file(s) here:**"
      ],
      "metadata": {
        "id": "aE8ot-AyiPdC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title FASTA file containing additional sequences\n",
        "\n",
        "fasta_file = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_fasta_file.fa' or 'my_fasta_file.fasta'.\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "E-lg25zave0H"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Metadata\n",
        "\n",
        "#@markdown **Option 1: The metadata is the same for all sequences in your FASTA file**\n",
        "metadata = {'Collection Date': 'YYYY-MM-DD', 'Geo Location': 'South Korea'}  #@param {type:\"raw\"}\n",
        "#@markdown - The 'Collection Date' field is required. Optional: you can add as many additional columns as you wish, e.g. 'Geo Location': 'South Korea'.\n",
        "#@markdown - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "#@markdown **Option 2: Input a CSV file with metadata for each sequence**\n",
        "metadata_csv = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_metadata.csv'. This file must include at least an 'Accession' and 'Collection Date' column.\n",
        "#@markdown  - NOTE: Make sure the IDs in the \"Accession\" column match the IDs of the sequences in the provided FASTA file\n",
        "#@markdown  - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "# Convert empty strings to None\n",
        "fasta_file = arg_str_to_bool(fasta_file)\n",
        "metadata_csv = arg_str_to_bool(metadata_csv)"
      ],
      "metadata": {
        "id": "LKgxMFBgiP64",
        "cellView": "form"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Optional: Specify Delphy arguments"
      ],
      "metadata": {
        "id": "zQF5zZ2dzS60"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Delphy options:\n",
        "mutation_rate = 0.01  #@param {type:\"raw\"}\n",
        "#@markdown  - Virus mutation rate (mutations per site per year), e.g. 0.01. If set to `None`, the mutation rate will be estimated from the generated tree.\n",
        "\n",
        "delphy_steps = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Number of steps to run in the Delphy algorithm (default: 500,000 * number of sequences).\n",
        "\n",
        "delphy_samples = 200 #@param {type:\"integer\"}\n",
        "#@markdown  - Number of logging and tree updates (will log every `delphy_steps/delphy_samples` step).\n",
        "\n",
        "delphy_release = 0.9996  #@param {type:\"number\"}\n",
        "#@markdown  - Delphy version to use (see https://github.com/broadinstitute/delphy/releases).\n",
        "\n",
        "#@title Delphy options:\n",
        "threads = 2  #@param {type:\"integer\"}\n",
        "#@markdown  - Number of threads to use for the MSA (Mafft) and phylogenetic tree (Delphy) generation."
      ],
      "metadata": {
        "cellView": "form",
        "id": "bEM5nlzZzSih"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# All done! Select `Runtime` at the top of this notebook, then click `Run all` and lean back...\n",
        "A completion message will be displayed below once the notebook has been successfully executed.  \n",
        "**ðŸ’¡ Tip: Click on the folder icon on the left to view/download the files that are being generated.**\n",
        "  \n",
        "<br>\n",
        "\n",
        "____\n",
        "____"
      ],
      "metadata": {
        "id": "UTL55NF2jAj-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "#@title # Generating the phylogenetic tree...\n",
        "import subprocess\n",
        "from IPython.display import display, HTML\n",
        "\n",
        "def log_message(text):\n",
        "    display(HTML(f\"<h2 style='color: green;'>{text}</h2>\"))\n",
        "def log_message_error(text):\n",
        "    display(HTML(f\"<h2 style='color: red;'>{text}</h2>\"))\n",
        "\n",
        "\n",
        "log_message(\"1/5 Installing software...\")\n",
        "\n",
        "# Install gget\n",
        "try:\n",
        "    result = subprocess.run(['pip', 'install', '-q', 'mysql-connector-python', 'biopython'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout)\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while installing mysql-connector-python and biopython: \")\n",
        "    print(e.stderr)\n",
        "\n",
        "try:\n",
        "    # After the release, this will just be: pip install gget (dependence on biopython will be removed)\n",
        "    result = subprocess.run(['pip', 'install', '-q', 'git+https://github.com/pachterlab/gget.git@delphy_dev'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout)\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while installing gget: \")\n",
        "    print(e.stderr)\n",
        "\n",
        "\n",
        "import gget\n",
        "from Bio import SeqIO\n",
        "import pandas as pd\n",
        "import re\n",
        "from datetime import datetime\n",
        "\n",
        "log_message(\"1/5 Software installation complete.\")\n",
        "\n",
        "\n",
        "# Downloading virus genomes from NCBI Virus\n",
        "log_message(\"2/5 Download data from NCBI Virus... This might take a few minutes depending on the internet connection and how busy the NCBI server is.\")\n",
        "gget.ncbi_virus(\n",
        "    virus = virus,\n",
        "    accession = accession,\n",
        "    host = host,\n",
        "    min_seq_length = min_seq_length,\n",
        "    max_seq_length = max_seq_length,\n",
        "    min_gene_count = min_gene_count,\n",
        "    max_gene_count = max_gene_count,\n",
        "    nuc_completeness = nuc_completeness,\n",
        "    has_proteins = has_proteins,\n",
        "    proteins_complete = proteins_complete,\n",
        "    host_taxid = host_taxid,\n",
        "    lab_passaged = lab_passaged,\n",
        "    geographic_region = geographic_region,\n",
        "    geographic_location = geographic_location,\n",
        "    submitter_country = submitter_country,\n",
        "    min_collection_date = min_collection_date,\n",
        "    max_collection_date = max_collection_date,\n",
        "    annotated = annotated,\n",
        "    source_database = source_database,\n",
        "    min_release_date = min_release_date,\n",
        "    max_release_date = max_release_date,\n",
        "    min_mature_peptide_count = min_mature_peptide_count,\n",
        "    max_mature_peptide_count = max_mature_peptide_count,\n",
        "    min_protein_count = min_protein_count,\n",
        "    max_protein_count = max_protein_count,\n",
        "    max_ambiguous_chars = max_ambiguous_chars\n",
        ")\n",
        "\n",
        "ncbi_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences.fasta\"\n",
        "ncbi_metadata = f\"{'_'.join(str(virus).split(' '))}_metadata.csv\"\n",
        "\n",
        "log_message(f\"2/5 Data download from NCBI Virus complete. Sequences were saved in the {ncbi_fasta_file} file and the associated metadata in {ncbi_metadata}/.jsonl\")\n",
        "\n",
        "\n",
        "# Merging sequencing and metadata files if additional file(s) were provided\n",
        "if fasta_file:\n",
        "  log_message(\"Adding user-provided fasta file and metadata to the data from NCBI Virus...\")\n",
        "\n",
        "  # Combine sequence files\n",
        "  combined_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences_combined.fasta\"\n",
        "  !cat $ncbi_fasta_file $fasta_file > $combined_fasta_file\n",
        "  input_fasta_file = combined_fasta_file\n",
        "\n",
        "  # Combine metadata\n",
        "  combined_metadata_file = f\"{'_'.join(virus.split(' '))}_metadata_combined.csv\"\n",
        "  ncbi_metadata_df = pd.read_csv(ncbi_metadata)\n",
        "  if metadata_csv:\n",
        "    # Combine provided metadata and NCBI metadata csv files\n",
        "    user_metada_df = pd.read_csv(metadata_csv)\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  else:\n",
        "    # Extract sequence accessions from the provided FASTA file\n",
        "    headers = [record.id.split(\" \")[0] for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
        "\n",
        "    # Create a metadata dataframe with the accessions from the FASTA file and the provided metadata\n",
        "    user_metada_df = pd.DataFrame(headers, columns=[\"Accession\"])\n",
        "    for key, value in metadata.items():\n",
        "      user_metada_df[key] = value\n",
        "\n",
        "    # Combine with NCBI metadata\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  log_message(f\"Merging user-provided and NCBI Virus data complete. The combined sequence and metadata files were saved as {combined_fasta_file} and {combined_metadata_file}.\")\n",
        "\n",
        "else:\n",
        "  input_fasta_file = ncbi_fasta_file\n",
        "  metadata_file = ncbi_metadata\n",
        "\n",
        "\n",
        "# Create MSA\n",
        "log_message(\"3/5 Multiple Sequence Aligment (MSA): Aligning the sequences to each other so they are all in the same frame...\")\n",
        "\n",
        "aligned_fasta_file = f\"{'_'.join(str(virus).split(' '))}_aligned.afa\"\n",
        "\n",
        "# # Option 1: Using the MUSCLE algorithm (this works well for a few hundred sequences, but is too slow when dealing with a few thousand sequences)\n",
        "# gget.muscle(input_fasta_file, super5=True, out=aligned_fasta_file)\n",
        "\n",
        "# Option 2: Using mafft\n",
        "# TO-DO: Wrap the following code into gget module and replace with command `gget.mafft(input_fasta_file, out=aligned_fasta_file)`\n",
        "\n",
        "#Install MAFFT\n",
        "try:\n",
        "    # Download MAFFT .deb package\n",
        "    subprocess.run(\n",
        "        ['wget', 'https://mafft.cbrc.jp/alignment/software/mafft_7.526-1_amd64.deb'],\n",
        "        check=True\n",
        "    )\n",
        "\n",
        "    # Install the .deb package using dpkg\n",
        "    subprocess.run(\n",
        "        ['dpkg', '-i', 'mafft_7.526-1_amd64.deb'],\n",
        "        check=True\n",
        "    )\n",
        "\n",
        "    # Verify the MAFFT installation\n",
        "    result = subprocess.run(\n",
        "        ['mafft', '--version'],\n",
        "        check=True,\n",
        "        capture_output=True,\n",
        "        text=True\n",
        "    )\n",
        "    print(\"MAFFT version:\", result.stderr.strip())\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    print(\"An error occurred:\", e.stderr)\n",
        "\n",
        "# Aligning sequences to each other using mafft\n",
        "# TO-DO: Add --parttree argument when aligning >10,000 sequences\n",
        "try:\n",
        "    with open(aligned_fasta_file, 'w') as output:\n",
        "        result = subprocess.run(['mafft', '--auto', '--thread', str(threads), input_fasta_file], stdout=output, stderr=subprocess.PIPE, check=True)\n",
        "    log_message(f\"3/5 MSA complete and saved in the {aligned_fasta_file} file.\")\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message(f\"An error occurred while generating the MSA using Mafft: \")\n",
        "    print(e.stderr)\n",
        "\n",
        "\n",
        "# TO-DO: Wrap the following code into a gget module and replace with command `gget.delphy(aligned_fasta_file, metadata_file)`\n",
        "# Adjust the headers in the aligned fasta file to match header format required by Delphy (accession|YYYY-MM-DD):\n",
        "log_message(\"4/5 Reformatting sequence files to match Delphy format...\")\n",
        "\n",
        "# Reformat collection date\n",
        "default_day = '01'\n",
        "default_month = '01'\n",
        "def extract_and_format_date(date_string):\n",
        "    # Define regular expressions for various date formats\n",
        "    year_only = re.compile(r'(?P<year>\\d{4})')\n",
        "    year_month = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})')\n",
        "    full_date = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})[-/.](?P<day>\\d{1,2})')\n",
        "\n",
        "    # Try to match the full date first\n",
        "    match = full_date.search(date_string)\n",
        "    if match:\n",
        "        year = match.group('year')\n",
        "        month = match.group('month').zfill(2)\n",
        "        day = match.group('day').zfill(2)\n",
        "    else:\n",
        "        # Try to match year and month\n",
        "        match = year_month.search(date_string)\n",
        "        if match:\n",
        "            year = match.group('year')\n",
        "            month = match.group('month').zfill(2)\n",
        "            day = default_day\n",
        "        else:\n",
        "            # Try to match only the year\n",
        "            match = year_only.search(date_string)\n",
        "            if match:\n",
        "                year = match.group('year')\n",
        "                month = default_month\n",
        "                day = default_day\n",
        "            else:\n",
        "                # If no match, return None\n",
        "                return None\n",
        "\n",
        "    # Format the extracted date into YYYY-MM-DD\n",
        "    formatted_date = f\"{year}-{month}-{day}\"\n",
        "\n",
        "    try:\n",
        "        # Validate date by trying to convert it to a datetime object\n",
        "        datetime.strptime(formatted_date, '%Y-%m-%d')\n",
        "    except ValueError:\n",
        "        return None  # Return None if the date is invalid\n",
        "\n",
        "    return formatted_date\n",
        "\n",
        "def update_fasta_headers(fasta_file, csv_file, output_fasta):\n",
        "    # Read the CSV file into a pandas DataFrame\n",
        "    df = pd.read_csv(csv_file)\n",
        "\n",
        "    # Create a dictionary from the DataFrame to map accession to date\n",
        "    accession_to_date = pd.Series(df['Collection Date'].values, index=df['accession']).to_dict()\n",
        "\n",
        "    # Open the input fasta file and output fasta file\n",
        "    with open(fasta_file) as fasta_input, open(output_fasta, 'w') as fasta_output:\n",
        "        # Iterate through each sequence record in the fasta file\n",
        "        for record in SeqIO.parse(fasta_input, 'fasta'):\n",
        "            accession = record.id\n",
        "\n",
        "            # Check if the accession is in the pandas dictionary and has a non-NaN date\n",
        "            if accession in accession_to_date and pd.notna(accession_to_date[accession]):\n",
        "                date = accession_to_date[accession]\n",
        "\n",
        "                # Format the date if necessary\n",
        "                formatted_date = extract_and_format_date(date)\n",
        "\n",
        "                if formatted_date is None:\n",
        "                  # Skip the entry if date is NaN or accession not found\n",
        "                  print(f\"Skipping accession {accession} due to unrecognized date format: '{date}'\")\n",
        "                  continue\n",
        "\n",
        "                # Update the seq header\n",
        "                record.id = f\"{accession}|{formatted_date}\"\n",
        "                record.description = ''  # Remove the original description to avoid duplication\n",
        "            else:\n",
        "                # Skip the entry if date is NaN or accession not found\n",
        "                print(f\"Skipping accession {accession} due to missing or NaN date.\")\n",
        "                continue\n",
        "\n",
        "            # Write the updated record to the output fasta file\n",
        "            SeqIO.write(record, fasta_output, 'fasta')\n",
        "\n",
        "aligned_fasta_file_clean = f\"{'_'.join(str(virus).split(' '))}_aligned_headers_adjusted.afa\"\n",
        "update_fasta_headers(aligned_fasta_file, metadata_file, aligned_fasta_file_clean)\n",
        "\n",
        "log_message(\"4/5 Reformatting complete.\")\n",
        "\n",
        "\n",
        "# Run Delphy\n",
        "log_message(\"5/5 Running Delphy...\")\n",
        "\n",
        "# Download Delphy binary\n",
        "try:\n",
        "    result = subprocess.run(['wget', f'https://github.com/broadinstitute/delphy/releases/download/{delphy_release}/delphy-ubuntu-x86_64'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout)\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while downloading Delphy: \")\n",
        "    print(e.stderr)\n",
        "\n",
        "# Give permissions\n",
        "try:\n",
        "    result = subprocess.run(['chmod', 'u+x', './delphy-ubuntu-x86_64'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout)\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while changing permissions for Delphy: \")\n",
        "    print(e.stderr)\n",
        "\n",
        "beast_log_out = f\"{'_'.join(str(virus).split(' '))}_delphy_beast_log.txt\"\n",
        "delphy_beast_tree_out = f\"{'_'.join(str(virus).split(' '))}_delphy_beast_tree.nwk\"\n",
        "dphy_out = f\"{'_'.join(str(virus).split(' '))}_delphy_out.dphy\"\n",
        "\n",
        "# Define Delphy arguments\n",
        "def count_fasta_sequences(fasta_file):\n",
        "    with open(fasta_file, 'r') as f:\n",
        "        return sum(1 for line in f if line.startswith('>'))\n",
        "\n",
        "if delphy_steps is None:\n",
        "  # Get number of sequences\n",
        "  num_seqs = count_fasta_sequences(aligned_fasta_file_clean)\n",
        "  delphy_steps = 500000 * num_seqs\n",
        "\n",
        "log_every = int(delphy_steps / delphy_samples)\n",
        "tree_every = int(delphy_steps / delphy_samples)\n",
        "\n",
        "# Define the command\n",
        "if mutation_rate:\n",
        "    cmd = [\n",
        "        './delphy-ubuntu-x86_64',\n",
        "        '--v0-fix-mutation-rate',\n",
        "        '--v0-init-mutation-rate', str(mutation_rate),\n",
        "        '--v0-log-every', str(log_every),\n",
        "        '--v0-tree-every', str(tree_every),\n",
        "        '--v0-threads', str(threads),\n",
        "        '--v0-steps', str(delphy_steps),\n",
        "        '--v0-in-fasta', aligned_fasta_file_clean,\n",
        "        '--v0-out-log-file', beast_log_out,\n",
        "        '--v0-out-trees-file', delphy_beast_tree_out,\n",
        "        '--v0-out-delphy-file', dphy_out\n",
        "    ]\n",
        "else:\n",
        "    cmd = [\n",
        "        './delphy-ubuntu-x86_64',\n",
        "        '--v0-log-every', str(log_every),\n",
        "        '--v0-tree-every', str(tree_every),\n",
        "        '--v0-threads', str(threads),\n",
        "        '--v0-steps', str(delphy_steps),\n",
        "        '--v0-in-fasta', aligned_fasta_file_clean,\n",
        "        '--v0-out-log-file', beast_log_out,\n",
        "        '--v0-out-trees-file', delphy_beast_tree_out,\n",
        "        '--v0-out-delphy-file', dphy_out\n",
        "    ]\n",
        "\n",
        "try:\n",
        "    result = subprocess.run(cmd, check=True, text=True, capture_output=True)\n",
        "    # print(result.stdout)\n",
        "\n",
        "    display(HTML(\"\"\"\n",
        "    <h1>All done! ðŸŽ‰</h1>\n",
        "    <h2>To download the files we generated in this notebook to your local computer, click on the folder icon on the left and download files by right clicking a file of interest and selecting 'Download'.</h2>\n",
        "    <h2>To further visualize your Delphy output, upload the <code>.dphy</code> (and <code>metadata.csv</code>) file(s) to <a href='https://delphy.fathom.info/' target='_blank'>https://delphy.fathom.info/</a></h2>\n",
        "    \"\"\"))\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while running Delphy: \")\n",
        "    print(e.stderr)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "HuVr7cb0x5VF",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "1f2e8a65-877e-4ce0-f417-566f7e1cc58b"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>1/5 Installing software...</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>1/5 Software installation complete.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>2/5 Download data from NCBI Virus... This might take a few minutes depending on the internet connection and how busy the NCBI server is.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "New version of client (17.1.0) available at https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/LATEST/linux-amd64/datasets.\n",
            "INFO:gget.utils:234 sequences passed the provided filters.\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>2/5 Data download from NCBI Virus complete. Sequences were saved in the Zaire_ebolavirus_sequences.fasta file and the associated metadata in Zaire_ebolavirus_metadata.csv/.jsonl</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>3/5 Multiple Sequence Aligment (MSA): Aligning the sequences to each other so they are all in the same frame...</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "MAFFT version: v7.526 (2024/Apr/26)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>3/5 MSA complete and saved in the Zaire_ebolavirus_aligned.afa file.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>4/5 Reformatting sequence files to match Delphy format...</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>4/5 Reformatting complete.</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: green;'>5/5 Running Delphy...</h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h2 style='color: red;'>An error occurred while running Delphy: </h2>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# Seed: 4088835061\n",
            "Reading fasta file Zaire_ebolavirus_aligned_headers_adjusted.afa\n",
            "Building rough initial tree...\n",
            "Warning: ambiguous state Y at site 9268 changed to N\n",
            "Warning: ambiguous state W at site 18911 changed to N\n",
            "Warning: ambiguous state R at site 18913 changed to N\n",
            "Warning: ambiguous state Y at site 2981 changed to N\n",
            "Warning: ambiguous state Y at site 12343 changed to N\n",
            "Attaching tip 2 to Candidate_region{branch=1, mut_idx=10, -2196.21<t<=-2191, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 3 to Candidate_region{branch=0, mut_idx=7, -2234.07<t<=-2191, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 4 to Candidate_region{branch=3, mut_idx=2, -2075.71<t<=-2047, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 5 to Candidate_region{branch=234, mut_idx=0, -1.79769e+308<t<=-2322, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 6 to Candidate_region{branch=238, mut_idx=0, -1.79769e+308<t<=-2323, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 7 to Candidate_region{branch=6, mut_idx=1, -2318.31<t<=-2046, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 8 to Candidate_region{branch=3, mut_idx=0, -2068.2<t<=-2047, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 9 to Candidate_region{branch=8, mut_idx=0, -2062.04<t<=-2045, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 10 to Candidate_region{branch=6, mut_idx=0, -2312.77<t<=-2046, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 11 to Candidate_region{branch=6, mut_idx=0, -2178.75<t<=-2046, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 12 to Candidate_region{branch=5, mut_idx=3, -2085.24<t<=-2046, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 13 to Candidate_region{branch=241, mut_idx=0, -2068.2<t<=-2062.04, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 14 to Candidate_region{branch=10, mut_idx=0, -2178.75<t<=-2046, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 15 to Candidate_region{branch=241, mut_idx=0, -2067.57<t<=-2062.04, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 16 to Candidate_region{branch=4, mut_idx=0, -2068.2<t<=-2047, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 17 to Candidate_region{branch=4, mut_idx=0, -2055.89<t<=-2047, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 18 to Candidate_region{branch=239, mut_idx=0, -1.79769e+308<t<=-2324, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 19 to Candidate_region{branch=251, mut_idx=0, -1.79769e+308<t<=-2325, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 20 to Candidate_region{branch=252, mut_idx=0, -1.79769e+308<t<=-2326, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 21 to Candidate_region{branch=253, mut_idx=0, -1.79769e+308<t<=-2327, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 22 to Candidate_region{branch=254, mut_idx=0, -1.79769e+308<t<=-2328, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 23 to Candidate_region{branch=255, mut_idx=0, -1.79769e+308<t<=-2329, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 24 to Candidate_region{branch=256, mut_idx=0, -1.79769e+308<t<=-2330, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 25 to Candidate_region{branch=257, mut_idx=0, -1.79769e+308<t<=-2331, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 26 to Candidate_region{branch=243, mut_idx=0, -2312.77<t<=-2178.75, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 27 to Candidate_region{branch=247, mut_idx=0, -2178.75<t<=-2076.28, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 28 to Candidate_region{branch=258, mut_idx=0, -1.79769e+308<t<=-2332, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 29 to Candidate_region{branch=261, mut_idx=0, -1.79769e+308<t<=-2333, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 30 to Candidate_region{branch=262, mut_idx=0, -1.79769e+308<t<=-2334, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 31 to Candidate_region{branch=263, mut_idx=0, -1.79769e+308<t<=-2335, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 32 to Candidate_region{branch=264, mut_idx=0, -1.79769e+308<t<=-2336, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 33 to Candidate_region{branch=265, mut_idx=0, -1.79769e+308<t<=-2337, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 34 to Candidate_region{branch=33, mut_idx=1, -2172.76<t<=-2032, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 35 to Candidate_region{branch=34, mut_idx=0, -2136.89<t<=-2030, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 36 to Candidate_region{branch=266, mut_idx=0, -1.79769e+308<t<=-2338, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 37 to Candidate_region{branch=27, mut_idx=0, -2077.03<t<=-2066.81, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 38 to Candidate_region{branch=260, mut_idx=0, -2178.75<t<=-2077.03, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 39 to Candidate_region{branch=269, mut_idx=0, -1.79769e+308<t<=-2339, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 40 to Candidate_region{branch=272, mut_idx=0, -1.79769e+308<t<=-2340, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 41 to Candidate_region{branch=273, mut_idx=0, -1.79769e+308<t<=-2341, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 42 to Candidate_region{branch=41, mut_idx=1, -2129.72<t<=-2031, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 43 to Candidate_region{branch=274, mut_idx=0, -1.79769e+308<t<=-2342, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 44 to Candidate_region{branch=31, mut_idx=1, -2221.88<t<=-2032, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 45 to Candidate_region{branch=37, mut_idx=0, -2067.59<t<=-2034, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 46 to Candidate_region{branch=276, mut_idx=0, -1.79769e+308<t<=-2343, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 47 to Candidate_region{branch=279, mut_idx=0, -1.79769e+308<t<=-2344, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 48 to Candidate_region{branch=6, mut_idx=0, -2095.16<t<=-2046, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 49 to Candidate_region{branch=48, mut_idx=2, -2055.91<t<=-2029, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 50 to Candidate_region{branch=49, mut_idx=0, -2032.36<t<=-2028, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 51 to Candidate_region{branch=282, mut_idx=2, -2055.91<t<=-2032.36, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 52 to Candidate_region{branch=25, mut_idx=2, -2255.92<t<=-2038, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 53 to Candidate_region{branch=25, mut_idx=0, -2206.61<t<=-2038, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 54 to Candidate_region{branch=280, mut_idx=0, -1.79769e+308<t<=-2345, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 55 to Candidate_region{branch=271, mut_idx=0, -2178.75<t<=-2129.4, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 56 to Candidate_region{branch=244, mut_idx=0, -2178.75<t<=-2095.16, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 57 to Candidate_region{branch=11, mut_idx=0, -2095.16<t<=-2045, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 58 to Candidate_region{branch=55, mut_idx=0, -2163.74<t<=-2027, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 59 to Candidate_region{branch=287, mut_idx=0, -1.79769e+308<t<=-2346, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 60 to Candidate_region{branch=11, mut_idx=0, -2082.26<t<=-2045, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 61 to Candidate_region{branch=38, mut_idx=0, -2129.4<t<=-2034, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 62 to Candidate_region{branch=292, mut_idx=0, -1.79769e+308<t<=-2347, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 63 to Candidate_region{branch=6, mut_idx=0, -2076.66<t<=-2046, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 64 to Candidate_region{branch=260, mut_idx=0, -2129.4<t<=-2077.03, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 65 to Candidate_region{branch=295, mut_idx=0, -1.79769e+308<t<=-2348, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 66 to Candidate_region{branch=298, mut_idx=0, -1.79769e+308<t<=-2349, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 67 to Candidate_region{branch=56, mut_idx=0, -2128.32<t<=-2027, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 68 to Candidate_region{branch=259, mut_idx=0, -2312.77<t<=-2227.42, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 69 to Candidate_region{branch=299, mut_idx=0, -1.79769e+308<t<=-2350, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 70 to Candidate_region{branch=68, mut_idx=0, -2246.33<t<=-2026, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 71 to Candidate_region{branch=70, mut_idx=2, -2038.69<t<=-2026, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 72 to Candidate_region{branch=302, mut_idx=0, -1.79769e+308<t<=-2351, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 73 to Candidate_region{branch=69, mut_idx=1, -2273.47<t<=-2026, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 74 to Candidate_region{branch=305, mut_idx=0, -1.79769e+308<t<=-2352, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 75 to Candidate_region{branch=307, mut_idx=0, -1.79769e+308<t<=-2353, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 76 to Candidate_region{branch=308, mut_idx=0, -1.79769e+308<t<=-2354, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 77 to Candidate_region{branch=56, mut_idx=0, -2101.85<t<=-2027, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 78 to Candidate_region{branch=306, mut_idx=1, -2273.47<t<=-2138.21, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 79 to Candidate_region{branch=78, mut_idx=0, -2273.14<t<=-2026, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 80 to Candidate_region{branch=259, mut_idx=0, -2246.33<t<=-2227.42, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 81 to Candidate_region{branch=44, mut_idx=0, -2162.21<t<=-2031, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 82 to Candidate_region{branch=31, mut_idx=0, -2162.21<t<=-2032, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 83 to Candidate_region{branch=309, mut_idx=0, -1.79769e+308<t<=-2355, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 84 to Candidate_region{branch=83, mut_idx=1, -2192.19<t<=-2025, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 85 to Candidate_region{branch=31, mut_idx=0, -2148.06<t<=-2032, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 86 to Candidate_region{branch=57, mut_idx=2, -2042.93<t<=-2027, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 87 to Candidate_region{branch=318, mut_idx=0, -2148.06<t<=-2063.13, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 88 to Candidate_region{branch=294, mut_idx=0, -2129.4<t<=-2075.68, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 89 to Candidate_region{branch=316, mut_idx=0, -1.79769e+308<t<=-2356, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 90 to Candidate_region{branch=14, mut_idx=0, -2076.28<t<=-2044, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 91 to Candidate_region{branch=322, mut_idx=0, -1.79769e+308<t<=-2357, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 92 to Candidate_region{branch=324, mut_idx=0, -1.79769e+308<t<=-2358, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 93 to Candidate_region{branch=44, mut_idx=0, -2081.01<t<=-2031, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 94 to Candidate_region{branch=325, mut_idx=0, -1.79769e+308<t<=-2359, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 95 to Candidate_region{branch=277, mut_idx=1, -2221.88<t<=-2162.21, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 96 to Candidate_region{branch=327, mut_idx=0, -1.79769e+308<t<=-2360, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 97 to Candidate_region{branch=96, mut_idx=1, -2029.35<t<=-2023, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 98 to Candidate_region{branch=95, mut_idx=0, -2217.24<t<=-2023, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 99 to Candidate_region{branch=329, mut_idx=0, -1.79769e+308<t<=-2361, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 100 to Candidate_region{branch=332, mut_idx=0, -1.79769e+308<t<=-2362, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 101 to Candidate_region{branch=333, mut_idx=0, -1.79769e+308<t<=-2363, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 102 to Candidate_region{branch=1, mut_idx=0, -2195.11<t<=-2191, min_muts=601, W_over_Wmax=0}\n",
            "Attaching tip 103 to Candidate_region{branch=2, mut_idx=4, -2193.1<t<=-2191, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 104 to Candidate_region{branch=1, mut_idx=0, -2192.1<t<=-2191, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 105 to Candidate_region{branch=237, mut_idx=0, -2223.01<t<=-2182.71, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 106 to Candidate_region{branch=97, mut_idx=0, -2028.74<t<=-2021, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 107 to Candidate_region{branch=61, mut_idx=0, -2075.68<t<=-2026, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 108 to Candidate_region{branch=106, mut_idx=1, -2015.81<t<=-1988.59, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 109 to Candidate_region{branch=82, mut_idx=0, -2148.06<t<=-2024, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 110 to Candidate_region{branch=303, mut_idx=0, -2246.33<t<=-2052.26, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 111 to Candidate_region{branch=110, mut_idx=7, -1934.12<t<=-1923, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 112 to Candidate_region{branch=111, mut_idx=0, -1931.92<t<=-1928.01, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 113 to Candidate_region{branch=112, mut_idx=0, -1930.28<t<=-1908, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 114 to Candidate_region{branch=26, mut_idx=0, -2227.42<t<=-2198.48, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 115 to Candidate_region{branch=114, mut_idx=6, -1929.09<t<=-1896, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 116 to Candidate_region{branch=115, mut_idx=7, -1883.87<t<=-1876, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 117 to Candidate_region{branch=349, mut_idx=7, -1883.87<t<=-1876.69, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 118 to Candidate_region{branch=102, mut_idx=601, -2191<t<=-2191, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 119 to Candidate_region{branch=118, mut_idx=4, -2092.2<t<=-2075.45, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 120 to Candidate_region{branch=334, mut_idx=0, -1.79769e+308<t<=-2364, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 121 to Candidate_region{branch=318, mut_idx=0, -2097.88<t<=-2063.13, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 122 to Candidate_region{branch=82, mut_idx=0, -2140.68<t<=-2024, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 123 to Candidate_region{branch=60, mut_idx=0, -2049.4<t<=-2026, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 124 to Candidate_region{branch=348, mut_idx=1, -2201.93<t<=-2192.53, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 125 to Candidate_region{branch=353, mut_idx=0, -1.79769e+308<t<=-2365, min_muts=18, W_over_Wmax=0}\n",
            "Attaching tip 126 to Candidate_region{branch=125, mut_idx=1, -2388.07<t<=-2371.65, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 127 to Candidate_region{branch=125, mut_idx=17, -2216.84<t<=-2191, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 128 to Candidate_region{branch=109, mut_idx=9, -1957.77<t<=-1942, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 129 to Candidate_region{branch=128, mut_idx=2, -1944.32<t<=-1942, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 130 to Candidate_region{branch=331, mut_idx=0, -2217.24<t<=-2191, min_muts=23, W_over_Wmax=0}\n",
            "Attaching tip 131 to Candidate_region{branch=357, mut_idx=1, -2201.93<t<=-2192.82, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 132 to Candidate_region{branch=98, mut_idx=0, -2143.09<t<=-2023, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 133 to Candidate_region{branch=277, mut_idx=0, -2217.24<t<=-2191, min_muts=13, W_over_Wmax=0}\n",
            "Attaching tip 134 to Candidate_region{branch=363, mut_idx=0, -2217.24<t<=-2191.87, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 135 to Candidate_region{branch=133, mut_idx=1, -2193.82<t<=-2193.77, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 136 to Candidate_region{branch=358, mut_idx=0, -1.79769e+308<t<=-2426, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 137 to Candidate_region{branch=135, mut_idx=3, -2191.84<t<=-2191.75, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 138 to Candidate_region{branch=328, mut_idx=1, -2221.88<t<=-2217.24, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 139 to Candidate_region{branch=366, mut_idx=0, -2217.24<t<=-2193.95, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 140 to Candidate_region{branch=137, mut_idx=0, -2191.75<t<=-2191.58, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 141 to Candidate_region{branch=137, mut_idx=0, -2191.73<t<=-2191.58, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 142 to Candidate_region{branch=138, mut_idx=1, -2215.74<t<=-2214, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 143 to Candidate_region{branch=370, mut_idx=3, -2191.84<t<=-2191.75, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 144 to Candidate_region{branch=134, mut_idx=3, -2204.24<t<=-2199.04, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 145 to Candidate_region{branch=142, mut_idx=5, -2210.97<t<=-2210.74, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 146 to Candidate_region{branch=145, mut_idx=9, -2191.74<t<=-2191, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 147 to Candidate_region{branch=140, mut_idx=4, -2191.27<t<=-2191.25, min_muts=28, W_over_Wmax=0}\n",
            "Attaching tip 148 to Candidate_region{branch=141, mut_idx=8, -2191.15<t<=-2191.15, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 149 to Candidate_region{branch=145, mut_idx=0, -2191.3<t<=-2191, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 150 to Candidate_region{branch=141, mut_idx=1, -2191.15<t<=-2191, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 151 to Candidate_region{branch=363, mut_idx=0, -2214<t<=-2191.87, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 152 to Candidate_region{branch=151, mut_idx=2, -2191.5<t<=-2191, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 153 to Candidate_region{branch=150, mut_idx=5, -2191.04<t<=-2191, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 154 to Candidate_region{branch=277, mut_idx=0, -2193.95<t<=-2191, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 155 to Candidate_region{branch=144, mut_idx=1, -2193.62<t<=-2192.31, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 156 to Candidate_region{branch=139, mut_idx=5, -2197.47<t<=-2195.68, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 157 to Candidate_region{branch=153, mut_idx=2, -2191<t<=-2191, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 158 to Candidate_region{branch=376, mut_idx=2, -2192<t<=-2191.84, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 159 to Candidate_region{branch=369, mut_idx=0, -1.79769e+308<t<=-2427, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 160 to Candidate_region{branch=392, mut_idx=0, -1.79769e+308<t<=-2428, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 161 to Candidate_region{branch=386, mut_idx=5, -2191.04<t<=-2191.01, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 162 to Candidate_region{branch=160, mut_idx=3, -2375.23<t<=-2373.81, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 163 to Candidate_region{branch=154, mut_idx=3, -2191.05<t<=-2191, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 164 to Candidate_region{branch=372, mut_idx=0, -2217.24<t<=-2206.66, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 165 to Candidate_region{branch=381, mut_idx=7, -2191.21<t<=-2191.15, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 166 to Candidate_region{branch=160, mut_idx=3, -2357.18<t<=-2328.37, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 167 to Candidate_region{branch=158, mut_idx=1, -2191.86<t<=-2191.76, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 168 to Candidate_region{branch=393, mut_idx=0, -1.79769e+308<t<=-2429, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 169 to Candidate_region{branch=133, mut_idx=8, -2192.2<t<=-2191.93, min_muts=11, W_over_Wmax=0}\n",
            "Attaching tip 170 to Candidate_region{branch=135, mut_idx=2, -2191.53<t<=-2191, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 171 to Candidate_region{branch=388, mut_idx=1, -2193.62<t<=-2192.73, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 172 to Candidate_region{branch=388, mut_idx=0, -2192.94<t<=-2192.73, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 173 to Candidate_region{branch=374, mut_idx=0, -2191.73<t<=-2191.58, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 174 to Candidate_region{branch=162, mut_idx=2, -2265.11<t<=-2217.13, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 175 to Candidate_region{branch=404, mut_idx=1, -2193.62<t<=-2192.94, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 176 to Candidate_region{branch=142, mut_idx=3, -2199.4<t<=-2196.31, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 177 to Candidate_region{branch=166, mut_idx=6, -2231.7<t<=-2191, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 178 to Candidate_region{branch=384, mut_idx=0, -2214<t<=-2198.15, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 179 to Candidate_region{branch=380, mut_idx=3, -2191.42<t<=-2191.27, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 180 to Candidate_region{branch=143, mut_idx=11, -2191.07<t<=-2191, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 181 to Candidate_region{branch=404, mut_idx=0, -2193.48<t<=-2192.94, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 182 to Candidate_region{branch=407, mut_idx=2, -2265.11<t<=-2221.3, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 183 to Candidate_region{branch=415, mut_idx=2, -2265.11<t<=-2238.46, min_muts=12, W_over_Wmax=0}\n",
            "Attaching tip 184 to Candidate_region{branch=385, mut_idx=1, -2195.8<t<=-2191.5, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 185 to Candidate_region{branch=366, mut_idx=0, -2206.66<t<=-2193.95, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 186 to Candidate_region{branch=417, mut_idx=1, -2195.8<t<=-2195.4, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 187 to Candidate_region{branch=391, mut_idx=1, -2193.51<t<=-2192, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 188 to Candidate_region{branch=140, mut_idx=4, -2191.04<t<=-2191, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 189 to Candidate_region{branch=182, mut_idx=0, -2238.46<t<=-2208.74, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 190 to Candidate_region{branch=176, mut_idx=6, -2191.54<t<=-2191.22, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 191 to Candidate_region{branch=175, mut_idx=4, -2192.49<t<=-2192.38, min_muts=9, W_over_Wmax=0}\n",
            "Attaching tip 192 to Candidate_region{branch=402, mut_idx=1, -2193.77<t<=-2193.57, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 193 to Candidate_region{branch=127, mut_idx=6, -2201.01<t<=-2198.56, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 194 to Candidate_region{branch=126, mut_idx=5, -2302.74<t<=-2291.55, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 195 to Candidate_region{branch=244, mut_idx=0, -2128.32<t<=-2095.16, min_muts=10, W_over_Wmax=0}\n",
            "Attaching tip 196 to Candidate_region{branch=303, mut_idx=0, -2235.19<t<=-2052.26, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 197 to Candidate_region{branch=270, mut_idx=0, -2077.03<t<=-2067.59, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 198 to Candidate_region{branch=104, mut_idx=2, -2135.63<t<=-2132, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 199 to Candidate_region{branch=105, mut_idx=4, -2150.02<t<=-2132, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 200 to Candidate_region{branch=103, mut_idx=2, -2132.8<t<=-2132, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 201 to Candidate_region{branch=198, mut_idx=0, -2134.64<t<=-2121.95, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 202 to Candidate_region{branch=199, mut_idx=0, -2142.61<t<=-2116, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 203 to Candidate_region{branch=172, mut_idx=0, -2192.87<t<=-2192.05, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 204 to Candidate_region{branch=203, mut_idx=12, -1921.54<t<=-1909.32, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 205 to Candidate_region{branch=204, mut_idx=0, -1914.42<t<=-1911.97, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 206 to Candidate_region{branch=437, mut_idx=0, -2192.15<t<=-2167.34, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 207 to Candidate_region{branch=204, mut_idx=8, -1885.49<t<=-1885, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 208 to Candidate_region{branch=80, mut_idx=0, -2244<t<=-2229.69, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 209 to Candidate_region{branch=124, mut_idx=1, -2192.73<t<=-2191.86, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 210 to Candidate_region{branch=344, mut_idx=1, -2225.19<t<=-2191, min_muts=8, W_over_Wmax=0}\n",
            "Attaching tip 211 to Candidate_region{branch=431, mut_idx=0, -2191.13<t<=-2161.55, min_muts=15, W_over_Wmax=0}\n",
            "Attaching tip 212 to Candidate_region{branch=211, mut_idx=15, -1886.77<t<=-1858, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 213 to Candidate_region{branch=445, mut_idx=15, -1886.77<t<=-1864, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 214 to Candidate_region{branch=212, mut_idx=1, -1861.32<t<=-1860.28, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 215 to Candidate_region{branch=214, mut_idx=0, -1860.85<t<=-1855.27, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 216 to Candidate_region{branch=445, mut_idx=0, -1869.59<t<=-1861.64, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 217 to Candidate_region{branch=216, mut_idx=1, -1864.74<t<=-1864, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 218 to Candidate_region{branch=446, mut_idx=9, -1952.4<t<=-1926.66, min_muts=16, W_over_Wmax=0}\n",
            "Attaching tip 219 to Candidate_region{branch=121, mut_idx=0, -2084.29<t<=-2073.55, min_muts=7, W_over_Wmax=0}\n",
            "Attaching tip 220 to Candidate_region{branch=219, mut_idx=5, -2002.4<t<=-1906, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 221 to Candidate_region{branch=220, mut_idx=4, -1907.33<t<=-1906, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 222 to Candidate_region{branch=437, mut_idx=1, -2167.34<t<=-2144.7, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 223 to Candidate_region{branch=209, mut_idx=0, -2192.35<t<=-1971, min_muts=4, W_over_Wmax=0}\n",
            "Attaching tip 224 to Candidate_region{branch=223, mut_idx=3, -1978.88<t<=-1948.06, min_muts=0, W_over_Wmax=0}\n",
            "Attaching tip 225 to Candidate_region{branch=223, mut_idx=0, -1978.44<t<=-1948.06, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 226 to Candidate_region{branch=95, mut_idx=0, -2143.09<t<=-2023, min_muts=2, W_over_Wmax=0}\n",
            "Attaching tip 227 to Candidate_region{branch=196, mut_idx=0, -2232.45<t<=-2100.97, min_muts=6, W_over_Wmax=0}\n",
            "Attaching tip 228 to Candidate_region{branch=227, mut_idx=6, -1849.73<t<=-1848, min_muts=1, W_over_Wmax=0}\n",
            "Attaching tip 229 to Candidate_region{branch=195, mut_idx=0, -2101.81<t<=-2063.41, min_muts=27, W_over_Wmax=0}\n",
            "Attaching tip 230 to Candidate_region{branch=218, mut_idx=16, -1869.73<t<=-1863, min_muts=3, W_over_Wmax=0}\n",
            "Attaching tip 231 to Candidate_region{branch=462, mut_idx=0, -2101.81<t<=-2071.51, min_muts=14, W_over_Wmax=0}\n",
            "Attaching tip 232 to Candidate_region{branch=236, mut_idx=6, -2246.73<t<=-2234.07, min_muts=5, W_over_Wmax=0}\n",
            "Attaching tip 233 to Candidate_region{branch=232, mut_idx=4, -2043.56<t<=-1999.09, min_muts=9, W_over_Wmax=0}\n",
            "Built initial tree with 234 tips\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "0.0 M steps/s, Step 0, log_posterior = -282044.77, log_G = -273606.03, log_coal = -12669.78, log_other_priors = 4231.03, num_muts = 1480, T = 444961.86, t_MRCA = 1992-02-09, n0 = 1095.0, g = 0.0000, mu = 10 * 10^-3 subst / site / year, M_ts = 1160, M_tv = 320, kappa = 1, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6056, 4053, 3753, 5099], num_muts_ab counts: [56, 276, 45, 41, 18, 298, 240, 33, 41, 42, 346, 44]\n",
            "\n",
            "\n",
            "*****\n",
            "REFRESHING PARTITION STENCILS\n",
            "*****\n",
            "\n",
            "\n",
            "\n",
            "*****\n",
            "\n",
            "Resetting coalescent prior resolution to 10.9351 days\n",
            "\n",
            "*******\n",
            "\n",
            "terminate called after throwing an instance of 'boost::wrapexcept<std::overflow_error>'\n",
            "  what():  Error in function boost::math::gamma_q_inv<double>(double, double): Overflow Error\n",
            "\n",
            "CPU times: user 1.68 s, sys: 255 ms, total: 1.94 s\n",
            "Wall time: 2min 13s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "-M6qmuH2uaga"
      },
      "execution_count": 5,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}