{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/broadinstitute/delphy/blob/main/tutorials/delphy_workflow_example.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ej4vLF8GBtT_"
      },
      "source": [
        "<div style=\"text-align: center;\">\n",
        "    <h1 style=\"font-size: 2.5em; color: #2C3E50;\">Running <a href=\"https://github.com/broadinstitute/delphy\" target=\"_blank\">Delphy</a> is as simple as 1, 2, 3 ‚Äì Ebolavirus Example</h1>\n",
        "    <h2 style=\"font-size: 1.5em; color: #34495E;\">Phylogenetic Tree Generation with Delphy on Ebolavirus Sequences</h2>\n",
        "</div>\n",
        "\n",
        "This notebook demonstrates how to generate a phylogenetic tree using [Delphy](https://github.com/broadinstitute/delphy) on viral sequences obtained from the [NCBI Virus Database](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/). Optionally, you can also upload your own sequences to be included in the analysis.\n",
        "\n",
        "We will utilize the following tools:\n",
        "- [**gget**](https://pachterlab.github.io/gget/en/virus.html) to download viral sequences from the [NCBI Virus Database](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/)\n",
        "- [**MAFFT**](https://mafft.cbrc.jp/alignment/server/index.html) for creating a Multiple-Sequence Alignment (MSA)\n",
        "- [**Delphy**](https://github.com/broadinstitute/delphy) to generate the phylogenetic tree\n",
        "\n",
        "In this example notebook, we already filled out the NCBI Virus filters and options below to download and analyze all **complete *Zaire ebolavirus* sequences from *Homo sapiens* hosts collected between January 01, 2014 and June 30, 2014**. All generated files will be saved in the **results/** folder, unless you specify a different output folder below.  \n",
        "\n",
        "Simply **click `Runtime` -> `Run all`** at the top of this notebook.\n",
        "\n",
        "If you encounter any problems or questions while using this notebook, please [report them here](https://github.com/broadinstitute/delphy/issues).\n",
        "\n",
        "Total runtime: ~4 min\n",
        "___\n",
        "___"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Apply filters to download sequences from [NCBI Virus](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/)"
      ],
      "metadata": {
        "id": "sWOSpHI-iGVO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title gget virus filtering options:\n",
        "#@markdown Set any non-boolean filter to `None` to disable it.\n",
        "\n",
        "def arg_str_to_bool(arg):\n",
        "    if arg == \"True\":\n",
        "        return True\n",
        "    elif arg == \"False\":\n",
        "        return False\n",
        "    elif arg == \"None\" or arg == \"\":\n",
        "        return None\n",
        "    else:\n",
        "        return arg\n",
        "\n",
        "# -------------------------\n",
        "# Virus query\n",
        "# -------------------------\n",
        "#@markdown ## **Virus**\n",
        "\n",
        "virus = \"Zaire ebolavirus\"  #@param {type:\"string\"}\n",
        "#@markdown  - Examples: 'Mammarenavirus lassaense' or 'coronaviridae' or 'NC_045512.2' or '142786' (Norovirus taxid).\n",
        "virus = arg_str_to_bool(virus)\n",
        "\n",
        "is_accession = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if `virus` is given as an NCBI accession, space-separated list of accessions, or a path to a text file of accessions (one per line).\n",
        "#@markdown  - For SARS-CoV-2 / Influenza A optimized downloads also check the corresponding box below (see `is_sars_cov2` / `is_alphainfluenza` below)\n",
        "is_accession = arg_str_to_bool(is_accession)\n",
        "\n",
        "download_all_accessions = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - ‚ö†Ô∏è Downloads ALL virus accessions (very large). When True, `virus` is ignored.\n",
        "download_all_accessions = arg_str_to_bool(download_all_accessions)\n",
        "\n",
        "# -------------------------\n",
        "# Host filters\n",
        "# -------------------------\n",
        "#@markdown ## **Host**\n",
        "\n",
        "host = \"homo sapiens\"  #@param {type:\"string\"}\n",
        "#@markdown  - Host organism name OR NCBI Taxonomy ID (e.g., 'human', 'Aedes aegypti', '1335626').\n",
        "#@markdown  - Input `None` to disable filtering by host.\n",
        "host = arg_str_to_bool(host)\n",
        "\n",
        "# -------------------------\n",
        "# Sequence & gene filters\n",
        "# -------------------------\n",
        "#@markdown ## **Sequence & gene filters**\n",
        "\n",
        "annotated = \"None\"  #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - True: only annotated sequences. False: only NOT annotated. None: no filter.\n",
        "annotated = arg_str_to_bool(annotated)\n",
        "\n",
        "refseq_only = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Limit search to RefSeq genomes only (curated).\n",
        "refseq_only = arg_str_to_bool(refseq_only)\n",
        "\n",
        "nuc_completeness = \"complete\"  #@param [\"None\", \"complete\", \"partial\"]\n",
        "#@markdown  - Nucleotide completeness.\n",
        "nuc_completeness = arg_str_to_bool(nuc_completeness)\n",
        "\n",
        "min_seq_length = None  #@param {type:\"raw\"}\n",
        "max_seq_length = None  #@param {type:\"raw\"}\n",
        "min_seq_length = arg_str_to_bool(min_seq_length)\n",
        "max_seq_length = arg_str_to_bool(max_seq_length)\n",
        "\n",
        "max_ambiguous_chars = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Max number of ambiguous nucleotide characters (N's).\n",
        "max_ambiguous_chars = arg_str_to_bool(max_ambiguous_chars)\n",
        "\n",
        "min_gene_count = None  #@param {type:\"raw\"}\n",
        "max_gene_count = None  #@param {type:\"raw\"}\n",
        "min_gene_count = arg_str_to_bool(min_gene_count)\n",
        "max_gene_count = arg_str_to_bool(max_gene_count)\n",
        "\n",
        "min_protein_count = None  #@param {type:\"raw\"}\n",
        "max_protein_count = None  #@param {type:\"raw\"}\n",
        "min_protein_count = arg_str_to_bool(min_protein_count)\n",
        "max_protein_count = arg_str_to_bool(max_protein_count)\n",
        "\n",
        "min_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "max_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "min_mature_peptide_count = arg_str_to_bool(min_mature_peptide_count)\n",
        "max_mature_peptide_count = arg_str_to_bool(max_mature_peptide_count)\n",
        "\n",
        "has_proteins = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Require specific proteins/genes (e.g. \"spike\") or list (e.g. [\"spike\",\"ORF1ab\"]).\n",
        "has_proteins = arg_str_to_bool(has_proteins)\n",
        "\n",
        "proteins_complete = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Only include sequences where all annotated proteins are complete.\n",
        "proteins_complete = arg_str_to_bool(proteins_complete)\n",
        "\n",
        "# -------------------------\n",
        "# Location & submitter filters\n",
        "# -------------------------\n",
        "#@markdown ## **Location & submitter filters**\n",
        "\n",
        "geographic_location = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic location of sample collection (e.g. 'USA', 'Asia').\n",
        "geographic_location = arg_str_to_bool(geographic_location)\n",
        "\n",
        "submitter_country = None  #@param {type:\"string\"}\n",
        "#@markdown  - Country of the sequence submitter.\n",
        "submitter_country = arg_str_to_bool(submitter_country)\n",
        "\n",
        "lab_passaged = \"None\"  #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - True: only lab-passaged. False: exclude lab-passaged. None: no filter.\n",
        "lab_passaged = arg_str_to_bool(lab_passaged)\n",
        "\n",
        "# -------------------------\n",
        "# Date filters\n",
        "# -------------------------\n",
        "#@markdown ## **Dates**\n",
        "#@markdown All dates should be supplied in YYYY-MM-DD format.\n",
        "\n",
        "min_collection_date = \"2014-01-01\"  #@param {type:\"string\"}\n",
        "max_collection_date = \"2014-06-30\"  #@param {type:\"string\"}\n",
        "min_release_date = \"2014-01-01\"  #@param {type:\"string\"}\n",
        "max_release_date = \"None\"  #@param {type:\"string\"}\n",
        "\n",
        "min_collection_date = arg_str_to_bool(min_collection_date)\n",
        "max_collection_date = arg_str_to_bool(max_collection_date)\n",
        "min_release_date = arg_str_to_bool(min_release_date)\n",
        "max_release_date = arg_str_to_bool(max_release_date)\n",
        "\n",
        "# -------------------------\n",
        "# SARS-CoV-2 specific filters / optimizations\n",
        "# -------------------------\n",
        "#@markdown ## **SARS-CoV-2 / Influenza A optimizations**\n",
        "\n",
        "lineage = None  #@param {type:\"string\"}\n",
        "#@markdown  - SARS-CoV-2 lineage filter (e.g. 'B.1.1.7'). (Only meaningful for SARS-CoV-2 queries.)\n",
        "lineage = arg_str_to_bool(lineage)\n",
        "\n",
        "is_sars_cov2 = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Use optimized cached downloads for SARS-CoV-2.\n",
        "is_sars_cov2 = arg_str_to_bool(is_sars_cov2)\n",
        "\n",
        "is_alphainfluenza = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Use optimized cached downloads for Influenza A virus (Alphainfluenza).\n",
        "is_alphainfluenza = arg_str_to_bool(is_alphainfluenza)\n",
        "\n",
        "# -------------------------\n",
        "# Optional GenBank metadata enrichment\n",
        "# -------------------------\n",
        "#@markdown ## **GenBank metadata enrichment**\n",
        "\n",
        "genbank_metadata = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Fetch additional detailed metadata from GenBank into a separate CSV/XML/CSV dumps.\n",
        "genbank_metadata = arg_str_to_bool(genbank_metadata)\n",
        "\n",
        "genbank_batch_size = 200  #@param {type:\"integer\"}\n",
        "#@markdown  - Batch size for GenBank metadata API requests. Larger may be faster but can time out (default: 200).\n",
        "genbank_batch_size = int(genbank_batch_size)\n",
        "\n",
        "# -------------------------\n",
        "# Workflow / output options\n",
        "# -------------------------\n",
        "#@markdown ## **Workflow / output options**\n",
        "\n",
        "keep_temp = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Keep intermediate temporary files.\n",
        "keep_temp = arg_str_to_bool(keep_temp)\n",
        "\n",
        "outfolder = \"results\"  #@param {type:\"string\"}\n",
        "#@markdown  - Output folder path (defaults to current working directory if None).\n",
        "outfolder = arg_str_to_bool(outfolder)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "6cazzIZfiJDU"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Optional: Upload a fasta file with your own sequences to add to the analysis\n",
        "  **1) Click on the folder icon on the left.  \n",
        "  2) Upload your file(s) to the Google Colab server by dragging in your file(s) (or use rightclick -> Upload).  \n",
        "  3) Specify the name of your file(s) here:**"
      ],
      "metadata": {
        "id": "aE8ot-AyiPdC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title FASTA file containing additional sequences\n",
        "\n",
        "fasta_file = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_fasta_file.fa' or 'my_fasta_file.fasta'.\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "E-lg25zave0H"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Metadata\n",
        "\n",
        "#@markdown **Option 1: The metadata is the same for all sequences in your FASTA file**\n",
        "metadata = {'Collection Date': 'YYYY-MM-DD', 'Geo Location': 'South Korea'}  #@param {type:\"raw\"}\n",
        "#@markdown - The 'Collection Date' field is required. Optional: you can add as many additional columns as you wish, e.g. 'Geo Location': 'South Korea'.\n",
        "#@markdown - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "#@markdown **Option 2: Input a CSV file with metadata for each sequence**\n",
        "metadata_csv = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_metadata.csv'. This file must include at least an 'Accession' and 'Collection Date' column.\n",
        "#@markdown  - NOTE: Make sure the IDs in the \"Accession\" column match the IDs of the sequences in the provided FASTA file\n",
        "#@markdown  - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "# Convert empty strings to None\n",
        "fasta_file = arg_str_to_bool(fasta_file)\n",
        "metadata_csv = arg_str_to_bool(metadata_csv)"
      ],
      "metadata": {
        "id": "LKgxMFBgiP64",
        "cellView": "form"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Optional: Specify Delphy arguments"
      ],
      "metadata": {
        "id": "zQF5zZ2dzS60"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Delphy options:\n",
        "mutation_rate = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Virus mutation rate (mutations per site per year), e.g. 0.01. If set to `None`, the mutation rate will be estimated from the generated tree.\n",
        "\n",
        "delphy_steps = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Number of steps to run in the Delphy algorithm (default: 500,000 * number of sequences).\n",
        "\n",
        "delphy_samples = 200 #@param {type:\"integer\"}\n",
        "#@markdown  - Number of logging and tree updates (will log every `delphy_steps/delphy_samples` step).\n",
        "\n",
        "delphy_release = \"1.2.2\"  #@param {type:\"string\"}\n",
        "#@markdown  - Delphy version to use (see https://github.com/broadinstitute/delphy/releases).\n",
        "\n",
        "#@title Delphy options:\n",
        "threads = 2  #@param {type:\"integer\"}\n",
        "#@markdown  - Number of threads to use for the MSA (Mafft) and phylogenetic tree (Delphy) generation."
      ],
      "metadata": {
        "cellView": "form",
        "id": "bEM5nlzZzSih"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# All done! Select `Runtime` at the top of this notebook, then click `Run all` and lean back...\n",
        "A completion message will be displayed below once the notebook has been successfully executed.  \n",
        "**üí° Tip: Click on the folder icon üìÅ on the left to view/download the files that are being generated.**\n",
        "  \n",
        "<br>\n",
        "\n",
        "____\n",
        "____\n",
        "____\n",
        "____"
      ],
      "metadata": {
        "id": "UTL55NF2jAj-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "#@title # Generating the phylogenetic tree...\n",
        "\n",
        "from IPython.display import display, HTML\n",
        "from pathlib import Path\n",
        "import subprocess\n",
        "import sys\n",
        "import shutil\n",
        "import glob\n",
        "import pandas as pd\n",
        "import re\n",
        "import os\n",
        "from datetime import datetime\n",
        "\n",
        "# ---------- Small UI helpers ----------\n",
        "def log_message(text):\n",
        "    display(HTML(f\"<h3 style='color:green;margin:6px 0'>{text}</h3>\"))\n",
        "\n",
        "def log_message_sub(text):\n",
        "    display(HTML(f\"<p style='color:black; font-weight:normal; margin:4px 0;'>{text}</p>\"))\n",
        "\n",
        "def log_message_error(text):\n",
        "    display(HTML(f\"<h3 style='color:red;margin:6px 0'>{text}</h3>\"))\n",
        "\n",
        "def run_checked(cmd, **kwargs):\n",
        "    \"\"\"Run subprocess.run with sensible defaults and return CompletedProcess or raise.\"\"\"\n",
        "    default = dict(check=True, text=True, capture_output=True)\n",
        "    default.update(kwargs)\n",
        "    return subprocess.run(cmd, **default)\n",
        "\n",
        "# ---------- Validate / create outfolder ----------\n",
        "outfolder = Path(outfolder) if 'outfolder' in globals() else Path(\"delphy_out\")\n",
        "outfolder.mkdir(parents=True, exist_ok=True)\n",
        "\n",
        "# ---------- 1) Install Python requirements (gget) ----------\n",
        "log_message(\"1/5 Installing Python dependencies (gget)...\")\n",
        "try:\n",
        "    !pip install --upgrade -q gget biopython\n",
        "    import gget\n",
        "    from Bio import SeqIO\n",
        "    log_message_sub(\"Python dependencies installed.\")\n",
        "except Exception as e:\n",
        "    log_message_error(\"Failed to install Python dependencies (gget).\")\n",
        "    print(e)\n",
        "    raise\n",
        "\n",
        "# ---------- 2) Download sequences from NCBI using gget ----------\n",
        "log_message(\"2/5 Downloading viral sequences from NCBI (gget.virus)... This may take a few minutes.\")\n",
        "try:\n",
        "    # The original code used many parameters assumed present in the notebook; call gget.virus with locals()\n",
        "    # Build kwargs only from variables that exist in globals() and are not None.\n",
        "    possible_kwargs = [\n",
        "        \"virus\",\"is_accession\",\"download_all_accessions\",\"host\",\n",
        "        \"nuc_completeness\",\"min_seq_length\",\"max_seq_length\",\n",
        "        \"min_gene_count\",\"max_gene_count\",\"min_mature_peptide_count\",\"max_mature_peptide_count\",\n",
        "        \"min_protein_count\",\"max_protein_count\",\"max_ambiguous_chars\",\"has_proteins\",\"proteins_complete\",\"annotated\",\"refseq_only\",\n",
        "        \"geographic_location\",\"submitter_country\",\"lab_passaged\",\n",
        "        \"min_collection_date\",\"max_collection_date\",\"min_release_date\",\"max_release_date\",\n",
        "        \"lineage\",\"is_sars_cov2\",\"is_alphainfluenza\",\n",
        "        \"genbank_metadata\",\"genbank_batch_size\",\"keep_temp\",\"outfolder\"\n",
        "    ]\n",
        "    gget_kwargs = {k: globals()[k] for k in possible_kwargs if k in globals() and globals()[k] is not None}\n",
        "    # Ensure outfolder is string path for gget\n",
        "    if \"outfolder\" not in gget_kwargs:\n",
        "        gget_kwargs[\"outfolder\"] = str(outfolder)\n",
        "    else:\n",
        "        gget_kwargs[\"outfolder\"] = str(Path(gget_kwargs[\"outfolder\"]))\n",
        "    # Call gget.virus\n",
        "    gget.virus(**gget_kwargs)\n",
        "\n",
        "    # Find the newest fasta and metadata file in outfolder\n",
        "    fasta_candidates = list(outfolder.glob(\"*.fasta\")) + list(outfolder.glob(\"*.fa\"))\n",
        "    if not fasta_candidates:\n",
        "        raise FileNotFoundError(f\"No fasta files were created in {outfolder}\")\n",
        "    ncbi_fasta_file = max(fasta_candidates, key=lambda p: p.stat().st_mtime)\n",
        "\n",
        "    metadata_candidates = list(outfolder.glob(\"*.csv\")) + list(outfolder.glob(\"*.jsonl\")) + list(outfolder.glob(\"*.tsv\"))\n",
        "    if not metadata_candidates:\n",
        "        raise FileNotFoundError(f\"No metadata files were created in {outfolder}\")\n",
        "    ncbi_metadata = max(metadata_candidates, key=lambda p: p.stat().st_mtime)\n",
        "\n",
        "    log_message_sub(f\"Download complete. FASTA: {ncbi_fasta_file.name}  Metadata: {ncbi_metadata.name}\")\n",
        "except Exception as e:\n",
        "    log_message_error(\"Error while downloading sequences from NCBI.\")\n",
        "    print(e)\n",
        "    raise\n",
        "\n",
        "# ---------- 2b) If user provided FASTA/metadata, merge them ----------\n",
        "input_fasta_file = Path(ncbi_fasta_file)\n",
        "metadata_file = Path(ncbi_metadata)\n",
        "\n",
        "if 'fasta_file' in globals() and fasta_file:\n",
        "    log_message_sub(\"Merging user-provided FASTA with NCBI FASTA...\")\n",
        "    try:\n",
        "        combined_fasta_file = outfolder / f\"{'_'.join(str(virus).split())}_sequences_combined.fasta\"\n",
        "        with open(combined_fasta_file, \"wb\") as outfh:\n",
        "            for src in [ncbi_fasta_file, Path(fasta_file)]:\n",
        "                with open(src, \"rb\") as fh:\n",
        "                    shutil.copyfileobj(fh, outfh)\n",
        "        input_fasta_file = combined_fasta_file\n",
        "\n",
        "        # Combine metadata\n",
        "        combined_metadata_file = outfolder / f\"{'_'.join(str(virus).split())}_metadata_combined.csv\"\n",
        "        ncbi_df = pd.read_csv(ncbi_metadata)\n",
        "\n",
        "        if 'metadata_csv' in globals() and metadata_csv:\n",
        "            user_meta_df = pd.read_csv(metadata_csv)\n",
        "            comb_meta_df = pd.concat([ncbi_df, user_meta_df], ignore_index=True, sort=False)\n",
        "            comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "            metadata_file = combined_metadata_file\n",
        "        else:\n",
        "            # If user provided `metadata` mapping and FASTA only\n",
        "            headers = [record.id.split()[0] for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
        "            user_meta_df = pd.DataFrame({\"accession\": headers})\n",
        "            if 'metadata' in globals() and isinstance(metadata, dict):\n",
        "                for k, v in metadata.items():\n",
        "                    # Broadcast scalar values or accept list-like\n",
        "                    if hasattr(v, \"__len__\") and not isinstance(v, str) and len(v) == len(headers):\n",
        "                        user_meta_df[k] = v\n",
        "                    else:\n",
        "                        user_meta_df[k] = [v] * len(headers)\n",
        "            comb_meta_df = pd.concat([ncbi_df, user_meta_df], ignore_index=True, sort=False)\n",
        "            comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "            metadata_file = combined_metadata_file\n",
        "\n",
        "        log_message_sub(f\"Merging complete. Combined FASTA: {combined_fasta_file.name}  Combined metadata: {combined_metadata_file.name}\")\n",
        "    except Exception as e:\n",
        "        log_message_error(\"Error while merging user-provided data.\")\n",
        "        print(e)\n",
        "        raise\n",
        "\n",
        "# ---------- 3) Multiple Sequence Alignment (MSA) using MAFFT (Colab-friendly, --parttree for >10k) ----------\n",
        "log_message(\"3/5 Multiple Sequence Alignment (MAFFT). This may take time depending on # sequences and threads.\")\n",
        "\n",
        "aligned_fasta_file = outfolder / f\"{'_'.join(str(virus).split())}_aligned.afa\"\n",
        "\n",
        "def count_fasta_records(path: Path):\n",
        "    return sum(1 for _ in SeqIO.parse(str(path), \"fasta\"))\n",
        "\n",
        "# 1) Check for mafft binary\n",
        "def mafft_is_available():\n",
        "    return bool(shutil.which(\"mafft\"))\n",
        "\n",
        "# 2) If not available, attempt the same .deb download + dpkg approach you had for Colab\n",
        "if not mafft_is_available():\n",
        "    try:\n",
        "        log_message_sub(\"MAFFT not found ‚Äî attempting to download & install MAFFT .deb (Colab-compatible).\")\n",
        "        deb_path = outfolder / \"mafft_7.526-1_amd64.deb\"\n",
        "        # download\n",
        "        subprocess.run([\"wget\", \"-q\", \"-O\", str(deb_path),\n",
        "                        \"https://mafft.cbrc.jp/alignment/software/mafft_7.526-1_amd64.deb\"],\n",
        "                       check=True, text=True)\n",
        "        # install\n",
        "        subprocess.run([\"dpkg\", \"-i\", str(deb_path)], check=True, text=True)\n",
        "        # If dpkg left missing deps, try apt-get -f install\n",
        "        subprocess.run([\"apt-get\", \"-y\", \"install\", \"-f\"], check=True, text=True)\n",
        "        if mafft_is_available():\n",
        "            log_message_sub(\"MAFFT installed successfully.\")\n",
        "        else:\n",
        "            log_message_error(\"MAFFT was not found after .deb install. Please install MAFFT manually in the Colab environment.\")\n",
        "            raise FileNotFoundError(\"mafft binary not found after .deb install\")\n",
        "    except subprocess.CalledProcessError as e:\n",
        "        log_message_error(\"An error occurred while attempting to install MAFFT via .deb.\")\n",
        "        # print stderr/stdout if available\n",
        "        try:\n",
        "            print(e.stderr or e.stdout)\n",
        "        except Exception:\n",
        "            print(e)\n",
        "        raise\n",
        "\n",
        "# 3) Prepare MAFFT command, adding --parttree for large datasets\n",
        "num_seqs = count_fasta_records(input_fasta_file)\n",
        "threads = int(threads) if 'threads' in globals() and threads else 1\n",
        "\n",
        "# Build command; place --parttree right after \"mafft\" (recommended)\n",
        "if num_seqs > 10_000:\n",
        "    log_message_sub(f\"Detected {num_seqs:,} sequences ‚Äî using MAFFT with --parttree optimized for large datasets.\")\n",
        "    mafft_cmd = [\"mafft\", \"--parttree\", \"--auto\", \"--thread\", str(threads), str(input_fasta_file)]\n",
        "else:\n",
        "    log_message_sub(f\"Detected {num_seqs:,} sequences ‚Äî using MAFFT standard --auto mode.\")\n",
        "    mafft_cmd = [\"mafft\", \"--auto\", \"--thread\", str(threads), str(input_fasta_file)]\n",
        "\n",
        "# 4) Run MAFFT and write alignment directly to file (no capture_output conflict)\n",
        "try:\n",
        "    log_message_sub(f\"Running: {' '.join(mafft_cmd[:6])} ... (command truncated in log)\")\n",
        "    with open(aligned_fasta_file, \"w\") as outfh:\n",
        "        # IMPORTANT: do NOT use capture_output=True here; provide stdout=outfh instead\n",
        "        subprocess.run(mafft_cmd, stdout=outfh, stderr=subprocess.PIPE, check=True, text=True)\n",
        "    log_message_sub(f\"MSA complete and saved to: {aligned_fasta_file.name}\")\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(\"Error while running MAFFT for MSA. See stderr below:\")\n",
        "    stderr = e.stderr.decode() if isinstance(e.stderr, (bytes, bytearray)) else e.stderr\n",
        "    print(stderr)\n",
        "    raise\n",
        "except Exception as e:\n",
        "    log_message_error(\"Unexpected error while running MAFFT.\")\n",
        "    print(e)\n",
        "    raise\n",
        "\n",
        "# ---------- 4) Reformat FASTA headers to match Delphy format (accession|YYYY-MM-DD) ----------\n",
        "log_message(\"4/5 Reformatting FASTA headers to 'accession|YYYY-MM-DD' (skipping entries without valid collection dates).\")\n",
        "\n",
        "def extract_and_format_date(date_string: str, accept_uncertain_dates: bool = False):\n",
        "    \"\"\"Return YYYY-MM-DD, YYYY-MM, or YYYY depending on accept_uncertain_dates. Return None if unrecognized.\"\"\"\n",
        "    if pd.isna(date_string):\n",
        "        return None\n",
        "    date_string = str(date_string).strip()\n",
        "    # Normalize separators to '-'\n",
        "    date_string = re.sub(r\"[/.]\", \"-\", date_string)\n",
        "    # Try parsing strict YYYY-MM-DD\n",
        "    full_date_re = re.compile(r\"^\\d{4}-\\d{1,2}-\\d{1,2}$\")\n",
        "    year_month_re = re.compile(r\"^\\d{4}-\\d{1,2}$\")\n",
        "    year_re = re.compile(r\"^\\d{4}$\")\n",
        "    if full_date_re.match(date_string):\n",
        "        return date_string\n",
        "    if year_month_re.match(date_string):\n",
        "        return date_string if accept_uncertain_dates else \"EXCLUDED\"\n",
        "    if year_re.match(date_string):\n",
        "        return date_string if accept_uncertain_dates else \"EXCLUDED\"\n",
        "    # Try to parse more flexible formats (e.g., 'Jan 2 2020' or '2 Jan 2020') using datetime\n",
        "    for fmt in (\"%Y-%m-%d\", \"%d-%b-%Y\", \"%d %b %Y\", \"%b %d %Y\", \"%Y/%m/%d\"):\n",
        "        try:\n",
        "            dt = datetime.strptime(date_string, fmt)\n",
        "            return dt.strftime(\"%Y-%m-%d\")\n",
        "        except Exception:\n",
        "            continue\n",
        "    return None\n",
        "\n",
        "def find_date_columns(df: pd.DataFrame):\n",
        "    \"\"\"Heuristic to find which column likely contains collection date and accession.\"\"\"\n",
        "    date_cols = [c for c in df.columns if re.search(r\"date|collection|collected\", c, flags=re.I)]\n",
        "    acc_cols = [c for c in df.columns if re.search(r\"acc(?:ession)?|id|seq_id\", c, flags=re.I)]\n",
        "    # Fallback sensible names\n",
        "    if not date_cols and \"Collection Date\" in df.columns:\n",
        "        date_cols = [\"Collection Date\"]\n",
        "    if not acc_cols and \"accession\" in df.columns:\n",
        "        acc_cols = [\"accession\"]\n",
        "    # Return first hits or None\n",
        "    return (date_cols[0] if date_cols else None, acc_cols[0] if acc_cols else None)\n",
        "\n",
        "def update_fasta_headers(fasta_in: Path, csv_in: Path, fasta_out: Path, accept_uncertain_dates: bool = False):\n",
        "    \"\"\"Rewrite fasta headers to 'accession|date' using CSV metadata. Exclude entries without valid date.\"\"\"\n",
        "    df = pd.read_csv(csv_in, dtype=str)\n",
        "    date_col, acc_col = find_date_columns(df)\n",
        "    if date_col is None or acc_col is None:\n",
        "        raise ValueError(f\"Could not find accession/date columns in {csv_in}. Found columns: {list(df.columns)}\")\n",
        "    accession_to_date = pd.Series(df[date_col].values, index=df[acc_col]).to_dict()\n",
        "\n",
        "    included = 0\n",
        "    excluded = 0\n",
        "    unrecognized = 0\n",
        "\n",
        "    with open(fasta_in) as fh_in, open(fasta_out, \"w\") as fh_out:\n",
        "        for record in SeqIO.parse(fh_in, \"fasta\"):\n",
        "            original_accession = record.id.split()[0]\n",
        "            date_val = accession_to_date.get(original_accession) or accession_to_date.get(original_accession.split(\"|\")[0])\n",
        "            formatted = extract_and_format_date(date_val, accept_uncertain_dates=accept_uncertain_dates)\n",
        "            if formatted == \"EXCLUDED\":\n",
        "                excluded += 1\n",
        "                continue\n",
        "            if formatted is None:\n",
        "                unrecognized += 1\n",
        "                continue\n",
        "            # Set header and remove description\n",
        "            record.id = f\"{original_accession}|{formatted}\"\n",
        "            record.description = \"\"\n",
        "            SeqIO.write(record, fh_out, \"fasta\")\n",
        "            included += 1\n",
        "\n",
        "    return {\"included\": included, \"excluded\": excluded, \"unrecognized\": unrecognized}\n",
        "\n",
        "aligned_fasta_file_clean = outfolder / f\"{'_'.join(str(virus).split())}_aligned_headers_adjusted.afa\"\n",
        "try:\n",
        "    accept_uncertain_dates = bool(globals().get(\"accept_uncertain_dates\", False))\n",
        "    stats = update_fasta_headers(aligned_fasta_file, metadata_file, aligned_fasta_file_clean, accept_uncertain_dates=accept_uncertain_dates)\n",
        "    log_message_sub(f\"Reformatting complete. Included {stats['included']} sequences; excluded (incomplete dates) {stats['excluded']}; unrecognized {stats['unrecognized']}\")\n",
        "except Exception as e:\n",
        "    log_message_error(\"Error while reformatting FASTA headers.\")\n",
        "    print(e)\n",
        "    raise\n",
        "\n",
        "# ---------- 5) Run Delphy ----------\n",
        "log_message(\"5/5 Preparing and running Delphy. This step can be long depending on steps/samples/sequence count.\")\n",
        "\n",
        "def count_fasta_records(path: Path):\n",
        "    return sum(1 for _ in SeqIO.parse(str(path), \"fasta\"))\n",
        "\n",
        "num_seqs = count_fasta_records(aligned_fasta_file_clean)\n",
        "if num_seqs == 0:\n",
        "    raise RuntimeError(f\"No sequences found in {aligned_fasta_file_clean}\")\n",
        "\n",
        "# sensible defaults for delphy params if not provided\n",
        "delphy_steps = globals().get(\"delphy_steps\")\n",
        "delphy_samples = globals().get(\"delphy_samples\", 100)\n",
        "if delphy_steps is None:\n",
        "    delphy_steps = int(500_000 * max(1, num_seqs))\n",
        "\n",
        "log_every = int(delphy_steps / delphy_samples) if delphy_samples else max(1, delphy_steps // 100)\n",
        "tree_every = log_every\n",
        "delphy_snapshot_every = log_every\n",
        "\n",
        "# Download delphy if not present\n",
        "delphy_bin = (Path.cwd() / \"delphy\").resolve()\n",
        "if not delphy_bin.exists():\n",
        "    delphy_release = globals().get(\"delphy_release\", \"1.2.2\")  # user may override\n",
        "    tarball = outfolder / \"delphy_release.tar.gz\"\n",
        "    try:\n",
        "        download_url = f\"https://github.com/broadinstitute/delphy/releases/download/{delphy_release}/delphy-linux-x86_64-ubuntu22.tar.gz\"\n",
        "        log_message_sub(f\"Downloading Delphy release {delphy_release}.\")\n",
        "        run_checked([\"wget\", \"-q\", \"-O\", str(tarball), download_url])\n",
        "        run_checked([\"tar\", \"-xzf\", str(tarball)], check=True)\n",
        "        # Try to find delphy binary in extracted files\n",
        "        found = list(Path(\".\").glob(\"**/delphy\"))\n",
        "        if found:\n",
        "            found[0].chmod(found[0].stat().st_mode | 0o111)\n",
        "            shutil.copy(found[0], delphy_bin)\n",
        "            log_message_sub(\"Delphy binary obtained.\")\n",
        "        else:\n",
        "            log_message_error(\"Delphy binary not found inside downloaded archive.\")\n",
        "            raise FileNotFoundError(\"delphy binary not found\")\n",
        "    except Exception as e:\n",
        "        log_message_error(\"Failed to download/extract Delphy. Please ensure `delphy` binary is available in the working directory.\")\n",
        "        print(e)\n",
        "        raise\n",
        "\n",
        "# Build Delphy command\n",
        "beast_log_out = outfolder / f\"{'_'.join(str(virus).split())}_delphy_beast.log\"\n",
        "delphy_beast_tree_out = outfolder / f\"{'_'.join(str(virus).split())}_delphy_beast.trees\"\n",
        "dphy_out = outfolder / f\"{'_'.join(str(virus).split())}_delphy_out.dphy\"\n",
        "\n",
        "mutation_rate = globals().get(\"mutation_rate\", None)\n",
        "threads = int(globals().get(\"threads\", 1))\n",
        "cmd = [\n",
        "    str(delphy_bin),\n",
        "    \"--v0-log-every\", str(log_every),\n",
        "    \"--v0-tree-every\", str(tree_every),\n",
        "    \"--v0-delphy-snapshot-every\", str(delphy_snapshot_every),\n",
        "    \"--v0-threads\", str(threads),\n",
        "    \"--v0-steps\", str(delphy_steps),\n",
        "    \"--v0-in-fasta\", str(aligned_fasta_file_clean),\n",
        "    \"--v0-out-log-file\", str(beast_log_out),\n",
        "    \"--v0-out-trees-file\", str(delphy_beast_tree_out),\n",
        "    \"--v0-out-delphy-file\", str(dphy_out),\n",
        "]\n",
        "if mutation_rate:\n",
        "    cmd = [str(delphy_bin), \"--v0-fix-mutation-rate\", \"--v0-init-mutation-rate\", str(mutation_rate)] + cmd[1:]\n",
        "\n",
        "log_message_sub(f\"Running Delphy with command: {' '.join(cmd[:8])} ... (truncated)\")\n",
        "\n",
        "try:\n",
        "    # Run delphy and stream output to notebook\n",
        "    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True, bufsize=1)\n",
        "    for line in proc.stdout:\n",
        "        # print lines incrementally so the notebook shows progress\n",
        "        print(line, end=\"\")\n",
        "    ret = proc.wait()\n",
        "    if ret != 0:\n",
        "        raise subprocess.CalledProcessError(ret, cmd)\n",
        "    # Success\n",
        "    display(HTML(\"\"\"\n",
        "    <h2 style='color:green'>All done! üéâ</h2>\n",
        "    <p>Files generated in this notebook are in the left-hand file browser (or in the folder you set as <code>outfolder</code>).</p>\n",
        "    <ul>\n",
        "      <li>Aligned FASTA: <code>{}</code></li>\n",
        "      <li>Delphy output (.dphy): <code>{}</code></li>\n",
        "      <li>Delphy log (.log): <code>{}</code></li>\n",
        "      <li>Delphy trees (.trees): <code>{}</code></li>\n",
        "    </ul>\n",
        "    <p>We recommend downloading the <code>.afa</code>, <code>.dphy</code>, and <code>metadata.csv</code> files. To visualize .dphy files online, try <a href='https://delphy.fathom.info/' target='_blank'>delphy.fathom.info</a>.</p>\n",
        "    \"\"\".format(aligned_fasta_file_clean.name, dphy_out.name, beast_log_out.name, delphy_beast_tree_out.name)))\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(\"An error occurred while running Delphy. See captured output above for details.\")\n",
        "    print(e)\n",
        "    raise\n",
        "except Exception as e:\n",
        "    log_message_error(\"Unexpected error while running Delphy.\")\n",
        "    print(e)\n",
        "    raise\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "HuVr7cb0x5VF",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "8aaae078-a757-4e11-e378-67e62f7f2ea5"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>1/5 Installing Python dependencies (gget)...</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Python dependencies installed.</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>2/5 Downloading viral sequences from NCBI (gget.virus)... This may take a few minutes.</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "INFO:gget.utils:Starting virus data retrieval process...\n",
            "INFO:gget.utils:Query parameters: virus='Zaire ebolavirus', is_accession=False, outfolder='results'\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 1: VALIDATING INPUT ARGUMENTS AND OUTPUT DIRECTORY SETUP...\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Input validation completed successfully\n",
            "INFO:gget.utils:Using specified output folder: results\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 2: CHECKING FOR SARS-CoV-2 AND INFLUENZA A QUERIES TO APPLY OPTIMIZED CACHED PATHWAY\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils: Skipping this step. No SARS-CoV-2 query detected.\n",
            "INFO:gget.utils: Skipping this step. No Alphainfluenza query detected.\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 3: Fetching virus metadata from NCBI API\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Streaming API metadata to temporary file: results/tmp_20260213_002733_744688/gget_metadata_20260213_002733_744688.jsonl\n",
            "INFO:gget.utils:Successfully retrieved 416 virus records from NCBI API across 1 pages\n",
            "INFO:gget.utils:Temporary metadata file size: 1.23 MB\n",
            "INFO:gget.utils:Processed 416 metadata records, skipped 0 records without accessions\n",
            "INFO:gget.utils:‚úÖ Saved API metadata JSONL: results/Zaire_ebolavirus_api_metadata.jsonl\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 4: Applying metadata-only filters\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Starting metadata-only filtering process...\n",
            "INFO:gget.utils:Processing 416 metadata records...\n",
            "INFO:gget.utils:‚úÖ Metadata-only filtering process complete.\n",
            "INFO:gget.utils:=================================\n",
            "INFO:gget.utils:Metadata-only filtering complete:\n",
            "INFO:gget.utils:  Total metadata records: 416\n",
            "INFO:gget.utils:  Records passing filters: 180\n",
            "INFO:gget.utils:Filter statistics (records excluded):\n",
            "INFO:gget.utils:  collection_date: 236 records\n",
            "INFO:gget.utils:‚úÖ Saved filtered metadata JSONL: results/Zaire_ebolavirus_metadata.jsonl\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 5: Downloading sequences for filtered accessions\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Downloading sequences for 180 accessions using E-utilities API\n",
            "INFO:gget.utils:Initiating E-utilities request for 180 accessions\n",
            "INFO:gget.utils:Received 180 sequences from E-utilities\n",
            "INFO:gget.utils:Successfully saved sequences to: results/tmp_20260213_002733_744688/virus_sequences_20260213_002733_744688.fasta (3.31 MB)\n",
            "INFO:gget.utils:Downloaded FASTA file: results/tmp_20260213_002733_744688/virus_sequences_20260213_002733_744688.fasta (3.31 MB)\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 6: Applying sequence-dependent filters and saving results\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:No sequence-dependent filters specified, skipping this step.\n",
            "INFO:gget.utils:All 180 downloaded sequences will be saved\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 7: Saving final output files\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Saving 180 filtered sequences and their metadata...\n",
            "INFO:gget.utils:‚úÖ FASTA file saved: results/Zaire_ebolavirus_sequences.fasta (3.31 MB)\n",
            "INFO:gget.utils:‚úÖ JSONL metadata file saved: results/Zaire_ebolavirus_metadata.jsonl (0.43 MB)\n",
            "INFO:gget.utils:Preparing metadata for CSV output...\n",
            "INFO:gget.utils:Processing metadata records...\n",
            "INFO:gget.utils:Creating DataFrame with 180 rows and 39 columns\n",
            "INFO:gget.utils:Successfully saved metadata CSV to: results/Zaire_ebolavirus_metadata.csv\n",
            "INFO:gget.utils:‚úÖ CSV metadata file saved: results/Zaire_ebolavirus_metadata.csv (0.37 MB)\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 8: Fetching detailed GenBank metadata\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Skipping this step since GenBank metadata retrieval was not requested.\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:PROCESS COMPLETED SUCCESSFULLY\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Results summary:\n",
            "INFO:gget.utils:  Total sequences (API metadata): 416\n",
            "INFO:gget.utils:  After metadata-only filtering: 180\n",
            "INFO:gget.utils:  After all filters (final): 180\n",
            "INFO:gget.utils:\n",
            "INFO:gget.utils:Output files saved to: results\n",
            "INFO:gget.utils:  üìÑ Sequences (FASTA): Zaire_ebolavirus_sequences.fasta\n",
            "INFO:gget.utils:  üìä Metadata (CSV):    Zaire_ebolavirus_metadata.csv\n",
            "INFO:gget.utils:  üîß Metadata (JSONL):  Zaire_ebolavirus_metadata.jsonl\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:‚úÖ Command summary saved: results/command_summary.txt\n",
            "INFO:gget.utils:NCBI virus data retrieval process completed.\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Download complete. FASTA: Zaire_ebolavirus_sequences.fasta  Metadata: Zaire_ebolavirus_metadata.csv</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>3/5 Multiple Sequence Alignment (MAFFT). This may take time depending on # sequences and threads.</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Detected 180 sequences ‚Äî using MAFFT standard --auto mode.</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Running: mafft --auto --thread 2 results/Zaire_ebolavirus_sequences.fasta ... (command truncated in log)</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>MSA complete and saved to: Zaire_ebolavirus_aligned.afa</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>4/5 Reformatting FASTA headers to 'accession|YYYY-MM-DD' (skipping entries without valid collection dates).</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Reformatting complete. Included 180 sequences; excluded (incomplete dates) 0; unrecognized 0</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>5/5 Preparing and running Delphy. This step can be long depending on steps/samples/sequence count.</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Running Delphy with command: /content/delphy --v0-log-every 450000 --v0-tree-every 450000 --v0-delphy-snapshot-every 450000 --v0-threads ... (truncated)</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# Seed: 2733193560\n",
            "Reading fasta file results/Zaire_ebolavirus_aligned_headers_adjusted.afa\n",
            "- read 1 sequences so far (19321 of 3473640 bytes = 0.6%)\n",
            "- read 2 sequences so far (38619 of 3473640 bytes = 1.1%)\n",
            "- read 3 sequences so far (57917 of 3473640 bytes = 1.7%)\n",
            "- read 4 sequences so far (77215 of 3473640 bytes = 2.2%)\n",
            "- read 5 sequences so far (96513 of 3473640 bytes = 2.8%)\n",
            "- read 6 sequences so far (115811 of 3473640 bytes = 3.3%)\n",
            "- read 7 sequences so far (135109 of 3473640 bytes = 3.9%)\n",
            "- read 8 sequences so far (154407 of 3473640 bytes = 4.4%)\n",
            "- read 9 sequences so far (173705 of 3473640 bytes = 5.0%)\n",
            "- read 10 sequences so far (193003 of 3473640 bytes = 5.6%)\n",
            "- read 11 sequences so far (212301 of 3473640 bytes = 6.1%)\n",
            "- read 12 sequences so far (231599 of 3473640 bytes = 6.7%)\n",
            "- read 13 sequences so far (250897 of 3473640 bytes = 7.2%)\n",
            "- read 14 sequences so far (270195 of 3473640 bytes = 7.8%)\n",
            "- read 15 sequences so far (289493 of 3473640 bytes = 8.3%)\n",
            "- read 16 sequences so far (308791 of 3473640 bytes = 8.9%)\n",
            "- read 17 sequences so far (328089 of 3473640 bytes = 9.4%)\n",
            "- read 18 sequences so far (347387 of 3473640 bytes = 10.0%)\n",
            "- read 19 sequences so far (366685 of 3473640 bytes = 10.6%)\n",
            "- read 20 sequences so far (385983 of 3473640 bytes = 11.1%)\n",
            "- read 21 sequences so far (405281 of 3473640 bytes = 11.7%)\n",
            "- read 22 sequences so far (424579 of 3473640 bytes = 12.2%)\n",
            "- read 23 sequences so far (443877 of 3473640 bytes = 12.8%)\n",
            "- read 24 sequences so far (463175 of 3473640 bytes = 13.3%)\n",
            "- read 25 sequences so far (482473 of 3473640 bytes = 13.9%)\n",
            "- read 26 sequences so far (501771 of 3473640 bytes = 14.4%)\n",
            "- read 27 sequences so far (521069 of 3473640 bytes = 15.0%)\n",
            "- read 28 sequences so far (540367 of 3473640 bytes = 15.6%)\n",
            "- read 29 sequences so far (559665 of 3473640 bytes = 16.1%)\n",
            "- read 30 sequences so far (578963 of 3473640 bytes = 16.7%)\n",
            "- read 31 sequences so far (598261 of 3473640 bytes = 17.2%)\n",
            "- read 32 sequences so far (617559 of 3473640 bytes = 17.8%)\n",
            "- read 33 sequences so far (636857 of 3473640 bytes = 18.3%)\n",
            "- read 34 sequences so far (656155 of 3473640 bytes = 18.9%)\n",
            "- read 35 sequences so far (675453 of 3473640 bytes = 19.4%)\n",
            "- read 36 sequences so far (694751 of 3473640 bytes = 20.0%)\n",
            "- read 37 sequences so far (714049 of 3473640 bytes = 20.6%)\n",
            "- read 38 sequences so far (733347 of 3473640 bytes = 21.1%)\n",
            "- read 39 sequences so far (752645 of 3473640 bytes = 21.7%)\n",
            "- read 40 sequences so far (771943 of 3473640 bytes = 22.2%)\n",
            "- read 41 sequences so far (791241 of 3473640 bytes = 22.8%)\n",
            "- read 42 sequences so far (810539 of 3473640 bytes = 23.3%)\n",
            "- read 43 sequences so far (829837 of 3473640 bytes = 23.9%)\n",
            "- read 44 sequences so far (849135 of 3473640 bytes = 24.4%)\n",
            "- read 45 sequences so far (868433 of 3473640 bytes = 25.0%)\n",
            "- read 46 sequences so far (887731 of 3473640 bytes = 25.6%)\n",
            "- read 47 sequences so far (907029 of 3473640 bytes = 26.1%)\n",
            "- read 48 sequences so far (926327 of 3473640 bytes = 26.7%)\n",
            "- read 49 sequences so far (945625 of 3473640 bytes = 27.2%)\n",
            "- read 50 sequences so far (964923 of 3473640 bytes = 27.8%)\n",
            "- read 51 sequences so far (984221 of 3473640 bytes = 28.3%)\n",
            "- read 52 sequences so far (1003519 of 3473640 bytes = 28.9%)\n",
            "- read 53 sequences so far (1022817 of 3473640 bytes = 29.4%)\n",
            "- read 54 sequences so far (1042115 of 3473640 bytes = 30.0%)\n",
            "- read 55 sequences so far (1061413 of 3473640 bytes = 30.6%)\n",
            "- read 56 sequences so far (1080711 of 3473640 bytes = 31.1%)\n",
            "- read 57 sequences so far (1100009 of 3473640 bytes = 31.7%)\n",
            "- read 58 sequences so far (1119307 of 3473640 bytes = 32.2%)\n",
            "- read 59 sequences so far (1138605 of 3473640 bytes = 32.8%)\n",
            "- read 60 sequences so far (1157903 of 3473640 bytes = 33.3%)\n",
            "- read 61 sequences so far (1177201 of 3473640 bytes = 33.9%)\n",
            "- read 62 sequences so far (1196499 of 3473640 bytes = 34.4%)\n",
            "- read 63 sequences so far (1215797 of 3473640 bytes = 35.0%)\n",
            "- read 64 sequences so far (1235095 of 3473640 bytes = 35.6%)\n",
            "- read 65 sequences so far (1254393 of 3473640 bytes = 36.1%)\n",
            "- read 66 sequences so far (1273691 of 3473640 bytes = 36.7%)\n",
            "- read 67 sequences so far (1292989 of 3473640 bytes = 37.2%)\n",
            "- read 68 sequences so far (1312287 of 3473640 bytes = 37.8%)\n",
            "- read 69 sequences so far (1331585 of 3473640 bytes = 38.3%)\n",
            "- read 70 sequences so far (1350883 of 3473640 bytes = 38.9%)\n",
            "- read 71 sequences so far (1370181 of 3473640 bytes = 39.4%)\n",
            "- read 72 sequences so far (1389479 of 3473640 bytes = 40.0%)\n",
            "- read 73 sequences so far (1408777 of 3473640 bytes = 40.6%)\n",
            "- read 74 sequences so far (1428075 of 3473640 bytes = 41.1%)\n",
            "- read 75 sequences so far (1447373 of 3473640 bytes = 41.7%)\n",
            "- read 76 sequences so far (1466671 of 3473640 bytes = 42.2%)\n",
            "- read 77 sequences so far (1485969 of 3473640 bytes = 42.8%)\n",
            "- read 78 sequences so far (1505267 of 3473640 bytes = 43.3%)\n",
            "- read 79 sequences so far (1524565 of 3473640 bytes = 43.9%)\n",
            "- read 80 sequences so far (1543863 of 3473640 bytes = 44.4%)\n",
            "- read 81 sequences so far (1563161 of 3473640 bytes = 45.0%)\n",
            "- read 82 sequences so far (1582459 of 3473640 bytes = 45.6%)\n",
            "- read 83 sequences so far (1601757 of 3473640 bytes = 46.1%)\n",
            "- read 84 sequences so far (1621055 of 3473640 bytes = 46.7%)\n",
            "- read 85 sequences so far (1640353 of 3473640 bytes = 47.2%)\n",
            "- read 86 sequences so far (1659651 of 3473640 bytes = 47.8%)\n",
            "- read 87 sequences so far (1678949 of 3473640 bytes = 48.3%)\n",
            "- read 88 sequences so far (1698247 of 3473640 bytes = 48.9%)\n",
            "- read 89 sequences so far (1717545 of 3473640 bytes = 49.4%)\n",
            "- read 90 sequences so far (1736843 of 3473640 bytes = 50.0%)\n",
            "- read 91 sequences so far (1756141 of 3473640 bytes = 50.6%)\n",
            "- read 92 sequences so far (1775439 of 3473640 bytes = 51.1%)\n",
            "- read 93 sequences so far (1794737 of 3473640 bytes = 51.7%)\n",
            "- read 94 sequences so far (1814035 of 3473640 bytes = 52.2%)\n",
            "- read 95 sequences so far (1833333 of 3473640 bytes = 52.8%)\n",
            "- read 96 sequences so far (1852631 of 3473640 bytes = 53.3%)\n",
            "- read 97 sequences so far (1871929 of 3473640 bytes = 53.9%)\n",
            "- read 98 sequences so far (1891227 of 3473640 bytes = 54.4%)\n",
            "- read 99 sequences so far (1910525 of 3473640 bytes = 55.0%)\n",
            "- read 100 sequences so far (1929823 of 3473640 bytes = 55.6%)\n",
            "- read 101 sequences so far (1949121 of 3473640 bytes = 56.1%)\n",
            "- read 102 sequences so far (1968419 of 3473640 bytes = 56.7%)\n",
            "- read 103 sequences so far (1987717 of 3473640 bytes = 57.2%)\n",
            "- read 104 sequences so far (2007015 of 3473640 bytes = 57.8%)\n",
            "- read 105 sequences so far (2026313 of 3473640 bytes = 58.3%)\n",
            "- read 106 sequences so far (2045611 of 3473640 bytes = 58.9%)\n",
            "- read 107 sequences so far (2064909 of 3473640 bytes = 59.4%)\n",
            "- read 108 sequences so far (2084207 of 3473640 bytes = 60.0%)\n",
            "- read 109 sequences so far (2103505 of 3473640 bytes = 60.6%)\n",
            "- read 110 sequences so far (2122803 of 3473640 bytes = 61.1%)\n",
            "- read 111 sequences so far (2142101 of 3473640 bytes = 61.7%)\n",
            "- read 112 sequences so far (2161399 of 3473640 bytes = 62.2%)\n",
            "- read 113 sequences so far (2180697 of 3473640 bytes = 62.8%)\n",
            "- read 114 sequences so far (2199995 of 3473640 bytes = 63.3%)\n",
            "- read 115 sequences so far (2219293 of 3473640 bytes = 63.9%)\n",
            "- read 116 sequences so far (2238591 of 3473640 bytes = 64.4%)\n",
            "- read 117 sequences so far (2257889 of 3473640 bytes = 65.0%)\n",
            "- read 118 sequences so far (2277187 of 3473640 bytes = 65.6%)\n",
            "- read 119 sequences so far (2296485 of 3473640 bytes = 66.1%)\n",
            "- read 120 sequences so far (2315783 of 3473640 bytes = 66.7%)\n",
            "- read 121 sequences so far (2335081 of 3473640 bytes = 67.2%)\n",
            "- read 122 sequences so far (2354379 of 3473640 bytes = 67.8%)\n",
            "- read 123 sequences so far (2373677 of 3473640 bytes = 68.3%)\n",
            "- read 124 sequences so far (2392975 of 3473640 bytes = 68.9%)\n",
            "- read 125 sequences so far (2412273 of 3473640 bytes = 69.4%)\n",
            "- read 126 sequences so far (2431571 of 3473640 bytes = 70.0%)\n",
            "- read 127 sequences so far (2450869 of 3473640 bytes = 70.6%)\n",
            "- read 128 sequences so far (2470167 of 3473640 bytes = 71.1%)\n",
            "- read 129 sequences so far (2489465 of 3473640 bytes = 71.7%)\n",
            "- read 130 sequences so far (2508763 of 3473640 bytes = 72.2%)\n",
            "- read 131 sequences so far (2528061 of 3473640 bytes = 72.8%)\n",
            "- read 132 sequences so far (2547359 of 3473640 bytes = 73.3%)\n",
            "- read 133 sequences so far (2566657 of 3473640 bytes = 73.9%)\n",
            "- read 134 sequences so far (2585955 of 3473640 bytes = 74.4%)\n",
            "- read 135 sequences so far (2605253 of 3473640 bytes = 75.0%)\n",
            "- read 136 sequences so far (2624551 of 3473640 bytes = 75.6%)\n",
            "- read 137 sequences so far (2643849 of 3473640 bytes = 76.1%)\n",
            "- read 138 sequences so far (2663147 of 3473640 bytes = 76.7%)\n",
            "- read 139 sequences so far (2682445 of 3473640 bytes = 77.2%)\n",
            "- read 140 sequences so far (2701743 of 3473640 bytes = 77.8%)\n",
            "- read 141 sequences so far (2721041 of 3473640 bytes = 78.3%)\n",
            "- read 142 sequences so far (2740339 of 3473640 bytes = 78.9%)\n",
            "- read 143 sequences so far (2759637 of 3473640 bytes = 79.4%)\n",
            "- read 144 sequences so far (2778935 of 3473640 bytes = 80.0%)\n",
            "- read 145 sequences so far (2798233 of 3473640 bytes = 80.6%)\n",
            "- read 146 sequences so far (2817531 of 3473640 bytes = 81.1%)\n",
            "- read 147 sequences so far (2836829 of 3473640 bytes = 81.7%)\n",
            "- read 148 sequences so far (2856127 of 3473640 bytes = 82.2%)\n",
            "- read 149 sequences so far (2875425 of 3473640 bytes = 82.8%)\n",
            "- read 150 sequences so far (2894723 of 3473640 bytes = 83.3%)\n",
            "- read 151 sequences so far (2914021 of 3473640 bytes = 83.9%)\n",
            "- read 152 sequences so far (2933319 of 3473640 bytes = 84.4%)\n",
            "- read 153 sequences so far (2952617 of 3473640 bytes = 85.0%)\n",
            "- read 154 sequences so far (2971915 of 3473640 bytes = 85.6%)\n",
            "- read 155 sequences so far (2991213 of 3473640 bytes = 86.1%)\n",
            "- read 156 sequences so far (3010511 of 3473640 bytes = 86.7%)\n",
            "- read 157 sequences so far (3029809 of 3473640 bytes = 87.2%)\n",
            "- read 158 sequences so far (3049107 of 3473640 bytes = 87.8%)\n",
            "- read 159 sequences so far (3068405 of 3473640 bytes = 88.3%)\n",
            "- read 160 sequences so far (3087703 of 3473640 bytes = 88.9%)\n",
            "- read 161 sequences so far (3107001 of 3473640 bytes = 89.4%)\n",
            "- read 162 sequences so far (3126299 of 3473640 bytes = 90.0%)\n",
            "- read 163 sequences so far (3145597 of 3473640 bytes = 90.6%)\n",
            "- read 164 sequences so far (3164895 of 3473640 bytes = 91.1%)\n",
            "- read 165 sequences so far (3184193 of 3473640 bytes = 91.7%)\n",
            "- read 166 sequences so far (3203491 of 3473640 bytes = 92.2%)\n",
            "- read 167 sequences so far (3222789 of 3473640 bytes = 92.8%)\n",
            "- read 168 sequences so far (3242087 of 3473640 bytes = 93.3%)\n",
            "- read 169 sequences so far (3261385 of 3473640 bytes = 93.9%)\n",
            "- read 170 sequences so far (3280683 of 3473640 bytes = 94.4%)\n",
            "- read 171 sequences so far (3299981 of 3473640 bytes = 95.0%)\n",
            "- read 172 sequences so far (3319279 of 3473640 bytes = 95.6%)\n",
            "- read 173 sequences so far (3338577 of 3473640 bytes = 96.1%)\n",
            "- read 174 sequences so far (3357875 of 3473640 bytes = 96.7%)\n",
            "- read 175 sequences so far (3377173 of 3473640 bytes = 97.2%)\n",
            "- read 176 sequences so far (3396471 of 3473640 bytes = 97.8%)\n",
            "- read 177 sequences so far (3415769 of 3473640 bytes = 98.3%)\n",
            "- read 178 sequences so far (3435067 of 3473640 bytes = 98.9%)\n",
            "- read 179 sequences so far (3454365 of 3473640 bytes = 99.4%)\n",
            "- read 180 sequences so far (3473640 of 3473640 bytes = 100.0%)\n",
            "Analysing fasta file results/Zaire_ebolavirus_aligned_headers_adjusted.afa\n",
            "- analysed 0 / 180 sequences\n",
            "- analysed 1 / 180 sequences\n",
            "- analysed 2 / 180 sequences\n",
            "- analysed 3 / 180 sequences\n",
            "- analysed 4 / 180 sequences\n",
            "- analysed 5 / 180 sequences\n",
            "- analysed 6 / 180 sequences\n",
            "- analysed 7 / 180 sequences\n",
            "- analysed 8 / 180 sequences\n",
            "- analysed 9 / 180 sequences\n",
            "- analysed 10 / 180 sequences\n",
            "- analysed 11 / 180 sequences\n",
            "- analysed 12 / 180 sequences\n",
            "- analysed 13 / 180 sequences\n",
            "- analysed 14 / 180 sequences\n",
            "- analysed 15 / 180 sequences\n",
            "- analysed 16 / 180 sequences\n",
            "- analysed 17 / 180 sequences\n",
            "- analysed 18 / 180 sequences\n",
            "- analysed 19 / 180 sequences\n",
            "- analysed 20 / 180 sequences\n",
            "- analysed 21 / 180 sequences\n",
            "- analysed 22 / 180 sequences\n",
            "- analysed 23 / 180 sequences\n",
            "- analysed 24 / 180 sequences\n",
            "- analysed 25 / 180 sequences\n",
            "- analysed 26 / 180 sequences\n",
            "- analysed 27 / 180 sequences\n",
            "- analysed 28 / 180 sequences\n",
            "- analysed 29 / 180 sequences\n",
            "- analysed 30 / 180 sequences\n",
            "- analysed 31 / 180 sequences\n",
            "- analysed 32 / 180 sequences\n",
            "- analysed 33 / 180 sequences\n",
            "- analysed 34 / 180 sequences\n",
            "- analysed 35 / 180 sequences\n",
            "- analysed 36 / 180 sequences\n",
            "- analysed 37 / 180 sequences\n",
            "- analysed 38 / 180 sequences\n",
            "- analysed 39 / 180 sequences\n",
            "- analysed 40 / 180 sequences\n",
            "- analysed 41 / 180 sequences\n",
            "- analysed 42 / 180 sequences\n",
            "- analysed 43 / 180 sequences\n",
            "- analysed 44 / 180 sequences\n",
            "- analysed 45 / 180 sequences\n",
            "- analysed 46 / 180 sequences\n",
            "- analysed 47 / 180 sequences\n",
            "- analysed 48 / 180 sequences\n",
            "- analysed 49 / 180 sequences\n",
            "- analysed 50 / 180 sequences\n",
            "- analysed 51 / 180 sequences\n",
            "- analysed 52 / 180 sequences\n",
            "- analysed 53 / 180 sequences\n",
            "- analysed 54 / 180 sequences\n",
            "- analysed 55 / 180 sequences\n",
            "- analysed 56 / 180 sequences\n",
            "- analysed 57 / 180 sequences\n",
            "- analysed 58 / 180 sequences\n",
            "- analysed 59 / 180 sequences\n",
            "- analysed 60 / 180 sequences\n",
            "- analysed 61 / 180 sequences\n",
            "- analysed 62 / 180 sequences\n",
            "- analysed 63 / 180 sequences\n",
            "- analysed 64 / 180 sequences\n",
            "- analysed 65 / 180 sequences\n",
            "- analysed 66 / 180 sequences\n",
            "- analysed 67 / 180 sequences\n",
            "- analysed 68 / 180 sequences\n",
            "- analysed 69 / 180 sequences\n",
            "- analysed 70 / 180 sequences\n",
            "- analysed 71 / 180 sequences\n",
            "- analysed 72 / 180 sequences\n",
            "- analysed 73 / 180 sequences\n",
            "- analysed 74 / 180 sequences\n",
            "- analysed 75 / 180 sequences\n",
            "WARNING (sequence 'KP096421.1|2014-11-10'): ambiguous state Y at site 9267 changed to N\n",
            "- analysed 76 / 180 sequences\n",
            "- analysed 77 / 180 sequences\n",
            "- analysed 78 / 180 sequences\n",
            "- analysed 79 / 180 sequences\n",
            "- analysed 80 / 180 sequences\n",
            "- analysed 81 / 180 sequences\n",
            "- analysed 82 / 180 sequences\n",
            "- analysed 83 / 180 sequences\n",
            "- analysed 84 / 180 sequences\n",
            "- analysed 85 / 180 sequences\n",
            "- analysed 86 / 180 sequences\n",
            "- analysed 87 / 180 sequences\n",
            "- analysed 88 / 180 sequences\n",
            "- analysed 89 / 180 sequences\n",
            "- analysed 90 / 180 sequences\n",
            "- analysed 91 / 180 sequences\n",
            "- analysed 92 / 180 sequences\n",
            "- analysed 93 / 180 sequences\n",
            "- analysed 94 / 180 sequences\n",
            "- analysed 95 / 180 sequences\n",
            "- analysed 96 / 180 sequences\n",
            "- analysed 97 / 180 sequences\n",
            "- analysed 98 / 180 sequences\n",
            "- analysed 99 / 180 sequences\n",
            "- analysed 100 / 180 sequences\n",
            "- analysed 101 / 180 sequences\n",
            "- analysed 102 / 180 sequences\n",
            "- analysed 103 / 180 sequences\n",
            "- analysed 104 / 180 sequences\n",
            "- analysed 105 / 180 sequences\n",
            "- analysed 106 / 180 sequences\n",
            "- analysed 107 / 180 sequences\n",
            "- analysed 108 / 180 sequences\n",
            "- analysed 109 / 180 sequences\n",
            "- analysed 110 / 180 sequences\n",
            "- analysed 111 / 180 sequences\n",
            "- analysed 112 / 180 sequences\n",
            "- analysed 113 / 180 sequences\n",
            "- analysed 114 / 180 sequences\n",
            "- analysed 115 / 180 sequences\n",
            "- analysed 116 / 180 sequences\n",
            "- analysed 117 / 180 sequences\n",
            "- analysed 118 / 180 sequences\n",
            "- analysed 119 / 180 sequences\n",
            "- analysed 120 / 180 sequences\n",
            "- analysed 121 / 180 sequences\n",
            "- analysed 122 / 180 sequences\n",
            "- analysed 123 / 180 sequences\n",
            "- analysed 124 / 180 sequences\n",
            "- analysed 125 / 180 sequences\n",
            "- analysed 126 / 180 sequences\n",
            "- analysed 127 / 180 sequences\n",
            "- analysed 128 / 180 sequences\n",
            "- analysed 129 / 180 sequences\n",
            "- analysed 130 / 180 sequences\n",
            "- analysed 131 / 180 sequences\n",
            "- analysed 132 / 180 sequences\n",
            "- analysed 133 / 180 sequences\n",
            "- analysed 134 / 180 sequences\n",
            "- analysed 135 / 180 sequences\n",
            "- analysed 136 / 180 sequences\n",
            "- analysed 137 / 180 sequences\n",
            "- analysed 138 / 180 sequences\n",
            "- analysed 139 / 180 sequences\n",
            "- analysed 140 / 180 sequences\n",
            "- analysed 141 / 180 sequences\n",
            "- analysed 142 / 180 sequences\n",
            "- analysed 143 / 180 sequences\n",
            "- analysed 144 / 180 sequences\n",
            "- analysed 145 / 180 sequences\n",
            "- analysed 146 / 180 sequences\n",
            "- analysed 147 / 180 sequences\n",
            "- analysed 148 / 180 sequences\n",
            "- analysed 149 / 180 sequences\n",
            "- analysed 150 / 180 sequences\n",
            "- analysed 151 / 180 sequences\n",
            "- analysed 152 / 180 sequences\n",
            "- analysed 153 / 180 sequences\n",
            "- analysed 154 / 180 sequences\n",
            "- analysed 155 / 180 sequences\n",
            "- analysed 156 / 180 sequences\n",
            "- analysed 157 / 180 sequences\n",
            "- analysed 158 / 180 sequences\n",
            "- analysed 159 / 180 sequences\n",
            "- analysed 160 / 180 sequences\n",
            "- analysed 161 / 180 sequences\n",
            "- analysed 162 / 180 sequences\n",
            "- analysed 163 / 180 sequences\n",
            "- analysed 164 / 180 sequences\n",
            "- analysed 165 / 180 sequences\n",
            "- analysed 166 / 180 sequences\n",
            "- analysed 167 / 180 sequences\n",
            "- analysed 168 / 180 sequences\n",
            "- analysed 169 / 180 sequences\n",
            "- analysed 170 / 180 sequences\n",
            "- analysed 171 / 180 sequences\n",
            "- analysed 172 / 180 sequences\n",
            "- analysed 173 / 180 sequences\n",
            "- analysed 174 / 180 sequences\n",
            "- analysed 175 / 180 sequences\n",
            "- analysed 176 / 180 sequences\n",
            "- analysed 177 / 180 sequences\n",
            "- analysed 178 / 180 sequences\n",
            "- analysed 179 / 180 sequences\n",
            "- analysed 180 / 180 sequences\n",
            "Building rough initial tree...\n",
            "- added 0 / 180 tips\n",
            "- added 2 / 180 tips\n",
            "- added 3 / 180 tips\n",
            "- added 4 / 180 tips\n",
            "- added 5 / 180 tips\n",
            "- added 6 / 180 tips\n",
            "- added 7 / 180 tips\n",
            "- added 8 / 180 tips\n",
            "- added 9 / 180 tips\n",
            "- added 10 / 180 tips\n",
            "- added 11 / 180 tips\n",
            "- added 12 / 180 tips\n",
            "- added 13 / 180 tips\n",
            "- added 14 / 180 tips\n",
            "- added 15 / 180 tips\n",
            "- added 16 / 180 tips\n",
            "- added 17 / 180 tips\n",
            "- added 18 / 180 tips\n",
            "- added 19 / 180 tips\n",
            "- added 20 / 180 tips\n",
            "- added 21 / 180 tips\n",
            "- added 22 / 180 tips\n",
            "- added 23 / 180 tips\n",
            "- added 24 / 180 tips\n",
            "- added 25 / 180 tips\n",
            "- added 26 / 180 tips\n",
            "- added 27 / 180 tips\n",
            "- added 28 / 180 tips\n",
            "- added 29 / 180 tips\n",
            "- added 30 / 180 tips\n",
            "- added 31 / 180 tips\n",
            "- added 32 / 180 tips\n",
            "- added 33 / 180 tips\n",
            "- added 34 / 180 tips\n",
            "- added 35 / 180 tips\n",
            "- added 36 / 180 tips\n",
            "- added 37 / 180 tips\n",
            "- added 38 / 180 tips\n",
            "- added 39 / 180 tips\n",
            "- added 40 / 180 tips\n",
            "- added 41 / 180 tips\n",
            "- added 42 / 180 tips\n",
            "- added 43 / 180 tips\n",
            "- added 44 / 180 tips\n",
            "- added 45 / 180 tips\n",
            "- added 46 / 180 tips\n",
            "- added 47 / 180 tips\n",
            "- added 48 / 180 tips\n",
            "- added 49 / 180 tips\n",
            "- added 50 / 180 tips\n",
            "- added 51 / 180 tips\n",
            "- added 52 / 180 tips\n",
            "- added 53 / 180 tips\n",
            "- added 54 / 180 tips\n",
            "- added 55 / 180 tips\n",
            "- added 56 / 180 tips\n",
            "- added 57 / 180 tips\n",
            "- added 58 / 180 tips\n",
            "- added 59 / 180 tips\n",
            "- added 60 / 180 tips\n",
            "- added 61 / 180 tips\n",
            "- added 62 / 180 tips\n",
            "- added 63 / 180 tips\n",
            "- added 64 / 180 tips\n",
            "- added 65 / 180 tips\n",
            "- added 66 / 180 tips\n",
            "- added 67 / 180 tips\n",
            "- added 68 / 180 tips\n",
            "- added 69 / 180 tips\n",
            "- added 70 / 180 tips\n",
            "- added 71 / 180 tips\n",
            "- added 72 / 180 tips\n",
            "- added 73 / 180 tips\n",
            "- added 74 / 180 tips\n",
            "- added 75 / 180 tips\n",
            "- added 76 / 180 tips\n",
            "- added 77 / 180 tips\n",
            "- added 78 / 180 tips\n",
            "- added 79 / 180 tips\n",
            "- added 80 / 180 tips\n",
            "- added 81 / 180 tips\n",
            "- added 82 / 180 tips\n",
            "- added 83 / 180 tips\n",
            "- added 84 / 180 tips\n",
            "- added 85 / 180 tips\n",
            "- added 86 / 180 tips\n",
            "- added 87 / 180 tips\n",
            "- added 88 / 180 tips\n",
            "- added 89 / 180 tips\n",
            "- added 90 / 180 tips\n",
            "- added 91 / 180 tips\n",
            "- added 92 / 180 tips\n",
            "- added 93 / 180 tips\n",
            "- added 94 / 180 tips\n",
            "- added 95 / 180 tips\n",
            "- added 96 / 180 tips\n",
            "- added 97 / 180 tips\n",
            "- added 98 / 180 tips\n",
            "- added 99 / 180 tips\n",
            "- added 100 / 180 tips\n",
            "- added 101 / 180 tips\n",
            "- added 102 / 180 tips\n",
            "- added 103 / 180 tips\n",
            "- added 104 / 180 tips\n",
            "- added 105 / 180 tips\n",
            "- added 106 / 180 tips\n",
            "- added 107 / 180 tips\n",
            "- added 108 / 180 tips\n",
            "- added 109 / 180 tips\n",
            "- added 110 / 180 tips\n",
            "- added 111 / 180 tips\n",
            "- added 112 / 180 tips\n",
            "- added 113 / 180 tips\n",
            "- added 114 / 180 tips\n",
            "- added 115 / 180 tips\n",
            "- added 116 / 180 tips\n",
            "- added 117 / 180 tips\n",
            "- added 118 / 180 tips\n",
            "- added 119 / 180 tips\n",
            "- added 120 / 180 tips\n",
            "- added 121 / 180 tips\n",
            "- added 122 / 180 tips\n",
            "- added 123 / 180 tips\n",
            "- added 124 / 180 tips\n",
            "- added 125 / 180 tips\n",
            "- added 126 / 180 tips\n",
            "- added 127 / 180 tips\n",
            "- added 128 / 180 tips\n",
            "- added 129 / 180 tips\n",
            "- added 130 / 180 tips\n",
            "- added 131 / 180 tips\n",
            "- added 132 / 180 tips\n",
            "- added 133 / 180 tips\n",
            "- added 134 / 180 tips\n",
            "- added 135 / 180 tips\n",
            "- added 136 / 180 tips\n",
            "- added 137 / 180 tips\n",
            "- added 138 / 180 tips\n",
            "- added 139 / 180 tips\n",
            "- added 140 / 180 tips\n",
            "- added 141 / 180 tips\n",
            "- added 142 / 180 tips\n",
            "- added 143 / 180 tips\n",
            "- added 144 / 180 tips\n",
            "- added 145 / 180 tips\n",
            "- added 146 / 180 tips\n",
            "- added 147 / 180 tips\n",
            "- added 148 / 180 tips\n",
            "- added 149 / 180 tips\n",
            "- added 150 / 180 tips\n",
            "- added 151 / 180 tips\n",
            "- added 152 / 180 tips\n",
            "- added 153 / 180 tips\n",
            "- added 154 / 180 tips\n",
            "- added 155 / 180 tips\n",
            "- added 156 / 180 tips\n",
            "- added 157 / 180 tips\n",
            "- added 158 / 180 tips\n",
            "- added 159 / 180 tips\n",
            "- added 160 / 180 tips\n",
            "- added 161 / 180 tips\n",
            "- added 162 / 180 tips\n",
            "- added 163 / 180 tips\n",
            "- added 164 / 180 tips\n",
            "- added 165 / 180 tips\n",
            "- added 166 / 180 tips\n",
            "- added 167 / 180 tips\n",
            "- added 168 / 180 tips\n",
            "- added 169 / 180 tips\n",
            "- added 170 / 180 tips\n",
            "- added 171 / 180 tips\n",
            "- added 172 / 180 tips\n",
            "- added 173 / 180 tips\n",
            "- added 174 / 180 tips\n",
            "- added 175 / 180 tips\n",
            "- added 176 / 180 tips\n",
            "- added 177 / 180 tips\n",
            "- added 178 / 180 tips\n",
            "- added 179 / 180 tips\n",
            "- added 180 / 180 tips\n",
            "Built initial tree with 180 tips\n",
            "# Parallelism: 2\n",
            "0.0 M steps/s, Step 0, log_posterior = -68847.31, log_G = -63733.92, log_coal = -9343.97, log_other_priors = 4230.58, num_muts = 1259, T = 391436.77, t_MRCA = 1992-09-15, n0 = 1095.0, g = 0.0000, mu = 1 * 10^-3 subst / site / year, M_ts = 988, M_tv = 271, kappa = 1, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4053, 3753, 5099], num_muts_ab counts: [52, 246, 37, 36, 15, 232, 227, 39, 36, 29, 283, 27]\n",
            "0.0 M steps/s, Step 450000, log_posterior = -41973.22, log_G = -44510.39, log_coal = -1690.03, log_other_priors = 4227.20, num_muts = 1161, T = 101191.21, t_MRCA = 1937-10-21, n0 = 4398.7, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 904, M_tv = 257, kappa = 7.36, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4048, 3739, 5106], num_muts_ab counts: [56, 232, 34, 26, 13, 214, 207, 38, 26, 30, 251, 34]\n",
            "0.3 M steps/s, Step 900000, log_posterior = -41852.31, log_G = -44401.87, log_coal = -1677.70, log_other_priors = 4227.26, num_muts = 1157, T = 97224.74, t_MRCA = 1938-05-25, n0 = 4401.9, g = -0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 903, M_tv = 254, kappa = 7.09, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6058, 4068, 3738, 5095], num_muts_ab counts: [49, 228, 35, 33, 16, 224, 209, 34, 26, 25, 242, 36]\n",
            "0.2 M steps/s, Step 1350000, log_posterior = -41556.56, log_G = -44095.14, log_coal = -1688.69, log_other_priors = 4227.27, num_muts = 1136, T = 98502.49, t_MRCA = 1938-09-23, n0 = 3799.3, g = -0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 890, M_tv = 246, kappa = 7.72, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6051, 4045, 3758, 5105], num_muts_ab counts: [52, 216, 34, 30, 12, 203, 215, 33, 33, 26, 256, 26]\n",
            "0.2 M steps/s, Step 1800000, log_posterior = -41531.56, log_G = -44086.54, log_coal = -1672.15, log_other_priors = 4227.14, num_muts = 1136, T = 97830.84, t_MRCA = 1936-07-20, n0 = 4180.3, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 890, M_tv = 246, kappa = 7.88, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6051, 4045, 3758, 5105], num_muts_ab counts: [52, 215, 35, 29, 13, 204, 215, 33, 32, 26, 256, 26]\n",
            "0.2 M steps/s, Step 2250000, log_posterior = -41576.13, log_G = -44122.01, log_coal = -1681.23, log_other_priors = 4227.12, num_muts = 1138, T = 98168.95, t_MRCA = 1938-10-02, n0 = 4709.5, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 890, M_tv = 248, kappa = 7.42, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6039, 4067, 3753, 5100], num_muts_ab counts: [46, 212, 29, 34, 14, 222, 218, 34, 26, 31, 238, 34]\n",
            "0.3 M steps/s, Step 2700000, log_posterior = -41567.24, log_G = -44109.97, log_coal = -1684.40, log_other_priors = 4227.13, num_muts = 1137, T = 98286.08, t_MRCA = 1939-09-20, n0 = 4833.1, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 889, M_tv = 248, kappa = 7.25, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6039, 4067, 3753, 5100], num_muts_ab counts: [46, 212, 29, 34, 14, 221, 218, 34, 26, 31, 238, 34]\n",
            "0.2 M steps/s, Step 3150000, log_posterior = -41443.25, log_G = -44013.91, log_coal = -1656.63, log_other_priors = 4227.29, num_muts = 1134, T = 93892.95, t_MRCA = 1939-12-10, n0 = 4045.4, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 887, M_tv = 247, kappa = 7.32, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4039, 3746, 5111], num_muts_ab counts: [54, 223, 39, 27, 13, 199, 208, 34, 26, 23, 257, 31]\n",
            "0.2 M steps/s, Step 3600000, log_posterior = -41477.50, log_G = -44031.12, log_coal = -1673.50, log_other_priors = 4227.13, num_muts = 1135, T = 93999.46, t_MRCA = 1942-06-29, n0 = 4949.7, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 887, M_tv = 248, kappa = 7.15, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4039, 3746, 5111], num_muts_ab counts: [54, 223, 38, 27, 13, 199, 208, 34, 28, 23, 257, 31]\n",
            "0.2 M steps/s, Step 4050000, log_posterior = -41440.34, log_G = -44003.92, log_coal = -1663.82, log_other_priors = 4227.40, num_muts = 1134, T = 93097.70, t_MRCA = 1941-07-26, n0 = 4235.8, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 887, M_tv = 247, kappa = 6.65, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4039, 3746, 5111], num_muts_ab counts: [54, 223, 39, 27, 13, 199, 208, 34, 26, 23, 257, 31]\n",
            "0.3 M steps/s, Step 4500000, log_posterior = -41351.81, log_G = -43931.80, log_coal = -1647.43, log_other_priors = 4227.42, num_muts = 1132, T = 89695.60, t_MRCA = 1942-05-12, n0 = 3524.9, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 885, M_tv = 247, kappa = 7.35, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4039, 3746, 5111], num_muts_ab counts: [55, 223, 39, 27, 13, 198, 208, 33, 26, 23, 256, 31]\n",
            "0.3 M steps/s, Step 4950000, log_posterior = -41379.98, log_G = -43956.42, log_coal = -1651.22, log_other_priors = 4227.66, num_muts = 1132, T = 91576.28, t_MRCA = 1941-02-05, n0 = 3157.9, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 885, M_tv = 247, kappa = 6.79, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4039, 3746, 5111], num_muts_ab counts: [55, 223, 38, 27, 13, 198, 208, 33, 27, 23, 256, 31]\n",
            "0.3 M steps/s, Step 5400000, log_posterior = -41419.14, log_G = -43981.23, log_coal = -1665.28, log_other_priors = 4227.37, num_muts = 1131, T = 94912.18, t_MRCA = 1939-11-12, n0 = 4268.7, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 884, M_tv = 247, kappa = 6.74, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4039, 3746, 5111], num_muts_ab counts: [55, 223, 38, 27, 13, 198, 208, 33, 27, 23, 255, 31]\n",
            "0.3 M steps/s, Step 5850000, log_posterior = -41516.42, log_G = -44042.17, log_coal = -1701.58, log_other_priors = 4227.33, num_muts = 1131, T = 100200.80, t_MRCA = 1939-12-31, n0 = 3804.9, g = -0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 884, M_tv = 247, kappa = 7.41, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4039, 3746, 5111], num_muts_ab counts: [55, 223, 38, 27, 13, 198, 208, 33, 27, 23, 255, 31]\n",
            "0.3 M steps/s, Step 6300000, log_posterior = -41464.97, log_G = -44027.80, log_coal = -1664.40, log_other_priors = 4227.23, num_muts = 1134, T = 94978.59, t_MRCA = 1940-04-28, n0 = 4215.3, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 887, M_tv = 247, kappa = 7.4, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4039, 3746, 5111], num_muts_ab counts: [55, 223, 38, 27, 13, 199, 208, 33, 27, 23, 257, 31]\n",
            "0.3 M steps/s, Step 6750000, log_posterior = -41477.70, log_G = -44012.27, log_coal = -1692.52, log_other_priors = 4227.10, num_muts = 1129, T = 100084.83, t_MRCA = 1940-01-18, n0 = 4966.5, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 882, M_tv = 247, kappa = 7.28, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6045, 4064, 3758, 5092], num_muts_ab counts: [47, 213, 35, 32, 15, 217, 214, 32, 32, 28, 238, 26]\n",
            "0.3 M steps/s, Step 7200000, log_posterior = -41413.68, log_G = -43966.53, log_coal = -1674.47, log_other_priors = 4227.31, num_muts = 1129, T = 96064.71, t_MRCA = 1940-08-20, n0 = 4037.3, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 882, M_tv = 247, kappa = 7.26, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4056, 3739, 5107], num_muts_ab counts: [48, 232, 29, 35, 12, 209, 198, 32, 30, 30, 243, 31]\n",
            "0.3 M steps/s, Step 7650000, log_posterior = -41314.56, log_G = -43871.85, log_coal = -1669.95, log_other_priors = 4227.25, num_muts = 1124, T = 94420.05, t_MRCA = 1941-02-09, n0 = 4010.9, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 878, M_tv = 246, kappa = 7.56, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6050, 4073, 3744, 5092], num_muts_ab counts: [47, 222, 33, 35, 14, 220, 206, 31, 27, 28, 230, 31]\n",
            "0.3 M steps/s, Step 8100000, log_posterior = -41256.83, log_G = -43792.96, log_coal = -1691.16, log_other_priors = 4227.28, num_muts = 1115, T = 99142.81, t_MRCA = 1940-01-05, n0 = 4236.0, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 870, M_tv = 245, kappa = 7.18, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6036, 4056, 3763, 5104], num_muts_ab counts: [49, 205, 32, 31, 12, 203, 223, 32, 30, 27, 239, 32]\n",
            "0.3 M steps/s, Step 8550000, log_posterior = -41492.98, log_G = -44039.50, log_coal = -1680.56, log_other_priors = 4227.08, num_muts = 1132, T = 97519.67, t_MRCA = 1938-08-21, n0 = 4027.3, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 882, M_tv = 250, kappa = 8.31, pi = [0.33, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4056, 3744, 5102], num_muts_ab counts: [48, 233, 32, 33, 15, 204, 200, 34, 33, 31, 245, 24]\n",
            "0.3 M steps/s, Step 9000000, log_posterior = -41459.67, log_G = -43990.22, log_coal = -1696.72, log_other_priors = 4227.27, num_muts = 1124, T = 104787.45, t_MRCA = 1933-04-16, n0 = 4592.1, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 878, M_tv = 246, kappa = 6.85, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6064, 4049, 3736, 5110], num_muts_ab counts: [50, 233, 35, 31, 13, 199, 197, 33, 26, 27, 249, 31]\n",
            "0.3 M steps/s, Step 9450000, log_posterior = -41406.72, log_G = -43916.06, log_coal = -1717.54, log_other_priors = 4226.88, num_muts = 1119, T = 104423.76, t_MRCA = 1938-09-20, n0 = 6043.8, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 871, M_tv = 248, kappa = 7.38, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6058, 4044, 3748, 5109], num_muts_ab counts: [55, 226, 31, 28, 11, 197, 203, 35, 30, 29, 245, 29]\n",
            "0.4 M steps/s, Step 9900000, log_posterior = -41279.83, log_G = -43797.84, log_coal = -1708.95, log_other_priors = 4226.96, num_muts = 1113, T = 102608.16, t_MRCA = 1937-07-23, n0 = 5820.8, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 871, M_tv = 242, kappa = 7.19, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6046, 4063, 3760, 5090], num_muts_ab counts: [53, 211, 32, 27, 13, 214, 214, 31, 32, 28, 232, 26]\n",
            "0.4 M steps/s, Step 10350000, log_posterior = -41420.19, log_G = -43967.54, log_coal = -1679.76, log_other_priors = 4227.12, num_muts = 1128, T = 96899.34, t_MRCA = 1939-01-03, n0 = 3816.2, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 880, M_tv = 248, kappa = 8.41, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4062, 3754, 5096], num_muts_ab counts: [54, 219, 31, 26, 12, 216, 217, 36, 25, 31, 228, 33]\n",
            "0.4 M steps/s, Step 10800000, log_posterior = -41250.98, log_G = -43773.22, log_coal = -1704.79, log_other_priors = 4227.03, num_muts = 1111, T = 103298.59, t_MRCA = 1936-08-10, n0 = 5191.1, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 870, M_tv = 241, kappa = 7.35, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6050, 4075, 3753, 5081], num_muts_ab counts: [51, 214, 34, 29, 12, 224, 212, 32, 28, 25, 220, 30]\n",
            "0.4 M steps/s, Step 11250000, log_posterior = -41243.45, log_G = -43804.93, log_coal = -1666.04, log_other_priors = 4227.52, num_muts = 1117, T = 97599.51, t_MRCA = 1937-04-18, n0 = 3681.8, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 874, M_tv = 243, kappa = 6.72, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6050, 4058, 3761, 5090], num_muts_ab counts: [51, 212, 37, 25, 14, 210, 217, 33, 32, 24, 235, 27]\n",
            "0.4 M steps/s, Step 11700000, log_posterior = -41354.16, log_G = -43886.55, log_coal = -1694.87, log_other_priors = 4227.26, num_muts = 1120, T = 100986.15, t_MRCA = 1937-10-01, n0 = 3554.5, g = -0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 877, M_tv = 243, kappa = 8.08, pi = [0.32, 0.21, 0.2, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6046, 4063, 3754, 5096], num_muts_ab counts: [50, 213, 35, 28, 13, 214, 216, 32, 25, 28, 234, 32]\n",
            "0.4 M steps/s, Step 12150000, log_posterior = -41404.85, log_G = -43920.35, log_coal = -1711.34, log_other_priors = 4226.84, num_muts = 1120, T = 104256.04, t_MRCA = 1937-01-21, n0 = 6031.0, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 877, M_tv = 243, kappa = 7.55, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6046, 4063, 3754, 5096], num_muts_ab counts: [50, 213, 35, 28, 13, 214, 216, 32, 25, 28, 234, 32]\n",
            "0.4 M steps/s, Step 12600000, log_posterior = -41347.19, log_G = -43881.79, log_coal = -1692.45, log_other_priors = 4227.06, num_muts = 1120, T = 100408.77, t_MRCA = 1936-05-28, n0 = 5018.7, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 877, M_tv = 243, kappa = 7.41, pi = [0.31, 0.22, 0.21, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6046, 4063, 3754, 5096], num_muts_ab counts: [50, 213, 36, 28, 13, 215, 216, 32, 24, 28, 233, 32]\n",
            "0.4 M steps/s, Step 13050000, log_posterior = -41362.84, log_G = -43893.27, log_coal = -1696.88, log_other_priors = 4227.31, num_muts = 1120, T = 101822.22, t_MRCA = 1936-10-05, n0 = 4697.2, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 877, M_tv = 243, kappa = 6.59, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6046, 4063, 3754, 5096], num_muts_ab counts: [50, 213, 35, 28, 13, 215, 216, 32, 25, 28, 233, 32]\n",
            "0.3 M steps/s, Step 13500000, log_posterior = -41283.14, log_G = -43813.92, log_coal = -1696.24, log_other_priors = 4227.02, num_muts = 1115, T = 101852.26, t_MRCA = 1936-04-10, n0 = 4478.2, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 876, M_tv = 239, kappa = 8.1, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 213, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.3 M steps/s, Step 13950000, log_posterior = -41277.55, log_G = -43799.74, log_coal = -1705.02, log_other_priors = 4227.21, num_muts = 1114, T = 102101.40, t_MRCA = 1936-03-18, n0 = 4322.1, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 7.38, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.3 M steps/s, Step 14400000, log_posterior = -41238.08, log_G = -43779.61, log_coal = -1685.61, log_other_priors = 4227.15, num_muts = 1114, T = 100284.59, t_MRCA = 1937-03-25, n0 = 4568.7, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 7.42, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.3 M steps/s, Step 14850000, log_posterior = -41325.51, log_G = -43833.38, log_coal = -1719.17, log_other_priors = 4227.04, num_muts = 1114, T = 105230.64, t_MRCA = 1937-05-30, n0 = 4956.0, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 7.55, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.3 M steps/s, Step 15300000, log_posterior = -41270.82, log_G = -43805.29, log_coal = -1692.41, log_other_priors = 4226.89, num_muts = 1114, T = 102521.42, t_MRCA = 1935-07-11, n0 = 4807.0, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 8.4, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 34, 24, 12, 208, 215, 31, 25, 29, 240, 31]\n",
            "0.3 M steps/s, Step 15750000, log_posterior = -41337.59, log_G = -43848.22, log_coal = -1716.19, log_other_priors = 4226.82, num_muts = 1114, T = 106632.66, t_MRCA = 1935-04-20, n0 = 5445.0, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 8.13, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 34, 24, 12, 208, 215, 31, 25, 29, 240, 31]\n",
            "0.3 M steps/s, Step 16200000, log_posterior = -41334.83, log_G = -43849.21, log_coal = -1712.60, log_other_priors = 4226.98, num_muts = 1114, T = 106395.81, t_MRCA = 1934-04-07, n0 = 5900.6, g = 0.0001, mu = 0.2 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 7.03, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.3 M steps/s, Step 16650000, log_posterior = -41344.29, log_G = -43860.25, log_coal = -1710.93, log_other_priors = 4226.88, num_muts = 1115, T = 106029.72, t_MRCA = 1934-07-21, n0 = 5357.2, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 876, M_tv = 239, kappa = 7.91, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 209, 215, 31, 26, 29, 240, 31]\n",
            "0.4 M steps/s, Step 17100000, log_posterior = -41323.50, log_G = -43846.75, log_coal = -1703.95, log_other_priors = 4227.19, num_muts = 1115, T = 104766.62, t_MRCA = 1935-05-27, n0 = 4543.5, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 876, M_tv = 239, kappa = 7.24, pi = [0.32, 0.21, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 209, 215, 31, 26, 29, 240, 31]\n",
            "0.4 M steps/s, Step 17550000, log_posterior = -41279.10, log_G = -43817.04, log_coal = -1689.16, log_other_priors = 4227.10, num_muts = 1114, T = 103382.30, t_MRCA = 1934-11-02, n0 = 3893.9, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 8.38, pi = [0.31, 0.21, 0.2, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.4 M steps/s, Step 18000000, log_posterior = -41232.78, log_G = -43772.69, log_coal = -1687.24, log_other_priors = 4227.15, num_muts = 1114, T = 99614.36, t_MRCA = 1937-04-09, n0 = 4511.0, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 7.45, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.4 M steps/s, Step 18450000, log_posterior = -41242.75, log_G = -43787.81, log_coal = -1682.15, log_other_priors = 4227.22, num_muts = 1115, T = 99684.70, t_MRCA = 1937-05-31, n0 = 4413.8, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 876, M_tv = 239, kappa = 7.25, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 209, 215, 31, 26, 29, 240, 31]\n",
            "0.4 M steps/s, Step 18900000, log_posterior = -41276.77, log_G = -43803.30, log_coal = -1700.72, log_other_priors = 4227.25, num_muts = 1114, T = 101990.67, t_MRCA = 1938-06-12, n0 = 4425.3, g = -0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 7.09, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.4 M steps/s, Step 19350000, log_posterior = -41270.24, log_G = -43790.02, log_coal = -1707.28, log_other_priors = 4227.06, num_muts = 1114, T = 101060.95, t_MRCA = 1940-11-14, n0 = 5736.8, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 875, M_tv = 239, kappa = 6.81, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4053, 3755, 5103], num_muts_ab counts: [53, 212, 33, 24, 12, 208, 215, 31, 26, 29, 240, 31]\n",
            "0.4 M steps/s, Step 19800000, log_posterior = -41288.75, log_G = -43838.20, log_coal = -1677.93, log_other_priors = 4227.37, num_muts = 1120, T = 97005.14, t_MRCA = 1940-03-24, n0 = 4108.0, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 879, M_tv = 241, kappa = 6.89, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 215, 31, 30, 10, 209, 212, 33, 29, 31, 243, 28]\n",
            "0.4 M steps/s, Step 20250000, log_posterior = -41280.80, log_G = -43834.50, log_coal = -1673.60, log_other_priors = 4227.30, num_muts = 1120, T = 96815.60, t_MRCA = 1939-11-01, n0 = 4103.4, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 879, M_tv = 241, kappa = 7.2, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 215, 31, 30, 10, 209, 212, 33, 29, 31, 243, 28]\n",
            "0.4 M steps/s, Step 20700000, log_posterior = -41330.00, log_G = -43863.45, log_coal = -1693.65, log_other_priors = 4227.10, num_muts = 1121, T = 98219.56, t_MRCA = 1940-03-16, n0 = 5160.8, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 880, M_tv = 241, kappa = 7.09, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 215, 31, 30, 10, 209, 212, 34, 29, 30, 244, 28]\n",
            "0.4 M steps/s, Step 21150000, log_posterior = -41334.30, log_G = -43872.11, log_coal = -1689.30, log_other_priors = 4227.11, num_muts = 1121, T = 98866.06, t_MRCA = 1940-01-14, n0 = 4879.6, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 880, M_tv = 241, kappa = 7.3, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 215, 31, 30, 10, 209, 212, 34, 29, 30, 244, 28]\n",
            "0.4 M steps/s, Step 21600000, log_posterior = -41286.08, log_G = -43836.52, log_coal = -1676.75, log_other_priors = 4227.20, num_muts = 1120, T = 97091.05, t_MRCA = 1939-07-29, n0 = 3905.7, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 879, M_tv = 241, kappa = 7.92, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 215, 31, 30, 10, 209, 212, 34, 29, 30, 243, 28]\n",
            "0.4 M steps/s, Step 22050000, log_posterior = -41235.14, log_G = -43770.66, log_coal = -1691.44, log_other_priors = 4226.95, num_muts = 1116, T = 96730.51, t_MRCA = 1941-10-28, n0 = 5493.0, g = 0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 876, M_tv = 240, kappa = 7.47, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 214, 31, 30, 10, 209, 210, 33, 29, 30, 243, 28]\n",
            "0.4 M steps/s, Step 22500000, log_posterior = -41244.81, log_G = -43777.58, log_coal = -1694.17, log_other_priors = 4226.95, num_muts = 1116, T = 97097.30, t_MRCA = 1943-02-04, n0 = 4727.5, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 876, M_tv = 240, kappa = 8.2, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 214, 31, 30, 10, 209, 210, 33, 29, 30, 243, 28]\n",
            "0.3 M steps/s, Step 22950000, log_posterior = -41240.71, log_G = -43770.04, log_coal = -1697.72, log_other_priors = 4227.06, num_muts = 1116, T = 96605.00, t_MRCA = 1943-02-04, n0 = 4621.3, g = -0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 876, M_tv = 240, kappa = 7.78, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 214, 31, 30, 10, 209, 210, 34, 29, 29, 243, 28]\n",
            "0.3 M steps/s, Step 23400000, log_posterior = -41238.99, log_G = -43773.31, log_coal = -1692.91, log_other_priors = 4227.23, num_muts = 1116, T = 96777.60, t_MRCA = 1943-10-28, n0 = 4550.7, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 876, M_tv = 240, kappa = 7.07, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 214, 31, 30, 10, 209, 210, 34, 29, 29, 243, 28]\n",
            "0.3 M steps/s, Step 23850000, log_posterior = -41291.57, log_G = -43807.22, log_coal = -1711.27, log_other_priors = 4226.92, num_muts = 1116, T = 99926.11, t_MRCA = 1942-12-16, n0 = 5475.9, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 876, M_tv = 240, kappa = 7.63, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 214, 31, 30, 10, 209, 210, 34, 29, 29, 243, 28]\n",
            "0.3 M steps/s, Step 24300000, log_posterior = -41294.47, log_G = -43826.04, log_coal = -1695.34, log_other_priors = 4226.92, num_muts = 1118, T = 98605.09, t_MRCA = 1942-04-26, n0 = 4888.9, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 878, M_tv = 240, kappa = 8.16, pi = [0.31, 0.22, 0.2, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4055, 3757, 5103], num_muts_ab counts: [49, 214, 31, 30, 10, 210, 210, 34, 29, 29, 244, 28]\n",
            "0.3 M steps/s, Step 24750000, log_posterior = -41227.43, log_G = -43740.61, log_coal = -1713.98, log_other_priors = 4227.16, num_muts = 1110, T = 102007.77, t_MRCA = 1941-09-07, n0 = 5136.5, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 872, M_tv = 238, kappa = 6.84, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6058, 4052, 3749, 5100], num_muts_ab counts: [55, 221, 31, 22, 11, 210, 204, 34, 28, 28, 237, 29]\n",
            "0.3 M steps/s, Step 25200000, log_posterior = -41190.16, log_G = -43719.61, log_coal = -1697.85, log_other_priors = 4227.31, num_muts = 1110, T = 100140.61, t_MRCA = 1941-03-06, n0 = 4519.4, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 872, M_tv = 238, kappa = 6.77, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6058, 4052, 3749, 5100], num_muts_ab counts: [55, 221, 31, 22, 11, 210, 204, 34, 28, 28, 237, 29]\n",
            "0.3 M steps/s, Step 25650000, log_posterior = -41164.78, log_G = -43693.19, log_coal = -1698.60, log_other_priors = 4227.01, num_muts = 1110, T = 97723.19, t_MRCA = 1941-03-23, n0 = 4507.3, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 872, M_tv = 238, kappa = 8.14, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6049, 4064, 3752, 5094], num_muts_ab counts: [47, 222, 29, 32, 11, 213, 204, 34, 32, 27, 233, 26]\n",
            "0.3 M steps/s, Step 26100000, log_posterior = -41084.83, log_G = -43638.57, log_coal = -1673.69, log_other_priors = 4227.43, num_muts = 1109, T = 94148.12, t_MRCA = 1942-04-26, n0 = 4081.6, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 871, M_tv = 238, kappa = 6.69, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6049, 4064, 3752, 5094], num_muts_ab counts: [47, 222, 29, 32, 11, 212, 204, 33, 32, 28, 233, 26]\n",
            "0.3 M steps/s, Step 26550000, log_posterior = -41125.03, log_G = -43667.55, log_coal = -1684.60, log_other_priors = 4227.12, num_muts = 1109, T = 96857.85, t_MRCA = 1941-12-11, n0 = 4473.4, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 871, M_tv = 238, kappa = 7.65, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6049, 4064, 3752, 5094], num_muts_ab counts: [47, 222, 29, 32, 11, 212, 204, 33, 32, 28, 233, 26]\n",
            "0.3 M steps/s, Step 27000000, log_posterior = -41232.73, log_G = -43745.31, log_coal = -1714.26, log_other_priors = 4226.84, num_muts = 1110, T = 102594.83, t_MRCA = 1939-10-28, n0 = 5842.9, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 872, M_tv = 238, kappa = 7.71, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6049, 4064, 3752, 5094], num_muts_ab counts: [47, 222, 30, 32, 11, 213, 204, 33, 31, 28, 233, 26]\n",
            "0.3 M steps/s, Step 27450000, log_posterior = -41206.84, log_G = -43753.81, log_coal = -1680.56, log_other_priors = 4227.53, num_muts = 1114, T = 97087.86, t_MRCA = 1941-07-09, n0 = 3781.5, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 872, M_tv = 242, kappa = 6.56, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4036, 3749, 5120], num_muts_ab counts: [53, 222, 30, 26, 12, 188, 207, 32, 30, 28, 255, 31]\n",
            "0.2 M steps/s, Step 27900000, log_posterior = -41147.58, log_G = -43684.96, log_coal = -1690.06, log_other_priors = 4227.44, num_muts = 1109, T = 98383.68, t_MRCA = 1940-05-26, n0 = 3995.7, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 871, M_tv = 238, kappa = 6.74, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4065, 3740, 5088], num_muts_ab counts: [55, 226, 35, 24, 11, 221, 199, 32, 24, 25, 225, 32]\n",
            "0.2 M steps/s, Step 28350000, log_posterior = -41125.15, log_G = -43675.45, log_coal = -1676.92, log_other_priors = 4227.21, num_muts = 1109, T = 97485.65, t_MRCA = 1941-03-03, n0 = 4060.7, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 871, M_tv = 238, kappa = 7.67, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4065, 3740, 5088], num_muts_ab counts: [55, 226, 35, 24, 11, 221, 199, 32, 24, 25, 225, 32]\n",
            "0.2 M steps/s, Step 28800000, log_posterior = -41170.64, log_G = -43697.94, log_coal = -1699.62, log_other_priors = 4226.92, num_muts = 1109, T = 99421.77, t_MRCA = 1939-08-16, n0 = 4447.7, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 871, M_tv = 238, kappa = 8.61, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4065, 3740, 5088], num_muts_ab counts: [55, 226, 35, 24, 11, 221, 199, 32, 24, 25, 225, 32]\n",
            "0.2 M steps/s, Step 29250000, log_posterior = -41182.34, log_G = -43710.56, log_coal = -1698.89, log_other_priors = 4227.11, num_muts = 1109, T = 100718.82, t_MRCA = 1939-03-25, n0 = 4455.9, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 871, M_tv = 238, kappa = 7.69, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4065, 3740, 5088], num_muts_ab counts: [55, 226, 35, 24, 11, 221, 199, 32, 24, 25, 225, 32]\n",
            "0.2 M steps/s, Step 29700000, log_posterior = -41222.57, log_G = -43742.17, log_coal = -1707.41, log_other_priors = 4227.02, num_muts = 1110, T = 102374.99, t_MRCA = 1939-03-10, n0 = 4585.9, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 872, M_tv = 238, kappa = 7.98, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4065, 3740, 5088], num_muts_ab counts: [55, 226, 35, 24, 11, 222, 199, 32, 24, 25, 225, 32]\n",
            "0.2 M steps/s, Step 30150000, log_posterior = -41130.14, log_G = -43675.17, log_coal = -1682.31, log_other_priors = 4227.34, num_muts = 1109, T = 97481.51, t_MRCA = 1938-02-03, n0 = 4058.5, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 871, M_tv = 238, kappa = 7.09, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4065, 3740, 5088], num_muts_ab counts: [55, 226, 36, 24, 11, 221, 199, 32, 23, 25, 225, 32]\n",
            "0.2 M steps/s, Step 30600000, log_posterior = -41241.41, log_G = -43742.59, log_coal = -1725.83, log_other_priors = 4227.00, num_muts = 1109, T = 103697.28, t_MRCA = 1939-07-19, n0 = 5676.9, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 871, M_tv = 238, kappa = 7.1, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4065, 3740, 5088], num_muts_ab counts: [55, 226, 36, 24, 11, 221, 199, 33, 23, 24, 225, 32]\n",
            "0.2 M steps/s, Step 31050000, log_posterior = -41122.80, log_G = -43642.48, log_coal = -1707.49, log_other_priors = 4227.16, num_muts = 1103, T = 102370.42, t_MRCA = 1938-08-08, n0 = 5032.6, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 865, M_tv = 238, kappa = 6.94, pi = [0.32, 0.21, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4065, 3740, 5088], num_muts_ab counts: [55, 226, 35, 24, 11, 220, 197, 32, 24, 25, 222, 32]\n",
            "0.3 M steps/s, Step 31500000, log_posterior = -41247.63, log_G = -43766.33, log_coal = -1708.13, log_other_priors = 4226.83, num_muts = 1109, T = 105338.05, t_MRCA = 1935-06-03, n0 = 5655.6, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 868, M_tv = 241, kappa = 7.89, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6067, 4043, 3733, 5116], num_muts_ab counts: [49, 233, 33, 29, 9, 195, 191, 34, 26, 25, 249, 36]\n",
            "0.3 M steps/s, Step 31950000, log_posterior = -41313.36, log_G = -43810.30, log_coal = -1729.97, log_other_priors = 4226.91, num_muts = 1109, T = 109520.52, t_MRCA = 1934-10-15, n0 = 5527.0, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 868, M_tv = 241, kappa = 7.62, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6067, 4043, 3733, 5116], num_muts_ab counts: [49, 233, 33, 29, 9, 195, 191, 34, 26, 25, 249, 36]\n",
            "0.4 M steps/s, Step 32400000, log_posterior = -41207.81, log_G = -43711.84, log_coal = -1722.85, log_other_priors = 4226.87, num_muts = 1104, T = 107642.18, t_MRCA = 1935-04-23, n0 = 5489.7, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 865, M_tv = 239, kappa = 7.83, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4058, 3755, 5099], num_muts_ab counts: [49, 211, 36, 30, 12, 206, 212, 31, 27, 24, 236, 30]\n",
            "0.4 M steps/s, Step 32850000, log_posterior = -41198.88, log_G = -43713.09, log_coal = -1712.83, log_other_priors = 4227.04, num_muts = 1104, T = 107717.93, t_MRCA = 1932-11-15, n0 = 4818.8, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 865, M_tv = 239, kappa = 7.65, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4058, 3755, 5099], num_muts_ab counts: [49, 211, 36, 30, 12, 206, 212, 31, 27, 24, 236, 30]\n",
            "0.3 M steps/s, Step 33300000, log_posterior = -41242.83, log_G = -43749.76, log_coal = -1720.24, log_other_priors = 4227.17, num_muts = 1104, T = 111259.14, t_MRCA = 1929-12-02, n0 = 5283.5, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 865, M_tv = 239, kappa = 6.67, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4058, 3755, 5099], num_muts_ab counts: [49, 211, 36, 30, 12, 206, 212, 31, 27, 24, 236, 30]\n",
            "0.3 M steps/s, Step 33750000, log_posterior = -41231.79, log_G = -43737.84, log_coal = -1720.67, log_other_priors = 4226.72, num_muts = 1104, T = 110041.72, t_MRCA = 1931-09-02, n0 = 5951.8, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 865, M_tv = 239, kappa = 8.15, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4058, 3755, 5099], num_muts_ab counts: [49, 211, 36, 30, 12, 206, 212, 31, 27, 24, 236, 30]\n",
            "0.3 M steps/s, Step 34200000, log_posterior = -41272.33, log_G = -43765.57, log_coal = -1733.30, log_other_priors = 4226.54, num_muts = 1104, T = 112749.50, t_MRCA = 1930-04-15, n0 = 6842.9, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 866, M_tv = 238, kappa = 8.36, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4058, 3755, 5099], num_muts_ab counts: [48, 211, 37, 30, 12, 207, 212, 31, 26, 24, 236, 30]\n",
            "0.3 M steps/s, Step 34650000, log_posterior = -41213.81, log_G = -43712.76, log_coal = -1728.05, log_other_priors = 4227.00, num_muts = 1101, T = 112697.07, t_MRCA = 1929-11-23, n0 = 4976.9, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 7.71, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 209, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 35100000, log_posterior = -41214.22, log_G = -43713.19, log_coal = -1728.35, log_other_priors = 4227.32, num_muts = 1101, T = 112388.16, t_MRCA = 1930-08-19, n0 = 5000.4, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 6.3, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 208, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 35550000, log_posterior = -41229.93, log_G = -43722.30, log_coal = -1734.56, log_other_priors = 4226.93, num_muts = 1101, T = 113617.05, t_MRCA = 1930-05-16, n0 = 5304.6, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 7.74, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 34, 23, 10, 209, 207, 32, 22, 25, 238, 34]\n",
            "0.3 M steps/s, Step 36000000, log_posterior = -41310.51, log_G = -43776.80, log_coal = -1760.59, log_other_priors = 4226.89, num_muts = 1101, T = 119429.68, t_MRCA = 1928-12-23, n0 = 5495.1, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 7.77, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 209, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 36450000, log_posterior = -41228.17, log_G = -43722.10, log_coal = -1732.80, log_other_priors = 4226.73, num_muts = 1100, T = 114723.80, t_MRCA = 1927-12-24, n0 = 5623.9, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 8.4, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 36900000, log_posterior = -41234.67, log_G = -43722.36, log_coal = -1739.42, log_other_priors = 4227.10, num_muts = 1100, T = 115227.15, t_MRCA = 1929-05-02, n0 = 5135.8, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.09, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 37350000, log_posterior = -41214.02, log_G = -43703.88, log_coal = -1737.26, log_other_priors = 4227.12, num_muts = 1100, T = 112886.08, t_MRCA = 1930-03-01, n0 = 5085.8, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.05, pi = [0.31, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 34, 23, 10, 208, 207, 32, 22, 25, 238, 34]\n",
            "0.3 M steps/s, Step 37800000, log_posterior = -41192.23, log_G = -43692.55, log_coal = -1726.75, log_other_priors = 4227.07, num_muts = 1100, T = 112063.44, t_MRCA = 1929-10-01, n0 = 5525.0, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 6.91, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 38250000, log_posterior = -41242.59, log_G = -43740.68, log_coal = -1729.08, log_other_priors = 4227.18, num_muts = 1101, T = 115394.24, t_MRCA = 1927-01-29, n0 = 5514.5, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 6.48, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 209, 207, 32, 23, 25, 238, 34]\n",
            "0.4 M steps/s, Step 38700000, log_posterior = -41278.19, log_G = -43764.22, log_coal = -1740.65, log_other_priors = 4226.67, num_muts = 1101, T = 117992.38, t_MRCA = 1926-10-24, n0 = 7058.2, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 7.6, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 34, 23, 10, 209, 207, 32, 22, 25, 238, 34]\n",
            "0.4 M steps/s, Step 39150000, log_posterior = -41300.88, log_G = -43778.39, log_coal = -1749.49, log_other_priors = 4227.00, num_muts = 1100, T = 121143.38, t_MRCA = 1923-12-02, n0 = 6069.5, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 6.83, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 39600000, log_posterior = -41298.86, log_G = -43768.56, log_coal = -1757.12, log_other_priors = 4226.83, num_muts = 1100, T = 120217.07, t_MRCA = 1925-03-20, n0 = 5907.5, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.7, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 40050000, log_posterior = -41241.50, log_G = -43738.98, log_coal = -1729.53, log_other_priors = 4227.01, num_muts = 1100, T = 116657.08, t_MRCA = 1924-06-10, n0 = 5715.4, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.05, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 34, 23, 10, 208, 207, 32, 22, 25, 238, 34]\n",
            "0.3 M steps/s, Step 40500000, log_posterior = -41260.30, log_G = -43746.75, log_coal = -1740.53, log_other_priors = 4226.99, num_muts = 1101, T = 116157.40, t_MRCA = 1926-08-23, n0 = 5132.1, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 7.62, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 209, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 40950000, log_posterior = -41278.52, log_G = -43754.05, log_coal = -1751.22, log_other_priors = 4226.75, num_muts = 1100, T = 118591.45, t_MRCA = 1926-06-13, n0 = 7437.7, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.02, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 41400000, log_posterior = -41285.14, log_G = -43762.46, log_coal = -1749.37, log_other_priors = 4226.69, num_muts = 1101, T = 117911.67, t_MRCA = 1926-06-17, n0 = 7089.1, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 7.5, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 239, 34]\n",
            "0.3 M steps/s, Step 41850000, log_posterior = -41253.16, log_G = -43738.68, log_coal = -1741.19, log_other_priors = 4226.70, num_muts = 1100, T = 116862.08, t_MRCA = 1927-02-18, n0 = 5968.2, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 8.26, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 42300000, log_posterior = -41277.45, log_G = -43755.46, log_coal = -1748.78, log_other_priors = 4226.79, num_muts = 1100, T = 118536.52, t_MRCA = 1925-12-28, n0 = 6351.9, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.54, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.2 M steps/s, Step 42750000, log_posterior = -41306.69, log_G = -43789.83, log_coal = -1743.62, log_other_priors = 4226.76, num_muts = 1102, T = 119265.15, t_MRCA = 1924-04-05, n0 = 6035.9, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 866, M_tv = 236, kappa = 7.92, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 212, 33, 23, 10, 208, 208, 32, 23, 25, 238, 34]\n",
            "0.2 M steps/s, Step 43200000, log_posterior = -41275.72, log_G = -43766.37, log_coal = -1736.20, log_other_priors = 4226.85, num_muts = 1100, T = 119605.81, t_MRCA = 1922-01-24, n0 = 5468.6, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.95, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 34, 23, 10, 208, 207, 32, 22, 25, 238, 34]\n",
            "0.2 M steps/s, Step 43650000, log_posterior = -41301.89, log_G = -43787.82, log_coal = -1740.90, log_other_priors = 4226.83, num_muts = 1100, T = 121836.59, t_MRCA = 1923-08-09, n0 = 5604.9, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.95, pi = [0.31, 0.22, 0.19, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4049, 3750, 5105], num_muts_ab counts: [56, 211, 33, 23, 10, 208, 207, 32, 23, 25, 238, 34]\n",
            "0.3 M steps/s, Step 44100000, log_posterior = -41354.48, log_G = -43822.13, log_coal = -1759.07, log_other_priors = 4226.72, num_muts = 1102, T = 122320.91, t_MRCA = 1922-11-22, n0 = 7176.1, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 865, M_tv = 237, kappa = 7.29, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4061, 3748, 5096], num_muts_ab counts: [51, 215, 34, 28, 12, 214, 203, 32, 25, 24, 233, 31]\n",
            "0.3 M steps/s, Step 44550000, log_posterior = -41308.89, log_G = -43797.44, log_coal = -1738.42, log_other_priors = 4226.97, num_muts = 1101, T = 121180.51, t_MRCA = 1922-03-13, n0 = 6103.5, g = 0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 864, M_tv = 237, kappa = 6.93, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4061, 3748, 5096], num_muts_ab counts: [51, 215, 33, 28, 12, 214, 203, 32, 26, 24, 232, 31]\n",
            "0.3 M steps/s, Step 45000000, log_posterior = -41400.35, log_G = -43879.17, log_coal = -1747.96, log_other_priors = 4226.78, num_muts = 1105, T = 123296.05, t_MRCA = 1921-06-22, n0 = 7240.7, g = 0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 865, M_tv = 240, kappa = 7.02, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4049, 3759, 5103], num_muts_ab counts: [48, 213, 33, 28, 9, 202, 207, 34, 34, 27, 243, 27]\n",
            "0.3 M steps/s, Step 45450000, log_posterior = -41358.48, log_G = -43850.91, log_coal = -1734.50, log_other_priors = 4226.93, num_muts = 1105, T = 120096.78, t_MRCA = 1921-04-11, n0 = 6382.5, g = 0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 865, M_tv = 240, kappa = 6.9, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6067, 4043, 3741, 5108], num_muts_ab counts: [55, 223, 38, 23, 9, 201, 199, 33, 23, 25, 242, 34]\n",
            "0.3 M steps/s, Step 45900000, log_posterior = -41427.52, log_G = -43897.99, log_coal = -1756.36, log_other_priors = 4226.83, num_muts = 1106, T = 123388.09, t_MRCA = 1920-10-14, n0 = 6773.5, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 866, M_tv = 240, kappa = 7.1, pi = [0.32, 0.21, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6059, 4048, 3743, 5109], num_muts_ab counts: [49, 224, 35, 30, 12, 198, 197, 32, 27, 28, 247, 27]\n",
            "0.3 M steps/s, Step 46350000, log_posterior = -41376.73, log_G = -43866.28, log_coal = -1737.46, log_other_priors = 4227.00, num_muts = 1105, T = 121703.21, t_MRCA = 1919-11-02, n0 = 6362.8, g = 0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 865, M_tv = 240, kappa = 6.61, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6059, 4048, 3743, 5109], num_muts_ab counts: [49, 224, 35, 30, 12, 198, 197, 32, 27, 28, 246, 27]\n",
            "0.3 M steps/s, Step 46800000, log_posterior = -41357.24, log_G = -43854.64, log_coal = -1729.68, log_other_priors = 4227.08, num_muts = 1105, T = 120455.75, t_MRCA = 1920-10-13, n0 = 5209.5, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 868, M_tv = 237, kappa = 7.13, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6060, 4058, 3749, 5092], num_muts_ab counts: [49, 225, 34, 26, 10, 211, 199, 34, 32, 24, 233, 28]\n",
            "0.3 M steps/s, Step 47250000, log_posterior = -41344.51, log_G = -43823.40, log_coal = -1748.06, log_other_priors = 4226.95, num_muts = 1102, T = 122073.04, t_MRCA = 1922-08-17, n0 = 5941.8, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 863, M_tv = 239, kappa = 7.13, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4053, 3749, 5105], num_muts_ab counts: [48, 217, 37, 30, 10, 204, 200, 32, 25, 27, 242, 30]\n",
            "0.3 M steps/s, Step 47700000, log_posterior = -41226.46, log_G = -43744.07, log_coal = -1709.59, log_other_priors = 4227.20, num_muts = 1102, T = 113950.60, t_MRCA = 1922-07-02, n0 = 4477.9, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 865, M_tv = 237, kappa = 7.28, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4065, 3746, 5096], num_muts_ab counts: [48, 221, 29, 28, 12, 217, 197, 32, 28, 30, 230, 30]\n",
            "0.3 M steps/s, Step 48150000, log_posterior = -41209.59, log_G = -43704.45, log_coal = -1732.03, log_other_priors = 4226.89, num_muts = 1096, T = 119230.20, t_MRCA = 1922-07-28, n0 = 6405.5, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 859, M_tv = 237, kappa = 7.06, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4081, 3752, 5074], num_muts_ab counts: [50, 216, 32, 29, 10, 233, 200, 33, 28, 26, 210, 29]\n",
            "0.3 M steps/s, Step 48600000, log_posterior = -41236.29, log_G = -43717.94, log_coal = -1745.35, log_other_priors = 4227.01, num_muts = 1097, T = 118998.38, t_MRCA = 1924-11-17, n0 = 5177.7, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 860, M_tv = 237, kappa = 7.48, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6072, 4051, 3735, 5101], num_muts_ab counts: [52, 234, 32, 24, 10, 207, 185, 35, 27, 26, 234, 31]\n",
            "0.3 M steps/s, Step 49050000, log_posterior = -41309.01, log_G = -43766.01, log_coal = -1769.72, log_other_priors = 4226.72, num_muts = 1097, T = 124822.93, t_MRCA = 1924-05-07, n0 = 7368.0, g = 0.0000, mu = 0.16 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 7.19, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6045, 4069, 3747, 5098], num_muts_ab counts: [46, 217, 28, 31, 13, 215, 202, 30, 26, 29, 228, 32]\n",
            "0.3 M steps/s, Step 49500000, log_posterior = -41340.19, log_G = -43833.92, log_coal = -1733.27, log_other_priors = 4227.00, num_muts = 1106, T = 116913.49, t_MRCA = 1926-10-24, n0 = 5489.7, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 867, M_tv = 239, kappa = 7.25, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4074, 3742, 5089], num_muts_ab counts: [49, 226, 31, 30, 12, 223, 199, 32, 24, 29, 219, 32]\n",
            "0.3 M steps/s, Step 49950000, log_posterior = -41324.57, log_G = -43818.83, log_coal = -1732.55, log_other_priors = 4226.82, num_muts = 1106, T = 115291.27, t_MRCA = 1927-02-14, n0 = 5760.8, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 867, M_tv = 239, kappa = 7.87, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4074, 3742, 5089], num_muts_ab counts: [49, 226, 31, 30, 12, 223, 199, 32, 24, 29, 219, 32]\n",
            "0.3 M steps/s, Step 50400000, log_posterior = -41346.22, log_G = -43840.19, log_coal = -1733.07, log_other_priors = 4227.04, num_muts = 1106, T = 117499.96, t_MRCA = 1926-10-07, n0 = 5783.4, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 867, M_tv = 239, kappa = 6.83, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4074, 3742, 5089], num_muts_ab counts: [49, 226, 31, 30, 12, 223, 199, 32, 24, 29, 219, 32]\n",
            "0.3 M steps/s, Step 50850000, log_posterior = -41301.63, log_G = -43801.65, log_coal = -1726.92, log_other_priors = 4226.94, num_muts = 1106, T = 113498.15, t_MRCA = 1929-02-12, n0 = 5123.1, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 867, M_tv = 239, kappa = 7.84, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4074, 3742, 5089], num_muts_ab counts: [49, 226, 30, 30, 12, 223, 199, 32, 25, 29, 219, 32]\n",
            "0.3 M steps/s, Step 51300000, log_posterior = -41255.43, log_G = -43766.54, log_coal = -1715.78, log_other_priors = 4226.90, num_muts = 1106, T = 110001.15, t_MRCA = 1930-09-10, n0 = 5548.9, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 867, M_tv = 239, kappa = 7.67, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4074, 3742, 5089], num_muts_ab counts: [49, 226, 30, 30, 12, 223, 199, 32, 25, 29, 219, 32]\n",
            "0.3 M steps/s, Step 51750000, log_posterior = -41293.12, log_G = -43793.69, log_coal = -1726.20, log_other_priors = 4226.77, num_muts = 1107, T = 111115.32, t_MRCA = 1931-07-16, n0 = 5918.6, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 868, M_tv = 239, kappa = 7.95, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4074, 3742, 5089], num_muts_ab counts: [49, 226, 30, 30, 12, 223, 199, 32, 25, 29, 220, 32]\n",
            "0.3 M steps/s, Step 52200000, log_posterior = -41121.62, log_G = -43621.45, log_coal = -1726.92, log_other_priors = 4226.75, num_muts = 1097, T = 108920.02, t_MRCA = 1933-12-27, n0 = 5831.2, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 860, M_tv = 237, kappa = 8.13, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4062, 3749, 5104], num_muts_ab counts: [46, 214, 30, 31, 14, 208, 202, 30, 26, 30, 236, 30]\n",
            "0.3 M steps/s, Step 52650000, log_posterior = -41095.29, log_G = -43613.61, log_coal = -1708.96, log_other_priors = 4227.28, num_muts = 1098, T = 106979.90, t_MRCA = 1933-11-02, n0 = 4680.0, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 863, M_tv = 235, kappa = 6.73, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 32, 29, 12, 194, 200, 31, 23, 27, 249, 33]\n",
            "0.3 M steps/s, Step 53100000, log_posterior = -41148.95, log_G = -43643.31, log_coal = -1732.56, log_other_priors = 4226.91, num_muts = 1097, T = 111662.88, t_MRCA = 1933-01-26, n0 = 5785.1, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 7.4, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 32, 29, 12, 193, 200, 31, 23, 27, 249, 33]\n",
            "0.3 M steps/s, Step 53550000, log_posterior = -41109.42, log_G = -43619.08, log_coal = -1717.28, log_other_priors = 4226.95, num_muts = 1097, T = 108523.50, t_MRCA = 1931-05-01, n0 = 4710.9, g = -0.0001, mu = 0.19 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 8.21, pi = [0.31, 0.22, 0.19, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 33, 29, 12, 193, 200, 31, 22, 27, 249, 33]\n",
            "0.3 M steps/s, Step 54000000, log_posterior = -41151.06, log_G = -43647.51, log_coal = -1730.52, log_other_priors = 4226.96, num_muts = 1098, T = 110524.41, t_MRCA = 1932-07-23, n0 = 5811.6, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 863, M_tv = 235, kappa = 7.16, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 31, 29, 13, 193, 200, 31, 23, 27, 250, 33]\n",
            "0.4 M steps/s, Step 54450000, log_posterior = -41122.26, log_G = -43627.65, log_coal = -1721.58, log_other_priors = 4226.96, num_muts = 1097, T = 110028.29, t_MRCA = 1933-01-09, n0 = 4793.3, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 8.06, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 32, 29, 13, 193, 200, 31, 22, 27, 249, 33]\n",
            "0.4 M steps/s, Step 54900000, log_posterior = -41137.04, log_G = -43634.90, log_coal = -1729.05, log_other_priors = 4226.91, num_muts = 1098, T = 109419.40, t_MRCA = 1933-04-26, n0 = 5620.1, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 863, M_tv = 235, kappa = 7.55, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 31, 29, 13, 194, 200, 31, 23, 27, 249, 33]\n",
            "0.4 M steps/s, Step 55350000, log_posterior = -41149.00, log_G = -43647.30, log_coal = -1728.81, log_other_priors = 4227.10, num_muts = 1099, T = 108972.39, t_MRCA = 1932-12-22, n0 = 5700.0, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 864, M_tv = 235, kappa = 6.64, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 32, 29, 13, 194, 200, 31, 22, 27, 250, 33]\n",
            "0.4 M steps/s, Step 55800000, log_posterior = -41127.13, log_G = -43632.12, log_coal = -1722.03, log_other_priors = 4227.01, num_muts = 1097, T = 110474.24, t_MRCA = 1933-03-18, n0 = 5584.6, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 7.12, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 31, 29, 13, 193, 200, 31, 23, 27, 249, 33]\n",
            "0.4 M steps/s, Step 56250000, log_posterior = -41107.51, log_G = -43621.62, log_coal = -1712.45, log_other_priors = 4226.56, num_muts = 1099, T = 106506.86, t_MRCA = 1933-08-26, n0 = 6464.3, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 864, M_tv = 235, kappa = 8.57, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 32, 29, 13, 194, 200, 31, 22, 27, 250, 33]\n",
            "0.4 M steps/s, Step 56700000, log_posterior = -41114.63, log_G = -43620.89, log_coal = -1720.56, log_other_priors = 4226.83, num_muts = 1097, T = 109286.03, t_MRCA = 1933-01-17, n0 = 5258.1, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 8.25, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 33, 29, 12, 193, 200, 31, 22, 27, 249, 33]\n",
            "0.4 M steps/s, Step 57150000, log_posterior = -41122.09, log_G = -43624.54, log_coal = -1724.36, log_other_priors = 4226.82, num_muts = 1097, T = 109684.05, t_MRCA = 1933-04-22, n0 = 5600.3, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 8, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 33, 29, 12, 193, 200, 31, 22, 27, 249, 33]\n",
            "0.4 M steps/s, Step 57600000, log_posterior = -41171.37, log_G = -43660.36, log_coal = -1738.04, log_other_priors = 4227.03, num_muts = 1097, T = 113433.68, t_MRCA = 1931-09-29, n0 = 5460.1, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 7.16, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 32, 29, 12, 193, 200, 31, 23, 27, 249, 33]\n",
            "0.4 M steps/s, Step 58050000, log_posterior = -41120.51, log_G = -43629.24, log_coal = -1718.20, log_other_priors = 4226.93, num_muts = 1097, T = 109899.13, t_MRCA = 1931-12-21, n0 = 4924.5, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 862, M_tv = 235, kappa = 8.07, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6052, 4045, 3743, 5119], num_muts_ab counts: [48, 220, 31, 29, 13, 193, 200, 31, 23, 27, 249, 33]\n",
            "0.4 M steps/s, Step 58500000, log_posterior = -41097.65, log_G = -43612.12, log_coal = -1712.44, log_other_priors = 4226.91, num_muts = 1098, T = 106803.86, t_MRCA = 1932-10-24, n0 = 5687.7, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 862, M_tv = 236, kappa = 7.5, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6046, 4061, 3753, 5099], num_muts_ab counts: [43, 217, 33, 33, 12, 205, 202, 32, 31, 26, 238, 26]\n",
            "0.4 M steps/s, Step 58950000, log_posterior = -41157.21, log_G = -43658.51, log_coal = -1725.62, log_other_priors = 4226.92, num_muts = 1098, T = 111201.83, t_MRCA = 1931-12-19, n0 = 5473.2, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 861, M_tv = 237, kappa = 7.64, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6050, 4052, 3758, 5099], num_muts_ab counts: [50, 216, 31, 27, 10, 204, 204, 34, 32, 27, 237, 26]\n",
            "0.4 M steps/s, Step 59400000, log_posterior = -41354.51, log_G = -43846.89, log_coal = -1734.53, log_other_priors = 4226.91, num_muts = 1109, T = 113261.31, t_MRCA = 1929-11-13, n0 = 5479.4, g = -0.0001, mu = 0.19 * 10^-3 subst / site / year, M_ts = 870, M_tv = 239, kappa = 7.69, pi = [0.32, 0.22, 0.19, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 33, 24, 10, 223, 208, 36, 28, 26, 223, 28]\n",
            "0.4 M steps/s, Step 59850000, log_posterior = -41335.06, log_G = -43839.89, log_coal = -1722.12, log_other_priors = 4226.94, num_muts = 1109, T = 112634.53, t_MRCA = 1929-10-30, n0 = 4965.2, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 870, M_tv = 239, kappa = 7.97, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 34, 24, 9, 223, 208, 36, 28, 26, 223, 28]\n",
            "0.4 M steps/s, Step 60300000, log_posterior = -41224.14, log_G = -43764.64, log_coal = -1686.64, log_other_priors = 4227.14, num_muts = 1108, T = 106857.24, t_MRCA = 1930-01-20, n0 = 4386.8, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 869, M_tv = 239, kappa = 7.64, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 34, 24, 9, 223, 208, 36, 28, 26, 222, 28]\n",
            "0.3 M steps/s, Step 60750000, log_posterior = -41288.79, log_G = -43804.79, log_coal = -1710.97, log_other_priors = 4226.97, num_muts = 1108, T = 110799.84, t_MRCA = 1929-07-20, n0 = 5481.4, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 869, M_tv = 239, kappa = 7.42, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 34, 24, 9, 223, 208, 36, 28, 26, 222, 28]\n",
            "0.3 M steps/s, Step 61200000, log_posterior = -41396.14, log_G = -43878.57, log_coal = -1744.53, log_other_priors = 4226.97, num_muts = 1108, T = 118155.32, t_MRCA = 1927-06-28, n0 = 5788.6, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 869, M_tv = 239, kappa = 7.16, pi = [0.32, 0.21, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 35, 24, 9, 223, 208, 36, 27, 26, 222, 28]\n",
            "0.4 M steps/s, Step 61650000, log_posterior = -41354.49, log_G = -43857.54, log_coal = -1723.92, log_other_priors = 4226.97, num_muts = 1110, T = 113331.88, t_MRCA = 1927-04-23, n0 = 5813.5, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 871, M_tv = 239, kappa = 7.12, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 34, 24, 9, 223, 210, 36, 28, 26, 222, 28]\n",
            "0.3 M steps/s, Step 62100000, log_posterior = -41357.24, log_G = -43854.58, log_coal = -1729.39, log_other_priors = 4226.73, num_muts = 1109, T = 114003.19, t_MRCA = 1926-07-15, n0 = 6195.6, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 870, M_tv = 239, kappa = 7.95, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 33, 24, 10, 223, 209, 36, 28, 26, 222, 28]\n",
            "0.3 M steps/s, Step 62550000, log_posterior = -41334.14, log_G = -43839.01, log_coal = -1722.15, log_other_priors = 4227.02, num_muts = 1109, T = 112759.66, t_MRCA = 1927-04-11, n0 = 5240.0, g = -0.0001, mu = 0.19 * 10^-3 subst / site / year, M_ts = 870, M_tv = 239, kappa = 7.35, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 34, 24, 9, 223, 209, 36, 28, 26, 222, 28]\n",
            "0.3 M steps/s, Step 63000000, log_posterior = -41348.37, log_G = -43844.65, log_coal = -1730.44, log_other_priors = 4226.72, num_muts = 1109, T = 113297.03, t_MRCA = 1929-02-07, n0 = 5974.5, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 870, M_tv = 239, kappa = 8.15, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 34, 24, 9, 223, 209, 36, 28, 26, 222, 28]\n",
            "0.3 M steps/s, Step 63450000, log_posterior = -41344.93, log_G = -43838.68, log_coal = -1733.10, log_other_priors = 4226.85, num_muts = 1109, T = 112835.90, t_MRCA = 1931-02-13, n0 = 5539.2, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 870, M_tv = 239, kappa = 7.9, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4064, 3758, 5084], num_muts_ab counts: [54, 216, 35, 24, 9, 223, 209, 36, 27, 26, 222, 28]\n",
            "0.3 M steps/s, Step 63900000, log_posterior = -41182.49, log_G = -43683.83, log_coal = -1725.49, log_other_priors = 4226.83, num_muts = 1100, T = 111138.42, t_MRCA = 1931-09-02, n0 = 5737.1, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 864, M_tv = 236, kappa = 7.83, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6051, 4046, 3749, 5113], num_muts_ab counts: [51, 217, 32, 29, 12, 195, 205, 29, 27, 26, 247, 30]\n",
            "0.4 M steps/s, Step 64350000, log_posterior = -41174.04, log_G = -43678.40, log_coal = -1722.54, log_other_priors = 4226.91, num_muts = 1101, T = 109087.19, t_MRCA = 1932-12-05, n0 = 5306.6, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 865, M_tv = 236, kappa = 7.82, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6051, 4046, 3749, 5113], num_muts_ab counts: [51, 217, 32, 29, 12, 195, 205, 29, 27, 26, 248, 30]\n",
            "0.4 M steps/s, Step 64800000, log_posterior = -41108.36, log_G = -43621.72, log_coal = -1713.55, log_other_priors = 4226.91, num_muts = 1099, T = 106590.55, t_MRCA = 1934-03-19, n0 = 6250.9, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 864, M_tv = 235, kappa = 7.06, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4061, 3746, 5091], num_muts_ab counts: [53, 220, 35, 23, 10, 217, 199, 32, 26, 25, 228, 31]\n",
            "0.3 M steps/s, Step 65250000, log_posterior = -41214.79, log_G = -43686.39, log_coal = -1755.53, log_other_priors = 4227.12, num_muts = 1098, T = 114598.46, t_MRCA = 1934-03-30, n0 = 5280.1, g = -0.0001, mu = 0.19 * 10^-3 subst / site / year, M_ts = 863, M_tv = 235, kappa = 6.88, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4061, 3746, 5091], num_muts_ab counts: [53, 220, 35, 23, 10, 217, 199, 32, 26, 25, 227, 31]\n",
            "0.4 M steps/s, Step 65700000, log_posterior = -41120.25, log_G = -43628.20, log_coal = -1718.86, log_other_priors = 4226.81, num_muts = 1098, T = 108637.37, t_MRCA = 1933-07-13, n0 = 4978.5, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 863, M_tv = 235, kappa = 8.63, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4061, 3746, 5091], num_muts_ab counts: [53, 220, 35, 23, 10, 217, 199, 32, 26, 25, 227, 31]\n",
            "0.4 M steps/s, Step 66150000, log_posterior = -41111.87, log_G = -43616.92, log_coal = -1721.73, log_other_priors = 4226.79, num_muts = 1098, T = 107482.89, t_MRCA = 1934-11-01, n0 = 5258.7, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 863, M_tv = 235, kappa = 8.45, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4061, 3746, 5091], num_muts_ab counts: [53, 220, 36, 23, 10, 217, 199, 32, 25, 25, 227, 31]\n",
            "0.4 M steps/s, Step 66600000, log_posterior = -41046.66, log_G = -43558.50, log_coal = -1715.07, log_other_priors = 4226.91, num_muts = 1096, T = 104687.77, t_MRCA = 1936-07-03, n0 = 4748.6, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 861, M_tv = 235, kappa = 8.36, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4058, 3742, 5106], num_muts_ab counts: [50, 220, 30, 28, 12, 207, 199, 30, 24, 28, 235, 33]\n",
            "0.4 M steps/s, Step 67050000, log_posterior = -41118.65, log_G = -43643.41, log_coal = -1702.34, log_other_priors = 4227.11, num_muts = 1102, T = 104204.65, t_MRCA = 1936-08-19, n0 = 4948.8, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 865, M_tv = 237, kappa = 7.23, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6058, 4061, 3745, 5095], num_muts_ab counts: [48, 225, 35, 29, 13, 208, 198, 30, 30, 26, 234, 26]\n",
            "0.4 M steps/s, Step 67500000, log_posterior = -41142.88, log_G = -43664.95, log_coal = -1704.88, log_other_priors = 4226.95, num_muts = 1103, T = 104654.40, t_MRCA = 1935-06-30, n0 = 5366.0, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 866, M_tv = 237, kappa = 7.56, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6075, 4056, 3729, 5099], num_muts_ab counts: [50, 239, 35, 27, 10, 209, 182, 32, 27, 26, 236, 30]\n",
            "0.4 M steps/s, Step 67950000, log_posterior = -41161.70, log_G = -43676.91, log_coal = -1711.75, log_other_priors = 4226.96, num_muts = 1103, T = 105624.08, t_MRCA = 1935-02-26, n0 = 4745.8, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 866, M_tv = 237, kappa = 8.13, pi = [0.31, 0.22, 0.2, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6075, 4056, 3729, 5099], num_muts_ab counts: [50, 239, 36, 27, 10, 209, 182, 32, 26, 26, 236, 30]\n",
            "0.4 M steps/s, Step 68400000, log_posterior = -41099.33, log_G = -43617.71, log_coal = -1708.54, log_other_priors = 4226.92, num_muts = 1100, T = 104139.35, t_MRCA = 1934-10-08, n0 = 5884.2, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 861, M_tv = 239, kappa = 7.29, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4059, 3738, 5096], num_muts_ab counts: [52, 227, 38, 26, 13, 208, 194, 29, 26, 25, 232, 30]\n",
            "0.4 M steps/s, Step 68850000, log_posterior = -41373.43, log_G = -43865.01, log_coal = -1735.32, log_other_priors = 4226.90, num_muts = 1111, T = 112571.96, t_MRCA = 1933-01-23, n0 = 5750.7, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 7.48, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 34, 27, 12, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 69300000, log_posterior = -41322.80, log_G = -43830.46, log_coal = -1719.11, log_other_priors = 4226.78, num_muts = 1111, T = 108889.46, t_MRCA = 1933-01-23, n0 = 5282.3, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 8.46, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 34, 27, 12, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 69750000, log_posterior = -41284.48, log_G = -43807.70, log_coal = -1703.79, log_other_priors = 4227.01, num_muts = 1111, T = 106468.89, t_MRCA = 1933-07-03, n0 = 5087.1, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 7.55, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 34, 27, 12, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 70200000, log_posterior = -41283.60, log_G = -43806.41, log_coal = -1704.04, log_other_priors = 4226.84, num_muts = 1111, T = 106506.46, t_MRCA = 1932-02-25, n0 = 5329.9, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 8.12, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 34, 27, 12, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 70650000, log_posterior = -41332.45, log_G = -43842.12, log_coal = -1716.73, log_other_priors = 4226.40, num_muts = 1112, T = 108254.60, t_MRCA = 1931-10-12, n0 = 6525.7, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 873, M_tv = 239, kappa = 9.33, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 34, 27, 12, 235, 213, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 71100000, log_posterior = -41362.59, log_G = -43855.05, log_coal = -1734.68, log_other_priors = 4227.15, num_muts = 1111, T = 111391.64, t_MRCA = 1931-04-19, n0 = 5045.9, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 6.98, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 33, 27, 13, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.3 M steps/s, Step 71550000, log_posterior = -41355.34, log_G = -43852.20, log_coal = -1729.81, log_other_priors = 4226.67, num_muts = 1111, T = 111122.69, t_MRCA = 1932-08-31, n0 = 7091.9, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 7.58, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 33, 27, 13, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.3 M steps/s, Step 72000000, log_posterior = -41428.96, log_G = -43910.10, log_coal = -1746.10, log_other_priors = 4227.24, num_muts = 1112, T = 114916.55, t_MRCA = 1931-06-27, n0 = 6066.7, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 873, M_tv = 239, kappa = 5.85, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 33, 27, 13, 236, 212, 32, 30, 27, 214, 29]\n",
            "0.3 M steps/s, Step 72450000, log_posterior = -41356.03, log_G = -43859.94, log_coal = -1723.25, log_other_priors = 4227.17, num_muts = 1111, T = 111858.65, t_MRCA = 1931-06-26, n0 = 4793.3, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 7.12, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 34, 27, 12, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 72900000, log_posterior = -41420.51, log_G = -43904.88, log_coal = -1742.44, log_other_priors = 4226.81, num_muts = 1112, T = 114514.34, t_MRCA = 1929-10-25, n0 = 6417.9, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 872, M_tv = 240, kappa = 7.4, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 34, 27, 13, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 73350000, log_posterior = -41432.25, log_G = -43906.88, log_coal = -1752.21, log_other_priors = 4226.84, num_muts = 1111, T = 116680.77, t_MRCA = 1929-05-20, n0 = 6928.6, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 6.96, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 33, 27, 13, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 73800000, log_posterior = -41349.65, log_G = -43854.52, log_coal = -1721.90, log_other_priors = 4226.77, num_muts = 1111, T = 111037.23, t_MRCA = 1929-08-02, n0 = 6013.1, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 7.9, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 33, 27, 13, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 74250000, log_posterior = -41365.55, log_G = -43868.20, log_coal = -1724.19, log_other_priors = 4226.84, num_muts = 1111, T = 112672.79, t_MRCA = 1929-08-25, n0 = 5406.8, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 872, M_tv = 239, kappa = 8.05, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4082, 3759, 5074], num_muts_ab counts: [48, 211, 34, 27, 12, 235, 212, 32, 30, 27, 214, 29]\n",
            "0.4 M steps/s, Step 74700000, log_posterior = -41336.21, log_G = -43814.75, log_coal = -1748.23, log_other_priors = 4226.77, num_muts = 1106, T = 115704.67, t_MRCA = 1928-08-24, n0 = 6344.8, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.64, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 75150000, log_posterior = -41323.01, log_G = -43807.38, log_coal = -1742.44, log_other_priors = 4226.81, num_muts = 1106, T = 114846.17, t_MRCA = 1929-07-05, n0 = 5722.6, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.91, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 75600000, log_posterior = -41311.15, log_G = -43808.90, log_coal = -1729.19, log_other_priors = 4226.95, num_muts = 1106, T = 114954.21, t_MRCA = 1927-04-04, n0 = 5608.6, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.39, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 76050000, log_posterior = -41315.44, log_G = -43815.17, log_coal = -1727.13, log_other_priors = 4226.85, num_muts = 1106, T = 115611.91, t_MRCA = 1926-08-31, n0 = 4892.7, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 8.49, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 76500000, log_posterior = -41303.53, log_G = -43815.03, log_coal = -1715.55, log_other_priors = 4227.05, num_muts = 1106, T = 115614.45, t_MRCA = 1925-05-13, n0 = 5334.2, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.14, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 76950000, log_posterior = -41354.01, log_G = -43839.03, log_coal = -1741.82, log_other_priors = 4226.85, num_muts = 1106, T = 118266.41, t_MRCA = 1925-09-23, n0 = 5842.5, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.65, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 77400000, log_posterior = -41275.55, log_G = -43793.45, log_coal = -1709.10, log_other_priors = 4226.99, num_muts = 1106, T = 113581.03, t_MRCA = 1926-02-03, n0 = 4910.3, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.82, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 77850000, log_posterior = -41355.36, log_G = -43833.44, log_coal = -1748.47, log_other_priors = 4226.55, num_muts = 1105, T = 119054.01, t_MRCA = 1926-02-03, n0 = 6519.1, g = 0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 8.57, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 78300000, log_posterior = -41312.41, log_G = -43811.04, log_coal = -1728.34, log_other_priors = 4226.97, num_muts = 1106, T = 114923.40, t_MRCA = 1926-04-21, n0 = 5368.3, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.49, pi = [0.32, 0.21, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 35, 25, 9, 203, 195, 34, 26, 25, 250, 31]\n",
            "0.4 M steps/s, Step 78750000, log_posterior = -41363.73, log_G = -43840.95, log_coal = -1749.50, log_other_priors = 4226.73, num_muts = 1105, T = 119951.78, t_MRCA = 1925-08-22, n0 = 6824.5, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 7.5, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 79200000, log_posterior = -41273.29, log_G = -43777.51, log_coal = -1722.95, log_other_priors = 4227.18, num_muts = 1105, T = 113269.00, t_MRCA = 1927-08-06, n0 = 5131.7, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 6.78, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 79650000, log_posterior = -41306.49, log_G = -43809.50, log_coal = -1723.82, log_other_priors = 4226.83, num_muts = 1106, T = 115198.75, t_MRCA = 1926-05-22, n0 = 5634.2, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.9, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 204, 195, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 80100000, log_posterior = -41355.77, log_G = -43840.12, log_coal = -1742.51, log_other_priors = 4226.87, num_muts = 1105, T = 119862.22, t_MRCA = 1922-10-18, n0 = 5779.3, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 7.63, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 35, 25, 9, 203, 195, 34, 26, 25, 249, 31]\n",
            "0.4 M steps/s, Step 80550000, log_posterior = -41433.73, log_G = -43899.23, log_coal = -1761.21, log_other_priors = 4226.71, num_muts = 1105, T = 126512.40, t_MRCA = 1920-12-26, n0 = 7686.2, g = 0.0000, mu = 0.16 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 7.06, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 81000000, log_posterior = -41388.47, log_G = -43875.49, log_coal = -1739.79, log_other_priors = 4226.81, num_muts = 1105, T = 123886.51, t_MRCA = 1918-03-25, n0 = 5721.0, g = -0.0001, mu = 0.17 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 7.93, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 81450000, log_posterior = -41413.29, log_G = -43888.93, log_coal = -1751.23, log_other_priors = 4226.87, num_muts = 1105, T = 125438.13, t_MRCA = 1919-04-02, n0 = 5804.8, g = -0.0001, mu = 0.17 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 7.57, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 81900000, log_posterior = -41446.01, log_G = -43919.13, log_coal = -1753.68, log_other_priors = 4226.80, num_muts = 1106, T = 126644.36, t_MRCA = 1918-07-15, n0 = 6748.8, g = -0.0000, mu = 0.16 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.25, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 35, 25, 9, 203, 196, 35, 26, 24, 249, 31]\n",
            "0.4 M steps/s, Step 82350000, log_posterior = -41418.44, log_G = -43896.89, log_coal = -1748.33, log_other_priors = 4226.78, num_muts = 1106, T = 124630.98, t_MRCA = 1917-05-18, n0 = 6211.5, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.7, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 35, 27, 24, 249, 31]\n",
            "0.4 M steps/s, Step 82800000, log_posterior = -41467.43, log_G = -43926.79, log_coal = -1767.30, log_other_priors = 4226.66, num_muts = 1106, T = 127930.16, t_MRCA = 1920-01-28, n0 = 6680.1, g = 0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.9, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 35, 25, 9, 203, 196, 35, 26, 24, 249, 31]\n",
            "0.4 M steps/s, Step 83250000, log_posterior = -41422.03, log_G = -43900.80, log_coal = -1747.85, log_other_priors = 4226.62, num_muts = 1106, T = 124948.02, t_MRCA = 1919-08-02, n0 = 7052.5, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.83, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 83700000, log_posterior = -41370.05, log_G = -43861.49, log_coal = -1735.45, log_other_priors = 4226.89, num_muts = 1106, T = 120627.13, t_MRCA = 1920-12-16, n0 = 6050.1, g = 0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.32, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 84150000, log_posterior = -41404.74, log_G = -43885.05, log_coal = -1746.28, log_other_priors = 4226.59, num_muts = 1106, T = 122887.33, t_MRCA = 1920-09-01, n0 = 7344.7, g = 0.0000, mu = 0.16 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.81, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.4 M steps/s, Step 84600000, log_posterior = -41404.14, log_G = -43887.53, log_coal = -1743.59, log_other_priors = 4226.98, num_muts = 1107, T = 121515.74, t_MRCA = 1923-08-08, n0 = 6274.5, g = 0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 872, M_tv = 235, kappa = 6.76, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 250, 31]\n",
            "0.3 M steps/s, Step 85050000, log_posterior = -41393.68, log_G = -43880.24, log_coal = -1740.20, log_other_priors = 4226.75, num_muts = 1106, T = 122660.94, t_MRCA = 1920-09-05, n0 = 6859.0, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.35, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 85500000, log_posterior = -41422.92, log_G = -43912.70, log_coal = -1736.99, log_other_priors = 4226.77, num_muts = 1108, T = 122905.43, t_MRCA = 1919-11-11, n0 = 5785.4, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 873, M_tv = 235, kappa = 8.08, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 224, 34, 25, 9, 203, 197, 35, 27, 24, 249, 31]\n",
            "0.3 M steps/s, Step 85950000, log_posterior = -41381.87, log_G = -43867.36, log_coal = -1741.33, log_other_priors = 4226.82, num_muts = 1106, T = 121130.18, t_MRCA = 1921-03-06, n0 = 6690.5, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.16, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 35, 27, 24, 249, 31]\n",
            "0.3 M steps/s, Step 86400000, log_posterior = -41429.63, log_G = -43898.52, log_coal = -1757.97, log_other_priors = 4226.86, num_muts = 1106, T = 124626.78, t_MRCA = 1921-06-27, n0 = 6677.6, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.01, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 35, 27, 24, 249, 31]\n",
            "0.3 M steps/s, Step 86850000, log_posterior = -41412.36, log_G = -43897.32, log_coal = -1742.14, log_other_priors = 4227.10, num_muts = 1107, T = 122815.92, t_MRCA = 1918-12-19, n0 = 5429.4, g = -0.0000, mu = 0.17 * 10^-3 subst / site / year, M_ts = 872, M_tv = 235, kappa = 6.87, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 35, 27, 24, 250, 31]\n",
            "0.3 M steps/s, Step 87300000, log_posterior = -41352.53, log_G = -43860.01, log_coal = -1719.58, log_other_priors = 4227.05, num_muts = 1106, T = 120239.44, t_MRCA = 1918-01-19, n0 = 5923.8, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 6.69, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 35, 25, 9, 203, 196, 34, 26, 25, 249, 31]\n",
            "0.3 M steps/s, Step 87750000, log_posterior = -41419.90, log_G = -43905.72, log_coal = -1741.11, log_other_priors = 4226.93, num_muts = 1106, T = 125368.92, t_MRCA = 1916-09-10, n0 = 6382.2, g = 0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 6.91, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 88200000, log_posterior = -41401.05, log_G = -43892.71, log_coal = -1734.95, log_other_priors = 4226.61, num_muts = 1106, T = 123887.32, t_MRCA = 1916-12-22, n0 = 6603.1, g = 0.0000, mu = 0.16 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 8.2, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 196, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 88650000, log_posterior = -41467.74, log_G = -43945.02, log_coal = -1749.43, log_other_priors = 4226.72, num_muts = 1108, T = 126268.96, t_MRCA = 1918-07-31, n0 = 5664.7, g = -0.0000, mu = 0.18 * 10^-3 subst / site / year, M_ts = 873, M_tv = 235, kappa = 8.42, pi = [0.32, 0.21, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 224, 34, 25, 9, 203, 197, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 89100000, log_posterior = -41452.21, log_G = -43919.81, log_coal = -1759.34, log_other_priors = 4226.94, num_muts = 1105, T = 128866.28, t_MRCA = 1917-06-23, n0 = 6255.0, g = -0.0000, mu = 0.16 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 6.95, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 249, 31]\n",
            "0.3 M steps/s, Step 89550000, log_posterior = -41406.96, log_G = -43904.75, log_coal = -1729.31, log_other_priors = 4227.10, num_muts = 1106, T = 125199.35, t_MRCA = 1915-07-21, n0 = 4495.0, g = -0.0001, mu = 0.16 * 10^-3 subst / site / year, M_ts = 871, M_tv = 235, kappa = 7.69, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 250, 31]\n",
            "0.3 M steps/s, Step 90000000, log_posterior = -41503.88, log_G = -43955.78, log_coal = -1774.63, log_other_priors = 4226.53, num_muts = 1105, T = 133260.45, t_MRCA = 1914-11-27, n0 = 7660.6, g = -0.0000, mu = 0.16 * 10^-3 subst / site / year, M_ts = 870, M_tv = 235, kappa = 7.88, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4044, 3745, 5108], num_muts_ab counts: [50, 223, 34, 25, 9, 203, 195, 34, 27, 25, 249, 31]\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "    <h2 style='color:green'>All done! üéâ</h2>\n",
              "    <p>Files generated in this notebook are in the left-hand file browser (or in the folder you set as <code>outfolder</code>).</p>\n",
              "    <ul>\n",
              "      <li>Aligned FASTA: <code>Zaire_ebolavirus_aligned_headers_adjusted.afa</code></li>\n",
              "      <li>Delphy output (.dphy): <code>Zaire_ebolavirus_delphy_out.dphy</code></li>\n",
              "      <li>Delphy log (.log): <code>Zaire_ebolavirus_delphy_beast.log</code></li>\n",
              "      <li>Delphy trees (.trees): <code>Zaire_ebolavirus_delphy_beast.trees</code></li>\n",
              "    </ul>\n",
              "    <p>We recommend downloading the <code>.afa</code>, <code>.dphy</code>, and <code>metadata.csv</code> files. To visualize .dphy files online, try <a href='https://delphy.fathom.info/' target='_blank'>delphy.fathom.info</a>.</p>\n",
              "    "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CPU times: user 4.07 s, sys: 536 ms, total: 4.61 s\n",
            "Wall time: 5min 31s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "HdHHacImXoDe"
      },
      "execution_count": 6,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
