{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/lauraluebbert/delphy_workflows/blob/main/delphy_workflow_sars_example.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ej4vLF8GBtT_"
      },
      "source": [
        "<div style=\"text-align: center;\">\n",
        "    <h1 style=\"font-size: 2.5em; color: #2C3E50;\">Running <a href=\"https://github.com/broadinstitute/delphy\" target=\"_blank\">Delphy</a> is as simple as 1, 2, 3 â€“ SARS-CoV2 Example</h1>\n",
        "    <h2 style=\"font-size: 1.5em; color: #34495E;\">Phylogenetic Tree Generation with Delphy on SARS-CoV2 Sequences</h2>\n",
        "</div>\n",
        "\n",
        "This notebook demonstrates how to generate a phylogenetic tree using [Delphy](https://github.com/broadinstitute/delphy) on viral sequences obtained from the [NCBI Virus Database](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/). Optionally, you can upload your own sequences to be included in the analysis.\n",
        "\n",
        "We will utilize the following tools:\n",
        "- [**gget**](https://github.com/pachterlab/gget/) to download sequences from the NCBI Virus Database\n",
        "- [**MAFFT**](https://mafft.cbrc.jp/alignment/server/index.html) for creating a Multiple-Sequence Alignment (MSA)\n",
        "- [**Delphy**](https://github.com/broadinstitute/delphy) to generate the phylogenetic tree\n",
        "\n",
        "In this example, we already filled out the filters and options below to download and analyze **SARS-CoV2 sequences from *Homo sapiens* hosts** collected between January 15, 2025 and February 15, 2025 bold text. **Simply click `Runtime` -> `Run all`** at the top of this notebook.\n",
        "\n",
        "If you encounter any problems or questions while using this notebook, please [report them here](https://github.com/broadinstitute/delphy/issues).\n",
        "\n",
        "Notebook runtime:\n",
        "___\n",
        "___"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Apply filters to download sequences from [NCBI Virus](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/)"
      ],
      "metadata": {
        "id": "sWOSpHI-iGVO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title NCBI Virus filtering options:\n",
        "#@markdown Set any filter to `None` to disable it.\n",
        "\n",
        "def arg_str_to_bool(arg):\n",
        "  if arg == \"True\":\n",
        "      return True\n",
        "  elif arg == \"False\":\n",
        "      return False\n",
        "  elif arg == \"None\" or arg == \"\":\n",
        "      return None\n",
        "  else:\n",
        "      return arg\n",
        "\n",
        "#@markdown ## **Virus**\n",
        "\n",
        "virus = 'sars-cov-2'  #@param {type:\"string\"}\n",
        "#@markdown  - Examples: 'Mammarenavirus lassaense' or 'coronaviridae' or 'NC_045512.2' or '142786' (Norovirus taxid).\n",
        "virus = arg_str_to_bool(virus)\n",
        "\n",
        "accession = False   #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if `virus` argument above is an NCBI accession (starts with 'NC'), e.g. 'NC_045512.2'.\n",
        "\n",
        "#@markdown ## **Host**\n",
        "\n",
        "host = 'homo sapiens'  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'homo sapiens' (alternative: use the `host_taxid` filter below). Input `None` to disable filtering by host.\n",
        "host = arg_str_to_bool(host)\n",
        "\n",
        "host_taxid = None  #@param {type:\"raw\"}\n",
        "#@markdown  - NCBI Taxonomy ID of host (e.g., 9443 for primates).\n",
        "host_taxid = arg_str_to_bool(host_taxid)\n",
        "\n",
        "#@markdown ## **Sequence completeness**\n",
        "\n",
        "annotated = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Set to `True` to only return sequences marked as 'annotated'; set to `False` to only return sequences NOT marked as 'annotated'.\n",
        "annotated = arg_str_to_bool(annotated)\n",
        "\n",
        "nuc_completeness = \"complete\"  #@param [\"None\", \"complete\", \"partial\"]\n",
        "#@markdown  - Choose between 'partial' or 'complete' nucleotide completeness.\n",
        "nuc_completeness = arg_str_to_bool(nuc_completeness)\n",
        "\n",
        "min_seq_length = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum sequence length, e.g. 6252.\n",
        "min_seq_length = arg_str_to_bool(min_seq_length)\n",
        "\n",
        "max_seq_length = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum sequence length, e.g. 7815.\n",
        "max_seq_length = arg_str_to_bool(max_seq_length)\n",
        "\n",
        "has_proteins = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Require sequences to contain specific proteins (e.g. 'GPC' **- include the quotation marks**) or a list of proteins (e.g. ['GPC', 'L']). Also accepts names of genes or segments.\n",
        "has_proteins = arg_str_to_bool(has_proteins)\n",
        "\n",
        "proteins_complete = False    #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if the proteins/genes/segments in `has_proteins` should be marked as 'complete'.\n",
        "proteins_complete = arg_str_to_bool(proteins_complete)\n",
        "\n",
        "max_ambiguous_chars = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum number of 'N' characters allowed in each sequence, e.g. 10.\n",
        "max_ambiguous_chars = arg_str_to_bool(max_ambiguous_chars)\n",
        "\n",
        "#@markdown ## **Gene/peptide/protein counts**\n",
        "\n",
        "min_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum gene count, e.g. 1.\n",
        "min_gene_count = arg_str_to_bool(min_gene_count)\n",
        "\n",
        "max_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum gene count, e.g. 40.\n",
        "max_gene_count = arg_str_to_bool(max_gene_count)\n",
        "\n",
        "min_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum peptide count, e.g. 2.\n",
        "min_mature_peptide_count = arg_str_to_bool(min_mature_peptide_count)\n",
        "\n",
        "max_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum peptide count, e.g. 15.\n",
        "max_mature_peptide_count = arg_str_to_bool(max_mature_peptide_count)\n",
        "\n",
        "min_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Minimum protein count, e.g. 2.\n",
        "min_protein_count = arg_str_to_bool(min_protein_count)\n",
        "\n",
        "max_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Maximum protein count, e.g. 10.\n",
        "max_protein_count = arg_str_to_bool(max_protein_count)\n",
        "\n",
        "#@markdown ## **Geographic location**\n",
        "\n",
        "geographic_location = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic location of sample collection, e.g. 'South Africa' or 'Germany'.\n",
        "geographic_location = arg_str_to_bool(geographic_location)\n",
        "\n",
        "geographic_region = None  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic region of sample collection, e.g. 'Africa' or 'Europe'.\n",
        "geographic_region = arg_str_to_bool(geographic_region)\n",
        "\n",
        "#@markdown ## **Dates**\n",
        "#@markdown All dates should be supplied in YYYY-MM-DD format.\n",
        "min_collection_date = \"2025-01-15\"  #@param {type:\"string\"}\n",
        "#@markdown  - Minimum collection date, e.g. '2000-01-01'.\n",
        "min_collection_date = arg_str_to_bool(min_collection_date)\n",
        "\n",
        "max_collection_date = \"2025-02-15\"  #@param {type:\"string\"}\n",
        "#@markdown  - Maximum collection date, e.g. '2014-12-04'.\n",
        "max_collection_date = arg_str_to_bool(max_collection_date)\n",
        "\n",
        "min_release_date = \"2025-01-01\"  #@param {type:\"string\"}\n",
        "#@markdown  - Minimum release date of the sequences, e.g. '2000-01-01'.\n",
        "min_release_date = arg_str_to_bool(min_release_date)\n",
        "\n",
        "max_release_date = \"None\"  #@param {type:\"string\"}\n",
        "#@markdown  - Maximum release date of the sequences, e.g. '2014-12-04'.\n",
        "max_release_date = arg_str_to_bool(max_release_date)\n",
        "\n",
        "#@markdown ## **Source**\n",
        "\n",
        "submitter_country = None  #@param {type:\"string\"}\n",
        "#@markdown  - Country that submitted the sequence, e.g. 'South Africa' or 'Germany'.\n",
        "submitter_country = arg_str_to_bool(submitter_country)\n",
        "\n",
        "lab_passaged = \"None\"   #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - Set to `True` to only return sequences that have been passaged in a laboratory setting; set to `False` to only return sequences that have NOT been passaged in a laboratory setting.\n",
        "lab_passaged = arg_str_to_bool(lab_passaged)\n",
        "\n",
        "source_database = None  #@param {type:\"string\"}\n",
        "#@markdown  - Source database of the sequence, e.g. 'GenBank' or 'RefSeq'.\n",
        "source_database = arg_str_to_bool(source_database)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "6cazzIZfiJDU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Optional: Upload a fasta file with your own sequences to add to the analysis\n",
        "  **1) Click on the folder icon on the left.  \n",
        "  2) Upload your file(s) to the Google Colab server by dragging in your file(s) (or use rightclick -> Upload).  \n",
        "  3) Specify the name of your file(s) here:**"
      ],
      "metadata": {
        "id": "aE8ot-AyiPdC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title FASTA file containing additional sequences\n",
        "\n",
        "fasta_file = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_fasta_file.fa' or 'my_fasta_file.fasta'.\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "E-lg25zave0H"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Metadata\n",
        "\n",
        "#@markdown **Option 1: The metadata is the same for all sequences in your FASTA file**\n",
        "metadata = {'Collection Date': 'YYYY-MM-DD', 'Geo Location': 'South Korea'}  #@param {type:\"raw\"}\n",
        "#@markdown - The 'Collection Date' field is required. Optional: you can add as many additional columns as you wish, e.g. 'Geo Location': 'South Korea'.\n",
        "#@markdown - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "#@markdown **Option 2: Input a CSV file with metadata for each sequence**\n",
        "metadata_csv = None  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_metadata.csv'. This file must include at least an 'Accession' and 'Collection Date' column.\n",
        "#@markdown  - NOTE: Make sure the IDs in the \"Accession\" column match the IDs of the sequences in the provided FASTA file\n",
        "#@markdown  - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "# Convert empty strings to None\n",
        "fasta_file = arg_str_to_bool(fasta_file)\n",
        "metadata_csv = arg_str_to_bool(metadata_csv)"
      ],
      "metadata": {
        "id": "LKgxMFBgiP64",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Optional: Specify Delphy arguments"
      ],
      "metadata": {
        "id": "zQF5zZ2dzS60"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Delphy options:\n",
        "mutation_rate = 0.01  #@param {type:\"raw\"}\n",
        "#@markdown  - Virus mutation rate (mutations per site per year), e.g. 0.01. If set to `None`, the mutation rate will be estimated from the generated tree.\n",
        "\n",
        "delphy_steps = 10000000  #@param {type:\"integer\"}\n",
        "#@markdown  - Number of steps to run in the Delphy algorithm (default: 10,000,000).\n",
        "\n",
        "log_freq = 100 #@param {type:\"integer\"}\n",
        "#@markdown  - Frequency of logging and tree updates, relative to the total number of steps (will log every `delphy_steps/log_freq` step).\n",
        "log_every = int(delphy_steps / log_freq)\n",
        "tree_every = int(delphy_steps / log_freq)\n",
        "\n",
        "delphy_release = 0.999932  #@param {type:\"number\"}\n",
        "#@markdown  - Delphy version to use (see https://github.com/broadinstitute/delphy/releases).\n",
        "\n",
        "#@title Delphy options:\n",
        "threads = 2  #@param {type:\"integer\"}\n",
        "#@markdown  - Number of threads to use for the MSA (Mafft) and phylogenetic tree (Delphy) generation."
      ],
      "metadata": {
        "cellView": "form",
        "id": "bEM5nlzZzSih"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# All done! Select `Runtime` at the top of this notebook, then click `Run all` and lean back...\n",
        "A completion message will be displayed below once the notebook has been successfully executed.  \n",
        "**ðŸ’¡ Tip: Click on the folder icon on the left to view/download the files that are being generated.**\n",
        "  \n",
        "<br>\n",
        "\n",
        "____\n",
        "____"
      ],
      "metadata": {
        "id": "UTL55NF2jAj-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "#@title # Generating the phylogenetic tree...\n",
        "import subprocess\n",
        "from IPython.display import display, HTML\n",
        "\n",
        "def log_message(text):\n",
        "    display(HTML(f\"<h2 style='color: green;'>{text}</h2>\"))\n",
        "def log_message_error(text):\n",
        "    display(HTML(f\"<h2 style='color: red;'>{text}</h2>\"))\n",
        "\n",
        "\n",
        "log_message(\"1/5 Installing software...\")\n",
        "\n",
        "# Install gget\n",
        "try:\n",
        "    result = subprocess.run(['pip', 'install', '-q', 'mysql-connector-python', 'biopython'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout)\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while installing mysql-connector-python and biopython: \")\n",
        "    print(e.stderr.decode())\n",
        "\n",
        "try:\n",
        "    # After the release, this will just be: pip install gget (dependence on biopython will be removed)\n",
        "    result = subprocess.run(['pip', 'install', '-q', 'git+https://github.com/pachterlab/gget.git@delphy_dev'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout)\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while installing gget: \")\n",
        "    print(e.stderr.decode())\n",
        "\n",
        "\n",
        "import gget\n",
        "from Bio import SeqIO\n",
        "import pandas as pd\n",
        "import re\n",
        "from datetime import datetime\n",
        "\n",
        "log_message(\"1/5 Software installation complete.\")\n",
        "\n",
        "\n",
        "# Downloading virus genomes from NCBI Virus\n",
        "log_message(\"2/5 Download data from NCBI Virus... This might take a few minutes depending on the internet connection and how busy the NCBI server is.\")\n",
        "gget.ncbi_virus(\n",
        "    virus = virus,\n",
        "    accession = accession,\n",
        "    host = host,\n",
        "    min_seq_length = min_seq_length,\n",
        "    max_seq_length = max_seq_length,\n",
        "    min_gene_count = min_gene_count,\n",
        "    max_gene_count = max_gene_count,\n",
        "    nuc_completeness = nuc_completeness,\n",
        "    has_proteins = has_proteins,\n",
        "    proteins_complete = proteins_complete,\n",
        "    host_taxid = host_taxid,\n",
        "    lab_passaged = lab_passaged,\n",
        "    geographic_region = geographic_region,\n",
        "    geographic_location = geographic_location,\n",
        "    submitter_country = submitter_country,\n",
        "    min_collection_date = min_collection_date,\n",
        "    max_collection_date = max_collection_date,\n",
        "    annotated = annotated,\n",
        "    source_database = source_database,\n",
        "    min_release_date = min_release_date,\n",
        "    max_release_date = max_release_date,\n",
        "    min_mature_peptide_count = min_mature_peptide_count,\n",
        "    max_mature_peptide_count = max_mature_peptide_count,\n",
        "    min_protein_count = min_protein_count,\n",
        "    max_protein_count = max_protein_count,\n",
        "    max_ambiguous_chars = max_ambiguous_chars\n",
        ")\n",
        "\n",
        "ncbi_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences.fasta\"\n",
        "ncbi_metadata = f\"{'_'.join(str(virus).split(' '))}_metadata.csv\"\n",
        "\n",
        "log_message(f\"2/5 Data download from NCBI Virus complete. Sequences were saved in the {ncbi_fasta_file} file and the associated metadata in {ncbi_metadata}/.jsonl\")\n",
        "\n",
        "\n",
        "# Merging sequencing and metadata files if additional file(s) were provided\n",
        "if fasta_file:\n",
        "  log_message(\"Adding user-provided fasta file and metadata to the data from NCBI Virus...\")\n",
        "\n",
        "  # Combine sequence files\n",
        "  combined_fasta_file = f\"{'_'.join(str(virus).split(' '))}_sequences_combined.fasta\"\n",
        "  !cat $ncbi_fasta_file $fasta_file > $combined_fasta_file\n",
        "  input_fasta_file = combined_fasta_file\n",
        "\n",
        "  # Combine metadata\n",
        "  combined_metadata_file = f\"{'_'.join(virus.split(' '))}_metadata_combined.csv\"\n",
        "  ncbi_metadata_df = pd.read_csv(ncbi_metadata)\n",
        "  if metadata_csv:\n",
        "    # Combine provided metadata and NCBI metadata csv files\n",
        "    user_metada_df = pd.read_csv(metadata_csv)\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  else:\n",
        "    # Extract sequence accessions from the provided FASTA file\n",
        "    headers = [record.id.split(\" \")[0] for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
        "\n",
        "    # Create a metadata dataframe with the accessions from the FASTA file and the provided metadata\n",
        "    user_metada_df = pd.DataFrame(headers, columns=[\"Accession\"])\n",
        "    for key, value in metadata.items():\n",
        "      user_metada_df[key] = value\n",
        "\n",
        "    # Combine with NCBI metadata\n",
        "    comb_meta_df = pd.concat([ncbi_metadata_df, user_metada_df])\n",
        "    comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "    metadata_file = combined_metadata_file\n",
        "\n",
        "  log_message(f\"Merging user-provided and NCBI Virus data complete. The combined sequence and metadata files were saved as {combined_fasta_file} and {combined_metadata_file}.\")\n",
        "\n",
        "else:\n",
        "  input_fasta_file = ncbi_fasta_file\n",
        "  metadata_file = ncbi_metadata\n",
        "\n",
        "\n",
        "# Create MSA\n",
        "log_message(\"3/5 Multiple Sequence Aligment (MSA): Aligning the sequences to each other so they are all in the same frame...\")\n",
        "\n",
        "aligned_fasta_file = f\"{'_'.join(str(virus).split(' '))}_aligned.afa\"\n",
        "\n",
        "# # Option 1: Using the MUSCLE algorithm (this works well for a few hundred sequences, but is too slow when dealing with a few thousand sequences)\n",
        "# gget.muscle(input_fasta_file, super5=True, out=aligned_fasta_file)\n",
        "\n",
        "# Option 2: Using mafft\n",
        "# TO-DO: Wrap the following code into gget module and replace with command `gget.mafft(input_fasta_file, out=aligned_fasta_file)`\n",
        "\n",
        "#Install MAFFT\n",
        "try:\n",
        "    result = subprocess.run(['apt-get', 'install', '-qq', '-y', 'mafft'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout.decode())\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while installing Mafft: \")\n",
        "    print(e.stderr.decode())\n",
        "\n",
        "# Aligning sequences to each other using mafft\n",
        "try:\n",
        "    with open(aligned_fasta_file, 'w') as output:\n",
        "        result = subprocess.run(['mafft', '--auto', '--thread', str(threads), input_fasta_file], stdout=output, stderr=subprocess.PIPE, check=True)\n",
        "    log_message(f\"3/5 MSA complete and saved in the {aligned_fasta_file} file.\")\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message(f\"An error occurred while generating the MSA using Mafft: \")\n",
        "    print(e.stderr.decode())\n",
        "\n",
        "\n",
        "# TO-DO: Wrap the following code into a gget module and replace with command `gget.delphy(aligned_fasta_file, metadata_file)`\n",
        "# Adjust the headers in the aligned fasta file to match header format required by Delphy (accession|YYYY-MM-DD):\n",
        "log_message(\"4/5 Reformatting sequence files to match Delphy format...\")\n",
        "\n",
        "# Reformat collection date\n",
        "default_day = '01'\n",
        "default_month = '01'\n",
        "def extract_and_format_date(date_string):\n",
        "    # Define regular expressions for various date formats\n",
        "    year_only = re.compile(r'(?P<year>\\d{4})')\n",
        "    year_month = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})')\n",
        "    full_date = re.compile(r'(?P<year>\\d{4})[-/.](?P<month>\\d{1,2})[-/.](?P<day>\\d{1,2})')\n",
        "\n",
        "    # Try to match the full date first\n",
        "    match = full_date.search(date_string)\n",
        "    if match:\n",
        "        year = match.group('year')\n",
        "        month = match.group('month').zfill(2)\n",
        "        day = match.group('day').zfill(2)\n",
        "    else:\n",
        "        # Try to match year and month\n",
        "        match = year_month.search(date_string)\n",
        "        if match:\n",
        "            year = match.group('year')\n",
        "            month = match.group('month').zfill(2)\n",
        "            day = default_day\n",
        "        else:\n",
        "            # Try to match only the year\n",
        "            match = year_only.search(date_string)\n",
        "            if match:\n",
        "                year = match.group('year')\n",
        "                month = default_month\n",
        "                day = default_day\n",
        "            else:\n",
        "                # If no match, return None\n",
        "                return None\n",
        "\n",
        "    # Format the extracted date into YYYY-MM-DD\n",
        "    formatted_date = f\"{year}-{month}-{day}\"\n",
        "\n",
        "    try:\n",
        "        # Validate date by trying to convert it to a datetime object\n",
        "        datetime.strptime(formatted_date, '%Y-%m-%d')\n",
        "    except ValueError:\n",
        "        return None  # Return None if the date is invalid\n",
        "\n",
        "    return formatted_date\n",
        "\n",
        "def update_fasta_headers(fasta_file, csv_file, output_fasta):\n",
        "    # Read the CSV file into a pandas DataFrame\n",
        "    df = pd.read_csv(csv_file)\n",
        "\n",
        "    # Create a dictionary from the DataFrame to map accession to date\n",
        "    accession_to_date = pd.Series(df['Collection Date'].values, index=df['accession']).to_dict()\n",
        "\n",
        "    # Open the input fasta file and output fasta file\n",
        "    with open(fasta_file) as fasta_input, open(output_fasta, 'w') as fasta_output:\n",
        "        # Iterate through each sequence record in the fasta file\n",
        "        for record in SeqIO.parse(fasta_input, 'fasta'):\n",
        "            accession = record.id\n",
        "\n",
        "            # Check if the accession is in the pandas dictionary and has a non-NaN date\n",
        "            if accession in accession_to_date and pd.notna(accession_to_date[accession]):\n",
        "                date = accession_to_date[accession]\n",
        "\n",
        "                # Format the date if necessary\n",
        "                formatted_date = extract_and_format_date(date)\n",
        "\n",
        "                if formatted_date is None:\n",
        "                  # Skip the entry if date is NaN or accession not found\n",
        "                  print(f\"Skipping accession {accession} due to unrecognized date format: '{date}'\")\n",
        "                  continue\n",
        "\n",
        "                # Update the seq header\n",
        "                record.id = f\"{accession}|{formatted_date}\"\n",
        "                record.description = ''  # Remove the original description to avoid duplication\n",
        "            else:\n",
        "                # Skip the entry if date is NaN or accession not found\n",
        "                print(f\"Skipping accession {accession} due to missing or NaN date.\")\n",
        "                continue\n",
        "\n",
        "            # Write the updated record to the output fasta file\n",
        "            SeqIO.write(record, fasta_output, 'fasta')\n",
        "\n",
        "aligned_fasta_file_clean = f\"{'_'.join(str(virus).split(' '))}_aligned_headers_adjusted.afa\"\n",
        "update_fasta_headers(aligned_fasta_file, metadata_file, aligned_fasta_file_clean)\n",
        "\n",
        "log_message(\"4/5 Reformatting complete.\")\n",
        "\n",
        "\n",
        "# Run Delphy\n",
        "log_message(\"5/5 Running Delphy...\")\n",
        "\n",
        "# Download Delphy binary\n",
        "try:\n",
        "    result = subprocess.run(['wget', f'https://github.com/broadinstitute/delphy/releases/download/{delphy_release}/delphy-ubuntu-x86_64'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout.decode())\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while downloading Delphy: \")\n",
        "    print(e.stderr.decode())\n",
        "\n",
        "# Give permissions\n",
        "try:\n",
        "    result = subprocess.run(['chmod', 'u+x', './delphy-ubuntu-x86_64'], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "    # print(result.stdout.decode())\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while changing permissions for Delphy: \")\n",
        "    print(e.stderr.decode())\n",
        "\n",
        "beast_log_out = f\"{'_'.join(str(virus).split(' '))}_delphy_beast_log.txt\"\n",
        "delphy_beast_tree_out = f\"{'_'.join(str(virus).split(' '))}_delphy_beast_tree.nwk\"\n",
        "dphy_out = f\"{'_'.join(str(virus).split(' '))}_delphy_out.dphy\"\n",
        "\n",
        "# Define the command\n",
        "if mutation_rate:\n",
        "    cmd = [\n",
        "        './delphy-ubuntu-x86_64',\n",
        "        '--v0-fix-mutation-rate',\n",
        "        '--v0-init-mutation-rate', str(mutation_rate),\n",
        "        '--v0-log-every', str(log_every),\n",
        "        '--v0-tree-every', str(tree_every),\n",
        "        '--v0-threads', str(threads),\n",
        "        '--v0-steps', str(delphy_steps),\n",
        "        '--v0-in-fasta', aligned_fasta_file_clean,\n",
        "        '--v0-out-log-file', beast_log_out,\n",
        "        '--v0-out-trees-file', delphy_beast_tree_out,\n",
        "        '--v0-out-delphy-file', dphy_out\n",
        "    ]\n",
        "else:\n",
        "    cmd = [\n",
        "        './delphy-ubuntu-x86_64',\n",
        "        '--v0-log-every', str(log_every),\n",
        "        '--v0-tree-every', str(tree_every),\n",
        "        '--v0-threads', str(threads),\n",
        "        '--v0-steps', str(delphy_steps),\n",
        "        '--v0-in-fasta', aligned_fasta_file_clean,\n",
        "        '--v0-out-log-file', beast_log_out,\n",
        "        '--v0-out-trees-file', delphy_beast_tree_out,\n",
        "        '--v0-out-delphy-file', dphy_out\n",
        "    ]\n",
        "\n",
        "try:\n",
        "    result = subprocess.run(cmd, check=True, text=True, capture_output=True)\n",
        "    # print(result.stdout.decode())\n",
        "\n",
        "    display(HTML(\"\"\"\n",
        "    <h1>All done! ðŸŽ‰</h1>\n",
        "    <h2>To download the files we generated in this notebook to your local computer, click on the folder icon on the left and download files by right clicking a file of interest and selecting 'Download'.</h2>\n",
        "    <h2>To further visualize your Delphy output, upload the <code>.dphy</code> (and <code>metadata.csv</code>) file(s) to <a href='https://delphy.fathom.info/' target='_blank'>https://delphy.fathom.info/</a></h2>\n",
        "    \"\"\"))\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(f\"An error occurred while running Delphy: \")\n",
        "    print(e.stderr.decode())"
      ],
      "metadata": {
        "cellView": "form",
        "id": "HuVr7cb0x5VF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "ihN2Y1S1cSYC"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}