{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/broadinstitute/delphy/blob/main/tutorials/delphy_workflow_example.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ej4vLF8GBtT_"
      },
      "source": [
        "<div style=\"text-align: center;\">\n",
        "    <h1 style=\"font-size: 2.5em; color: #2C3E50;\">Running <a href=\"https://github.com/broadinstitute/delphy\" target=\"_blank\">Delphy</a> is as simple as 1, 2, 3 ‚Äì Ebolavirus Example</h1>\n",
        "    <h2 style=\"font-size: 1.5em; color: #34495E;\">Phylogenetic Tree Generation with Delphy on Ebolavirus Sequences</h2>\n",
        "</div>\n",
        "\n",
        "This notebook demonstrates how to generate a phylogenetic tree using [Delphy](https://github.com/broadinstitute/delphy) on viral sequences obtained from the [NCBI Virus Database](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/). Optionally, you can also upload your own sequences to be included in the analysis.\n",
        "\n",
        "We will utilize the following tools:\n",
        "- [**gget**](https://pachterlab.github.io/gget/en/virus.html) to download viral sequences from the [NCBI Virus Database](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/)\n",
        "- [**MAFFT**](https://mafft.cbrc.jp/alignment/server/index.html) for creating a Multiple-Sequence Alignment (MSA)\n",
        "- [**Delphy**](https://github.com/broadinstitute/delphy) to generate the phylogenetic tree\n",
        "\n",
        "In this example notebook, we already filled out the NCBI Virus filters and options below to download and analyze all **complete *Zaire ebolavirus* sequences from *Homo sapiens* hosts collected between January 01, 2014 and June 30, 2014**. All generated files will be saved in the **results/** folder, unless you specify a different output folder below.  \n",
        "\n",
        "Simply **click `Runtime` -> `Run all`** at the top to run this example notebook.\n",
        "\n",
        "If you encounter any problems or questions while using this notebook, please [report them here](https://github.com/broadinstitute/delphy/issues).\n",
        "\n",
        "Total runtime: ~4 min\n",
        "___\n",
        "___"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Apply filters to download sequences from [NCBI Virus](https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/)"
      ],
      "metadata": {
        "id": "sWOSpHI-iGVO"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title gget virus filtering options:\n",
        "#@markdown Set any non-boolean filter to `None` to disable it.\n",
        "\n",
        "def arg_str_to_bool(arg):\n",
        "    if arg == \"True\":\n",
        "        return True\n",
        "    elif arg == \"False\":\n",
        "        return False\n",
        "    elif arg == \"None\" or arg == \"\":\n",
        "        return None\n",
        "    else:\n",
        "        return arg\n",
        "\n",
        "# -------------------------\n",
        "# Virus query\n",
        "# -------------------------\n",
        "#@markdown ## **Virus**\n",
        "\n",
        "virus = \"Zaire ebolavirus\"  #@param {type:\"string\"}\n",
        "#@markdown  - Examples: 'Mammarenavirus lassaense' or 'coronaviridae' or 'NC_045512.2' or '142786' (Norovirus taxid).\n",
        "virus = arg_str_to_bool(virus)\n",
        "\n",
        "is_accession = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Check this box if `virus` is given as an NCBI accession (starts with 'NC_').\n",
        "#@markdown  - For SARS-CoV-2 / Influenza A optimized downloads also check the corresponding box below (see `is_sars_cov2` / `is_alphainfluenza` below)\n",
        "is_accession = arg_str_to_bool(is_accession)\n",
        "\n",
        "download_all_accessions = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - ‚ö†Ô∏è Downloads ALL virus accessions from NCBI (very large database).\n",
        "download_all_accessions = arg_str_to_bool(download_all_accessions)\n",
        "\n",
        "# -------------------------\n",
        "# Host filters\n",
        "# -------------------------\n",
        "#@markdown ## **Host**\n",
        "\n",
        "host = \"Homo sapiens\"  #@param {type:\"string\"}\n",
        "#@markdown  - Host organism name OR NCBI Taxonomy ID (e.g., 'human', 'Aedes aegypti', '1335626').\n",
        "#@markdown  - Input 'None' to disable filtering by host.\n",
        "host = arg_str_to_bool(host)\n",
        "\n",
        "# -------------------------\n",
        "# Sequence & gene filters\n",
        "# -------------------------\n",
        "#@markdown ## **Sequence & gene filters**\n",
        "\n",
        "annotated = \"None\"  #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - True: only annotated sequences. False: only NOT annotated. None: no filter.\n",
        "annotated = arg_str_to_bool(annotated)\n",
        "\n",
        "refseq_only = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Limit search to RefSeq genomes only (curated).\n",
        "refseq_only = arg_str_to_bool(refseq_only)\n",
        "\n",
        "nuc_completeness = \"None\"  #@param [\"None\", \"complete\", \"partial\"]\n",
        "#@markdown  - Set to 'complete' to only return full-length viral genomes; set to 'partial' to only return sequences that are NOT full-length viral genomes.\n",
        "nuc_completeness = arg_str_to_bool(nuc_completeness)\n",
        "\n",
        "min_seq_length = None  #@param {type:\"raw\"}\n",
        "max_seq_length = None  #@param {type:\"raw\"}\n",
        "\n",
        "min_seq_length = arg_str_to_bool(min_seq_length)\n",
        "max_seq_length = arg_str_to_bool(max_seq_length)\n",
        "\n",
        "#@markdown\n",
        "max_ambiguous_chars = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Max number of ambiguous nucleotide characters (N's).\n",
        "max_ambiguous_chars = arg_str_to_bool(max_ambiguous_chars)\n",
        "\n",
        "min_gene_count = None  #@param {type:\"raw\"}\n",
        "max_gene_count = None  #@param {type:\"raw\"}\n",
        "#@markdown\n",
        "min_gene_count = arg_str_to_bool(min_gene_count)\n",
        "max_gene_count = arg_str_to_bool(max_gene_count)\n",
        "#@markdown\n",
        "min_protein_count = None  #@param {type:\"raw\"}\n",
        "max_protein_count = None  #@param {type:\"raw\"}\n",
        "#@markdown\n",
        "min_protein_count = arg_str_to_bool(min_protein_count)\n",
        "max_protein_count = arg_str_to_bool(max_protein_count)\n",
        "#@markdown\n",
        "min_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "max_mature_peptide_count = None  #@param {type:\"raw\"}\n",
        "#@markdown\n",
        "min_mature_peptide_count = arg_str_to_bool(min_mature_peptide_count)\n",
        "max_mature_peptide_count = arg_str_to_bool(max_mature_peptide_count)\n",
        "\n",
        "has_proteins = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Require specific proteins/genes (e.g. \"spike\") or list (e.g. [\"spike\",\"ORF1ab\"]). **Include quotation marks.**\n",
        "has_proteins = arg_str_to_bool(has_proteins)\n",
        "\n",
        "proteins_complete = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Only include sequences where all annotated proteins are complete.\n",
        "proteins_complete = arg_str_to_bool(proteins_complete)\n",
        "\n",
        "# -------------------------\n",
        "# Location & submitter filters\n",
        "# -------------------------\n",
        "#@markdown ## **Location & submitter filters**\n",
        "\n",
        "geographic_location = \"\"  #@param {type:\"string\"}\n",
        "#@markdown  - Geographic location of sample collection (e.g. 'USA', 'Asia').\n",
        "geographic_location = arg_str_to_bool(geographic_location)\n",
        "\n",
        "submitter_country = \"\"  #@param {type:\"string\"}\n",
        "#@markdown  - Country of the sequence submitter.\n",
        "submitter_country = arg_str_to_bool(submitter_country)\n",
        "\n",
        "lab_passaged = \"None\"  #@param [\"True\", \"False\", \"None\"]\n",
        "#@markdown  - True: only lab-passaged. False: exclude lab-passaged. None: no filter.\n",
        "lab_passaged = arg_str_to_bool(lab_passaged)\n",
        "\n",
        "# -------------------------\n",
        "# Date filters\n",
        "# -------------------------\n",
        "#@markdown ## **Dates**\n",
        "#@markdown All dates should be supplied in YYYY-MM-DD format.\n",
        "\n",
        "min_collection_date = \"2014-01-01\"  #@param {type:\"string\"}\n",
        "max_collection_date = \"2014-06-30\"  #@param {type:\"string\"}\n",
        "\n",
        "#@markdown\n",
        "min_release_date = \"\"  #@param {type:\"string\"}\n",
        "max_release_date = \"\"  #@param {type:\"string\"}\n",
        "\n",
        "min_collection_date = arg_str_to_bool(min_collection_date)\n",
        "max_collection_date = arg_str_to_bool(max_collection_date)\n",
        "\n",
        "min_release_date = arg_str_to_bool(min_release_date)\n",
        "max_release_date = arg_str_to_bool(max_release_date)\n",
        "\n",
        "# -------------------------\n",
        "# SARS-CoV-2 specific filters / optimizations\n",
        "# -------------------------\n",
        "#@markdown ## **SARS-CoV-2 / Influenza A optimizations**\n",
        "\n",
        "lineage = \"\"  #@param {type:\"string\"}\n",
        "#@markdown  - SARS-CoV-2 lineage filter (e.g. 'B.1.1.7'). (Only meaningful for SARS-CoV-2 queries.)\n",
        "lineage = arg_str_to_bool(lineage)\n",
        "\n",
        "is_sars_cov2 = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Use optimized cached downloads for SARS-CoV-2.\n",
        "is_sars_cov2 = arg_str_to_bool(is_sars_cov2)\n",
        "\n",
        "is_alphainfluenza = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Use optimized cached downloads for Influenza A virus (Alphainfluenza).\n",
        "is_alphainfluenza = arg_str_to_bool(is_alphainfluenza)\n",
        "\n",
        "# -------------------------\n",
        "# Optional GenBank metadata enrichment\n",
        "# -------------------------\n",
        "#@markdown ## **GenBank metadata enrichment**\n",
        "\n",
        "genbank_metadata = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Fetch additional detailed metadata from GenBank into a separate CSV/XML/CSV dumps.\n",
        "genbank_metadata = arg_str_to_bool(genbank_metadata)\n",
        "\n",
        "genbank_batch_size = 200  #@param {type:\"integer\"}\n",
        "#@markdown  - Batch size for GenBank metadata API requests. Larger may be faster but can time out (default: 200).\n",
        "genbank_batch_size = int(genbank_batch_size)\n",
        "\n",
        "# -------------------------\n",
        "# Workflow / output options\n",
        "# -------------------------\n",
        "#@markdown ## **Workflow / output options**\n",
        "\n",
        "keep_temp = False  #@param {type:\"boolean\"}\n",
        "#@markdown  - Keep intermediate temporary files.\n",
        "keep_temp = arg_str_to_bool(keep_temp)\n",
        "\n",
        "outfolder = \"results\"  #@param {type:\"string\"}\n",
        "#@markdown  - Output folder path.\n",
        "outfolder = arg_str_to_bool(outfolder)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "6cazzIZfiJDU"
      },
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Optional: Upload a fasta file with your own sequences to add to the analysis\n",
        "  **1) Click on the folder icon on the left.  \n",
        "  2) Upload your file to the Google Colab server by dragging in your file (or use rightclick -> Upload).  \n",
        "  3) Specify the name of your file here:**"
      ],
      "metadata": {
        "id": "aE8ot-AyiPdC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title FASTA file containing additional sequences\n",
        "\n",
        "fasta_file = \"\"  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_fasta_file.fa' or 'my_fasta_file.fasta'.\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "E-lg25zave0H"
      },
      "execution_count": 2,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Metadata\n",
        "\n",
        "#@markdown **Option 1: The metadata is the same for all sequences in your FASTA file**\n",
        "metadata = {'Collection Date': 'YYYY-MM-DD', 'Geo Location': 'South Korea'}  #@param {type:\"raw\"}\n",
        "#@markdown - The 'Collection Date' field is required. Optional: you can add as many additional columns as you wish, e.g. 'Geo Location': 'South Korea'.\n",
        "#@markdown - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "#@markdown **Option 2: Input a CSV file with metadata for each sequence**\n",
        "metadata_csv = \"\"  #@param {type:\"string\"}\n",
        "#@markdown  - Example: 'my_metadata.csv'. This file must include at least an 'Accession' and 'Collection Date' column.\n",
        "#@markdown  - NOTE: Make sure the IDs in the \"Accession\" column match the IDs of the sequences in the provided FASTA file\n",
        "#@markdown  - NOTE: Use NCBI column names where applicable (see https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus for example column names)\n",
        "\n",
        "# Convert empty strings to None\n",
        "fasta_file = arg_str_to_bool(fasta_file)\n",
        "metadata_csv = arg_str_to_bool(metadata_csv)"
      ],
      "metadata": {
        "id": "LKgxMFBgiP64",
        "cellView": "form"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Optional: Specify Delphy arguments"
      ],
      "metadata": {
        "id": "zQF5zZ2dzS60"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Delphy options:\n",
        "mutation_rate = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Virus mutation rate (mutations per site per year), e.g. 0.01. If set to `None`, the mutation rate will be estimated from the generated tree.\n",
        "\n",
        "delphy_steps = None  #@param {type:\"raw\"}\n",
        "#@markdown  - Number of steps to run in the Delphy algorithm (default: 500,000 * number of sequences).\n",
        "\n",
        "delphy_samples = 200 #@param {type:\"integer\"}\n",
        "#@markdown  - Number of logging and tree updates (will log every `delphy_steps/delphy_samples` step).\n",
        "\n",
        "delphy_release = \"1.2.2\"  #@param {type:\"string\"}\n",
        "#@markdown  - Delphy version to use (see https://github.com/broadinstitute/delphy/releases).\n",
        "\n",
        "#@title Delphy options:\n",
        "threads = 2  #@param {type:\"integer\"}\n",
        "#@markdown  - Number of threads to use for the MSA (Mafft) and phylogenetic tree (Delphy) generation."
      ],
      "metadata": {
        "cellView": "form",
        "id": "bEM5nlzZzSih"
      },
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Select `Runtime` at the top of this notebook, then click `Run all` and lean back...\n",
        "A completion message will be displayed below once the notebook has been successfully executed.  \n",
        "**üí° Tip: Click on the folder icon üìÅ on the left to view/download the files that are being generated.**\n",
        "  \n",
        "<br>\n",
        "\n",
        "____\n",
        "____\n",
        "____\n",
        "____"
      ],
      "metadata": {
        "id": "UTL55NF2jAj-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "%%time\n",
        "\n",
        "#@title # Generating the phylogenetic tree...\n",
        "\n",
        "from IPython.display import display, HTML\n",
        "from pathlib import Path\n",
        "import subprocess\n",
        "import sys\n",
        "import shutil\n",
        "import glob\n",
        "import pandas as pd\n",
        "import re\n",
        "import os\n",
        "from datetime import datetime\n",
        "\n",
        "# ---------- Small UI helpers ----------\n",
        "def log_message(text):\n",
        "    display(HTML(f\"<h3 style='color:green;margin:6px 0'>{text}</h3>\"))\n",
        "\n",
        "def log_message_sub(text):\n",
        "    display(HTML(f\"<p style='color:black; font-weight:normal; margin:4px 0;'>{text}</p>\"))\n",
        "\n",
        "def log_message_error(text):\n",
        "    display(HTML(f\"<h3 style='color:red;margin:6px 0'>{text}</h3>\"))\n",
        "\n",
        "def run_checked(cmd, **kwargs):\n",
        "    \"\"\"Run subprocess.run with sensible defaults and return CompletedProcess or raise.\"\"\"\n",
        "    default = dict(check=True, text=True, capture_output=True)\n",
        "    default.update(kwargs)\n",
        "    return subprocess.run(cmd, **default)\n",
        "\n",
        "# ---------- Validate / create outfolder ----------\n",
        "outfolder = Path(outfolder) if 'outfolder' in globals() else Path(\"delphy_out\")\n",
        "outfolder.mkdir(parents=True, exist_ok=True)\n",
        "\n",
        "# ---------- 1) Install Python requirements (gget) ----------\n",
        "log_message(\"1/5 Installing Python dependencies (gget)...\")\n",
        "try:\n",
        "    !pip install --upgrade -q gget biopython\n",
        "    import gget\n",
        "    from Bio import SeqIO\n",
        "    log_message_sub(\"Python dependencies installed.\")\n",
        "except Exception as e:\n",
        "    log_message_error(\"Failed to install Python dependencies (gget).\")\n",
        "    print(e)\n",
        "    raise\n",
        "\n",
        "# ---------- 2) Download sequences from NCBI using gget ----------\n",
        "log_message(\"2/5 Downloading viral sequences from NCBI (gget.virus)... This may take a few minutes.\")\n",
        "try:\n",
        "    # The original code used many parameters assumed present in the notebook; call gget.virus with locals()\n",
        "    # Build kwargs only from variables that exist in globals() and are not None.\n",
        "    possible_kwargs = [\n",
        "        \"virus\",\"is_accession\",\"download_all_accessions\",\"host\",\n",
        "        \"nuc_completeness\",\"min_seq_length\",\"max_seq_length\",\n",
        "        \"min_gene_count\",\"max_gene_count\",\"min_mature_peptide_count\",\"max_mature_peptide_count\",\n",
        "        \"min_protein_count\",\"max_protein_count\",\"max_ambiguous_chars\",\"has_proteins\",\"proteins_complete\",\"annotated\",\"refseq_only\",\n",
        "        \"geographic_location\",\"submitter_country\",\"lab_passaged\",\n",
        "        \"min_collection_date\",\"max_collection_date\",\"min_release_date\",\"max_release_date\",\n",
        "        \"lineage\",\"is_sars_cov2\",\"is_alphainfluenza\",\n",
        "        \"genbank_metadata\",\"genbank_batch_size\",\"keep_temp\",\"outfolder\"\n",
        "    ]\n",
        "    gget_kwargs = {k: globals()[k] for k in possible_kwargs if k in globals() and globals()[k] is not None}\n",
        "    # Ensure outfolder is string path for gget\n",
        "    if \"outfolder\" not in gget_kwargs:\n",
        "        gget_kwargs[\"outfolder\"] = str(outfolder)\n",
        "    else:\n",
        "        gget_kwargs[\"outfolder\"] = str(Path(gget_kwargs[\"outfolder\"]))\n",
        "    # Call gget.virus\n",
        "    gget.virus(**gget_kwargs)\n",
        "\n",
        "    # Find the newest fasta and metadata file in outfolder\n",
        "    fasta_candidates = list(outfolder.glob(\"*.fasta\")) + list(outfolder.glob(\"*.fa\"))\n",
        "    if not fasta_candidates:\n",
        "        raise FileNotFoundError(f\"No fasta files were created in {outfolder}\")\n",
        "    ncbi_fasta_file = max(fasta_candidates, key=lambda p: p.stat().st_mtime)\n",
        "\n",
        "    metadata_candidates = list(outfolder.glob(\"*.csv\")) + list(outfolder.glob(\"*.jsonl\")) + list(outfolder.glob(\"*.tsv\"))\n",
        "    if not metadata_candidates:\n",
        "        raise FileNotFoundError(f\"No metadata files were created in {outfolder}\")\n",
        "    ncbi_metadata = max(metadata_candidates, key=lambda p: p.stat().st_mtime)\n",
        "\n",
        "    log_message_sub(f\"Download complete. FASTA: {ncbi_fasta_file.name}  Metadata: {ncbi_metadata.name}\")\n",
        "except Exception as e:\n",
        "    log_message_error(\"Error while downloading sequences from NCBI.\")\n",
        "    print(e)\n",
        "    raise\n",
        "\n",
        "# ---------- 2b) If user provided FASTA/metadata, merge them ----------\n",
        "input_fasta_file = Path(ncbi_fasta_file)\n",
        "metadata_file = Path(ncbi_metadata)\n",
        "\n",
        "if 'fasta_file' in globals() and fasta_file:\n",
        "    log_message_sub(\"Merging user-provided FASTA with NCBI FASTA...\")\n",
        "    try:\n",
        "        combined_fasta_file = outfolder / f\"{'_'.join(str(virus).split())}_sequences_combined.fasta\"\n",
        "        with open(combined_fasta_file, \"wb\") as outfh:\n",
        "            for src in [ncbi_fasta_file, Path(fasta_file)]:\n",
        "                with open(src, \"rb\") as fh:\n",
        "                    shutil.copyfileobj(fh, outfh)\n",
        "        input_fasta_file = combined_fasta_file\n",
        "\n",
        "        # Combine metadata\n",
        "        combined_metadata_file = outfolder / f\"{'_'.join(str(virus).split())}_metadata_combined.csv\"\n",
        "        ncbi_df = pd.read_csv(ncbi_metadata)\n",
        "\n",
        "        if 'metadata_csv' in globals() and metadata_csv:\n",
        "            user_meta_df = pd.read_csv(metadata_csv)\n",
        "            comb_meta_df = pd.concat([ncbi_df, user_meta_df], ignore_index=True, sort=False)\n",
        "            comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "            metadata_file = combined_metadata_file\n",
        "        else:\n",
        "            # If user provided `metadata` mapping and FASTA only\n",
        "            headers = [record.id.split()[0] for record in SeqIO.parse(fasta_file, \"fasta\")]\n",
        "            user_meta_df = pd.DataFrame({\"accession\": headers})\n",
        "            if 'metadata' in globals() and isinstance(metadata, dict):\n",
        "                for k, v in metadata.items():\n",
        "                    # Broadcast scalar values or accept list-like\n",
        "                    if hasattr(v, \"__len__\") and not isinstance(v, str) and len(v) == len(headers):\n",
        "                        user_meta_df[k] = v\n",
        "                    else:\n",
        "                        user_meta_df[k] = [v] * len(headers)\n",
        "            comb_meta_df = pd.concat([ncbi_df, user_meta_df], ignore_index=True, sort=False)\n",
        "            comb_meta_df.to_csv(combined_metadata_file, index=False)\n",
        "            metadata_file = combined_metadata_file\n",
        "\n",
        "        log_message_sub(f\"Merging complete. Combined FASTA: {combined_fasta_file.name}  Combined metadata: {combined_metadata_file.name}\")\n",
        "    except Exception as e:\n",
        "        log_message_error(\"Error while merging user-provided data.\")\n",
        "        print(e)\n",
        "        raise\n",
        "\n",
        "# ---------- 3) Multiple Sequence Alignment (MSA) using MAFFT (Colab-friendly, --parttree for >10k) ----------\n",
        "log_message(\"3/5 Multiple Sequence Alignment (MAFFT). This may take time depending on # sequences and threads.\")\n",
        "\n",
        "aligned_fasta_file = outfolder / f\"{'_'.join(str(virus).split())}_aligned.afa\"\n",
        "\n",
        "def count_fasta_records(path: Path):\n",
        "    return sum(1 for _ in SeqIO.parse(str(path), \"fasta\"))\n",
        "\n",
        "# 1) Check for mafft binary\n",
        "def mafft_is_available():\n",
        "    return bool(shutil.which(\"mafft\"))\n",
        "\n",
        "# 2) If not available, attempt the same .deb download + dpkg approach you had for Colab\n",
        "if not mafft_is_available():\n",
        "    try:\n",
        "        log_message_sub(\"MAFFT not found ‚Äî attempting to download & install MAFFT .deb (Colab-compatible).\")\n",
        "        deb_path = outfolder / \"mafft_7.526-1_amd64.deb\"\n",
        "        # download\n",
        "        subprocess.run([\"wget\", \"-q\", \"-O\", str(deb_path),\n",
        "                        \"https://mafft.cbrc.jp/alignment/software/mafft_7.526-1_amd64.deb\"],\n",
        "                       check=True, text=True)\n",
        "        # install\n",
        "        subprocess.run([\"dpkg\", \"-i\", str(deb_path)], check=True, text=True)\n",
        "        # If dpkg left missing deps, try apt-get -f install\n",
        "        subprocess.run([\"apt-get\", \"-y\", \"install\", \"-f\"], check=True, text=True)\n",
        "        if mafft_is_available():\n",
        "            log_message_sub(\"MAFFT installed successfully.\")\n",
        "        else:\n",
        "            log_message_error(\"MAFFT was not found after .deb install. Please install MAFFT manually in the Colab environment.\")\n",
        "            raise FileNotFoundError(\"mafft binary not found after .deb install\")\n",
        "    except subprocess.CalledProcessError as e:\n",
        "        log_message_error(\"An error occurred while attempting to install MAFFT via .deb.\")\n",
        "        # print stderr/stdout if available\n",
        "        try:\n",
        "            print(e.stderr or e.stdout)\n",
        "        except Exception:\n",
        "            print(e)\n",
        "        raise\n",
        "\n",
        "# 3) Prepare MAFFT command, adding --parttree for large datasets\n",
        "num_seqs = count_fasta_records(input_fasta_file)\n",
        "threads = int(threads) if 'threads' in globals() and threads else 1\n",
        "\n",
        "# Build command; place --parttree right after \"mafft\" (recommended)\n",
        "if num_seqs > 10_000:\n",
        "    log_message_sub(f\"Detected {num_seqs:,} sequences ‚Äî using MAFFT with --parttree optimized for large datasets.\")\n",
        "    mafft_cmd = [\"mafft\", \"--parttree\", \"--auto\", \"--thread\", str(threads), str(input_fasta_file)]\n",
        "else:\n",
        "    log_message_sub(f\"Detected {num_seqs:,} sequences ‚Äî using MAFFT standard --auto mode.\")\n",
        "    mafft_cmd = [\"mafft\", \"--auto\", \"--thread\", str(threads), str(input_fasta_file)]\n",
        "\n",
        "# 4) Run MAFFT and write alignment directly to file (no capture_output conflict)\n",
        "try:\n",
        "    log_message_sub(f\"Running: {' '.join(mafft_cmd[:6])} ... (command truncated in log)\")\n",
        "    with open(aligned_fasta_file, \"w\") as outfh:\n",
        "        # IMPORTANT: do NOT use capture_output=True here; provide stdout=outfh instead\n",
        "        subprocess.run(mafft_cmd, stdout=outfh, stderr=subprocess.PIPE, check=True, text=True)\n",
        "    log_message_sub(f\"MSA complete and saved to: {aligned_fasta_file.name}\")\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(\"Error while running MAFFT for MSA. See stderr below:\")\n",
        "    stderr = e.stderr.decode() if isinstance(e.stderr, (bytes, bytearray)) else e.stderr\n",
        "    print(stderr)\n",
        "    raise\n",
        "except Exception as e:\n",
        "    log_message_error(\"Unexpected error while running MAFFT.\")\n",
        "    print(e)\n",
        "    raise\n",
        "\n",
        "# ---------- 4) Reformat FASTA headers to match Delphy format (accession|YYYY-MM-DD) ----------\n",
        "log_message(\"4/5 Reformatting FASTA headers to 'accession|YYYY-MM-DD' (skipping entries without valid collection dates).\")\n",
        "\n",
        "def extract_and_format_date(date_string: str, accept_uncertain_dates: bool = False):\n",
        "    \"\"\"Return YYYY-MM-DD, YYYY-MM, or YYYY depending on accept_uncertain_dates. Return None if unrecognized.\"\"\"\n",
        "    if pd.isna(date_string):\n",
        "        return None\n",
        "    date_string = str(date_string).strip()\n",
        "    # Normalize separators to '-'\n",
        "    date_string = re.sub(r\"[/.]\", \"-\", date_string)\n",
        "    # Try parsing strict YYYY-MM-DD\n",
        "    full_date_re = re.compile(r\"^\\d{4}-\\d{1,2}-\\d{1,2}$\")\n",
        "    year_month_re = re.compile(r\"^\\d{4}-\\d{1,2}$\")\n",
        "    year_re = re.compile(r\"^\\d{4}$\")\n",
        "    if full_date_re.match(date_string):\n",
        "        return date_string\n",
        "    if year_month_re.match(date_string):\n",
        "        return date_string if accept_uncertain_dates else \"EXCLUDED\"\n",
        "    if year_re.match(date_string):\n",
        "        return date_string if accept_uncertain_dates else \"EXCLUDED\"\n",
        "    # Try to parse more flexible formats (e.g., 'Jan 2 2020' or '2 Jan 2020') using datetime\n",
        "    for fmt in (\"%Y-%m-%d\", \"%d-%b-%Y\", \"%d %b %Y\", \"%b %d %Y\", \"%Y/%m/%d\"):\n",
        "        try:\n",
        "            dt = datetime.strptime(date_string, fmt)\n",
        "            return dt.strftime(\"%Y-%m-%d\")\n",
        "        except Exception:\n",
        "            continue\n",
        "    return None\n",
        "\n",
        "def find_date_columns(df: pd.DataFrame):\n",
        "    \"\"\"Heuristic to find which column likely contains collection date and accession.\"\"\"\n",
        "    date_cols = [c for c in df.columns if re.search(r\"date|collection|collected\", c, flags=re.I)]\n",
        "    acc_cols = [c for c in df.columns if re.search(r\"acc(?:ession)?|id|seq_id\", c, flags=re.I)]\n",
        "    # Fallback sensible names\n",
        "    if not date_cols and \"Collection Date\" in df.columns:\n",
        "        date_cols = [\"Collection Date\"]\n",
        "    if not acc_cols and \"accession\" in df.columns:\n",
        "        acc_cols = [\"accession\"]\n",
        "    # Return first hits or None\n",
        "    return (date_cols[0] if date_cols else None, acc_cols[0] if acc_cols else None)\n",
        "\n",
        "def update_fasta_headers(fasta_in: Path, csv_in: Path, fasta_out: Path, accept_uncertain_dates: bool = False):\n",
        "    \"\"\"Rewrite fasta headers to 'accession|date' using CSV metadata. Exclude entries without valid date.\"\"\"\n",
        "    df = pd.read_csv(csv_in, dtype=str)\n",
        "    date_col, acc_col = find_date_columns(df)\n",
        "    if date_col is None or acc_col is None:\n",
        "        raise ValueError(f\"Could not find accession/date columns in {csv_in}. Found columns: {list(df.columns)}\")\n",
        "    accession_to_date = pd.Series(df[date_col].values, index=df[acc_col]).to_dict()\n",
        "\n",
        "    included = 0\n",
        "    excluded = 0\n",
        "    unrecognized = 0\n",
        "\n",
        "    with open(fasta_in) as fh_in, open(fasta_out, \"w\") as fh_out:\n",
        "        for record in SeqIO.parse(fh_in, \"fasta\"):\n",
        "            original_accession = record.id.split()[0]\n",
        "            date_val = accession_to_date.get(original_accession) or accession_to_date.get(original_accession.split(\"|\")[0])\n",
        "            formatted = extract_and_format_date(date_val, accept_uncertain_dates=accept_uncertain_dates)\n",
        "            if formatted == \"EXCLUDED\":\n",
        "                excluded += 1\n",
        "                continue\n",
        "            if formatted is None:\n",
        "                unrecognized += 1\n",
        "                continue\n",
        "            # Set header and remove description\n",
        "            record.id = f\"{original_accession}|{formatted}\"\n",
        "            record.description = \"\"\n",
        "            SeqIO.write(record, fh_out, \"fasta\")\n",
        "            included += 1\n",
        "\n",
        "    return {\"included\": included, \"excluded\": excluded, \"unrecognized\": unrecognized}\n",
        "\n",
        "aligned_fasta_file_clean = outfolder / f\"{'_'.join(str(virus).split())}_aligned_headers_adjusted.afa\"\n",
        "try:\n",
        "    accept_uncertain_dates = bool(globals().get(\"accept_uncertain_dates\", False))\n",
        "    stats = update_fasta_headers(aligned_fasta_file, metadata_file, aligned_fasta_file_clean, accept_uncertain_dates=accept_uncertain_dates)\n",
        "    log_message_sub(f\"Reformatting complete. Included {stats['included']} sequences; excluded (incomplete dates) {stats['excluded']}; unrecognized {stats['unrecognized']}\")\n",
        "except Exception as e:\n",
        "    log_message_error(\"Error while reformatting FASTA headers.\")\n",
        "    print(e)\n",
        "    raise\n",
        "\n",
        "# ---------- 5) Run Delphy ----------\n",
        "log_message(\"5/5 Preparing and running Delphy. This step can be long depending on steps/samples/sequence count.\")\n",
        "\n",
        "def count_fasta_records(path: Path):\n",
        "    return sum(1 for _ in SeqIO.parse(str(path), \"fasta\"))\n",
        "\n",
        "num_seqs = count_fasta_records(aligned_fasta_file_clean)\n",
        "if num_seqs == 0:\n",
        "    raise RuntimeError(f\"No sequences found in {aligned_fasta_file_clean}\")\n",
        "\n",
        "# sensible defaults for delphy params if not provided\n",
        "delphy_steps = globals().get(\"delphy_steps\")\n",
        "delphy_samples = globals().get(\"delphy_samples\", 100)\n",
        "if delphy_steps is None:\n",
        "    delphy_steps = int(500_000 * max(1, num_seqs))\n",
        "\n",
        "log_every = int(delphy_steps / delphy_samples) if delphy_samples else max(1, delphy_steps // 100)\n",
        "tree_every = log_every\n",
        "delphy_snapshot_every = log_every\n",
        "\n",
        "# Download delphy if not present\n",
        "delphy_bin = (Path.cwd() / \"delphy\").resolve()\n",
        "if not delphy_bin.exists():\n",
        "    delphy_release = globals().get(\"delphy_release\", \"1.2.2\")  # user may override\n",
        "    tarball = outfolder / \"delphy_release.tar.gz\"\n",
        "    try:\n",
        "        download_url = f\"https://github.com/broadinstitute/delphy/releases/download/{delphy_release}/delphy-linux-x86_64-ubuntu22.tar.gz\"\n",
        "        log_message_sub(f\"Downloading Delphy release {delphy_release}.\")\n",
        "        run_checked([\"wget\", \"-q\", \"-O\", str(tarball), download_url])\n",
        "        run_checked([\"tar\", \"-xzf\", str(tarball)], check=True)\n",
        "        # Try to find delphy binary in extracted files\n",
        "        found = list(Path(\".\").glob(\"**/delphy\"))\n",
        "        if found:\n",
        "            src = found[0].resolve()\n",
        "            dst = delphy_bin.resolve()\n",
        "\n",
        "            # Make sure the found binary is executable\n",
        "            src.chmod(src.stat().st_mode | 0o111)\n",
        "\n",
        "            # Only copy if it's not already the same file\n",
        "            if src != dst:\n",
        "                shutil.copy(src, dst)\n",
        "                dst.chmod(dst.stat().st_mode | 0o111)\n",
        "                log_message_sub(\"Delphy binary obtained.\")\n",
        "            else:\n",
        "                log_message_sub(\"Delphy binary already present in working directory; skipping copy.\")\n",
        "\n",
        "    except Exception as e:\n",
        "        log_message_error(\"Failed to download/extract Delphy. Please ensure `delphy` binary is available in the working directory.\")\n",
        "        print(e)\n",
        "        raise\n",
        "\n",
        "# Build Delphy command\n",
        "beast_log_out = outfolder / f\"{'_'.join(str(virus).split())}_delphy_beast.log\"\n",
        "delphy_beast_tree_out = outfolder / f\"{'_'.join(str(virus).split())}_delphy_beast.trees\"\n",
        "dphy_out = outfolder / f\"{'_'.join(str(virus).split())}_delphy_out.dphy\"\n",
        "\n",
        "mutation_rate = globals().get(\"mutation_rate\", None)\n",
        "threads = int(globals().get(\"threads\", 1))\n",
        "cmd = [\n",
        "    str(delphy_bin),\n",
        "    \"--v0-log-every\", str(log_every),\n",
        "    \"--v0-tree-every\", str(tree_every),\n",
        "    \"--v0-delphy-snapshot-every\", str(delphy_snapshot_every),\n",
        "    \"--v0-threads\", str(threads),\n",
        "    \"--v0-steps\", str(delphy_steps),\n",
        "    \"--v0-in-fasta\", str(aligned_fasta_file_clean),\n",
        "    \"--v0-out-log-file\", str(beast_log_out),\n",
        "    \"--v0-out-trees-file\", str(delphy_beast_tree_out),\n",
        "    \"--v0-out-delphy-file\", str(dphy_out),\n",
        "]\n",
        "if mutation_rate:\n",
        "    cmd = [str(delphy_bin), \"--v0-fix-mutation-rate\", \"--v0-init-mutation-rate\", str(mutation_rate)] + cmd[1:]\n",
        "\n",
        "log_message_sub(f\"Running Delphy with command: {' '.join(cmd[:8])} ... (truncated)\")\n",
        "\n",
        "try:\n",
        "    # Run delphy and stream output to notebook\n",
        "    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True, bufsize=1)\n",
        "    for line in proc.stdout:\n",
        "        # print lines incrementally so the notebook shows progress\n",
        "        print(line, end=\"\")\n",
        "    ret = proc.wait()\n",
        "    if ret != 0:\n",
        "        raise subprocess.CalledProcessError(ret, cmd)\n",
        "\n",
        "    # Success\n",
        "    outfolder_display = str(outfolder) if outfolder not in [None, \"\", \"None\"] else None\n",
        "\n",
        "    if outfolder_display:\n",
        "        location_text = f\"Files generated in this notebook are in the left-hand file browser and in the <code>{outfolder_display}/</code> folder.\"\n",
        "    else:\n",
        "        location_text = \"Files generated in this notebook are in the left-hand file browser.\"\n",
        "\n",
        "    display(HTML(f\"\"\"\n",
        "    <h2 style='color:green'>All done! üéâ</h2>\n",
        "    <p>{location_text}</p>\n",
        "    <ul>\n",
        "      <li>Aligned FASTA: <code>{aligned_fasta_file_clean.name}</code></li>\n",
        "      <li>Delphy output (.dphy): <code>{dphy_out.name}</code></li>\n",
        "      <li>Delphy log (.log): <code>{beast_log_out.name}</code></li>\n",
        "      <li>Delphy trees (.trees): <code>{delphy_beast_tree_out.name}</code></li>\n",
        "    </ul>\n",
        "    <p><strong>We recommend downloading the <code>.afa</code>, <code>.dphy</code>, and <code>metadata.csv</code> files.\n",
        "    To visualize your Delphy output, upload the <code>.dphy</code> (and optionally <code>metadata.csv</code>) file(s) to\n",
        "    <a href='https://delphy.fathom.info/' target='_blank'>https://delphy.fathom.info/</a>.</strong></p>\n",
        "    \"\"\"))\n",
        "\n",
        "except subprocess.CalledProcessError as e:\n",
        "    log_message_error(\"An error occurred while running Delphy. See captured output above for details.\")\n",
        "    print(e)\n",
        "    raise\n",
        "except Exception as e:\n",
        "    log_message_error(\"Unexpected error while running Delphy.\")\n",
        "    print(e)\n",
        "    raise\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "HuVr7cb0x5VF",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "187ce5e2-094c-4fa4-c62b-0c59c5f010ce"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>1/5 Installing Python dependencies (gget)...</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Python dependencies installed.</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>2/5 Downloading viral sequences from NCBI (gget.virus)... This may take a few minutes.</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "INFO:gget.utils:Starting virus data retrieval process...\n",
            "INFO:gget.utils:Query parameters: virus='Zaire ebolavirus', is_accession=False, outfolder='results'\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 1: VALIDATING INPUT ARGUMENTS AND OUTPUT DIRECTORY SETUP...\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Input validation completed successfully\n",
            "INFO:gget.utils:Using specified output folder: results\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 2: CHECKING FOR SARS-CoV-2 AND INFLUENZA A QUERIES TO APPLY OPTIMIZED CACHED PATHWAY\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils: Skipping this step. No SARS-CoV-2 query detected.\n",
            "INFO:gget.utils: Skipping this step. No Alphainfluenza query detected.\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 3: Fetching virus metadata from NCBI API\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Streaming API metadata to temporary file: results/tmp_20260213_014744_919fd7/gget_metadata_20260213_014744_919fd7.jsonl\n",
            "INFO:gget.utils:Successfully retrieved 3279 virus records from NCBI API across 4 pages\n",
            "INFO:gget.utils:Temporary metadata file size: 10.12 MB\n",
            "INFO:gget.utils:Processed 3279 metadata records, skipped 0 records without accessions\n",
            "INFO:gget.utils:‚úÖ Saved API metadata JSONL: results/Zaire_ebolavirus_api_metadata.jsonl\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 4: Applying metadata-only filters\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Starting metadata-only filtering process...\n",
            "INFO:gget.utils:Processing 3279 metadata records...\n",
            "INFO:gget.utils:‚úÖ Metadata-only filtering process complete.\n",
            "INFO:gget.utils:=================================\n",
            "INFO:gget.utils:Metadata-only filtering complete:\n",
            "INFO:gget.utils:  Total metadata records: 3279\n",
            "INFO:gget.utils:  Records passing filters: 277\n",
            "INFO:gget.utils:Filter statistics (records excluded):\n",
            "INFO:gget.utils:  collection_date: 3002 records\n",
            "INFO:gget.utils:‚úÖ Saved filtered metadata JSONL: results/Zaire_ebolavirus_metadata.jsonl\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 5: Downloading sequences for filtered accessions\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Downloading sequences for 277 accessions using E-utilities API\n",
            "INFO:gget.utils:Large request detected (277 accessions). Using batch processing with batch size 200\n",
            "INFO:gget.utils:Downloading 277 accessions in 2 batches of size 200\n",
            "Downloading batches:   0%|          | 0/2 [00:01<?, ?batch/s]"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "‚úì Batch 1: Downloaded 200 sequences (3.68 MB)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Downloading batches: 100%|‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà| 2/2 [00:03<00:00,  1.65s/batch]\n",
            "INFO:gget.utils:Successfully saved 277 sequences to: results/tmp_20260213_014744_919fd7/virus_sequences_20260213_014744_919fd7.fasta (5.10 MB)\n",
            "INFO:gget.utils:Downloaded FASTA file: results/tmp_20260213_014744_919fd7/virus_sequences_20260213_014744_919fd7.fasta (5.10 MB)\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 6: Applying sequence-dependent filters and saving results\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:No sequence-dependent filters specified, skipping this step.\n",
            "INFO:gget.utils:All 277 downloaded sequences will be saved\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 7: Saving final output files\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Saving 277 filtered sequences and their metadata...\n",
            "INFO:gget.utils:‚úÖ FASTA file saved: results/Zaire_ebolavirus_sequences.fasta (5.10 MB)\n",
            "INFO:gget.utils:‚úÖ JSONL metadata file saved: results/Zaire_ebolavirus_metadata.jsonl (0.67 MB)\n",
            "INFO:gget.utils:Preparing metadata for CSV output...\n",
            "INFO:gget.utils:Processing metadata records...\n",
            "INFO:gget.utils:Creating DataFrame with 277 rows and 39 columns\n",
            "INFO:gget.utils:Successfully saved metadata CSV to: results/Zaire_ebolavirus_metadata.csv\n",
            "INFO:gget.utils:‚úÖ CSV metadata file saved: results/Zaire_ebolavirus_metadata.csv (0.56 MB)\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:STEP 8: Fetching detailed GenBank metadata\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Skipping this step since GenBank metadata retrieval was not requested.\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:PROCESS COMPLETED SUCCESSFULLY\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:Results summary:\n",
            "INFO:gget.utils:  Total sequences (API metadata): 3279\n",
            "INFO:gget.utils:  After metadata-only filtering: 277\n",
            "INFO:gget.utils:  After all filters (final): 277\n",
            "INFO:gget.utils:\n",
            "INFO:gget.utils:Output files saved to: results\n",
            "INFO:gget.utils:  üìÑ Sequences (FASTA): Zaire_ebolavirus_sequences.fasta\n",
            "INFO:gget.utils:  üìä Metadata (CSV):    Zaire_ebolavirus_metadata.csv\n",
            "INFO:gget.utils:  üîß Metadata (JSONL):  Zaire_ebolavirus_metadata.jsonl\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:============================================================\n",
            "INFO:gget.utils:‚úÖ Command summary saved: results/command_summary.txt\n",
            "INFO:gget.utils:NCBI virus data retrieval process completed.\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "‚úì Batch 2: Downloaded 77 sequences (1.42 MB)\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Download complete. FASTA: Zaire_ebolavirus_sequences.fasta  Metadata: Zaire_ebolavirus_metadata.csv</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>3/5 Multiple Sequence Alignment (MAFFT). This may take time depending on # sequences and threads.</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Detected 277 sequences ‚Äî using MAFFT standard --auto mode.</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Running: mafft --auto --thread 2 results/Zaire_ebolavirus_sequences.fasta ... (command truncated in log)</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>MSA complete and saved to: Zaire_ebolavirus_aligned.afa</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>4/5 Reformatting FASTA headers to 'accession|YYYY-MM-DD' (skipping entries without valid collection dates).</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Reformatting complete. Included 277 sequences; excluded (incomplete dates) 0; unrecognized 0</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<h3 style='color:green;margin:6px 0'>5/5 Preparing and running Delphy. This step can be long depending on steps/samples/sequence count.</h3>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "<p style='color:black; font-weight:normal; margin:4px 0;'>Running Delphy with command: /content/delphy --v0-log-every 692500 --v0-tree-every 692500 --v0-delphy-snapshot-every 692500 --v0-threads ... (truncated)</p>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "# Seed: 1025141311\n",
            "Reading fasta file results/Zaire_ebolavirus_aligned_headers_adjusted.afa\n",
            "- read 1 sequences so far (19324 of 5346377 bytes = 0.4%)\n",
            "- read 2 sequences so far (38625 of 5346377 bytes = 0.7%)\n",
            "- read 3 sequences so far (57926 of 5346377 bytes = 1.1%)\n",
            "- read 4 sequences so far (77227 of 5346377 bytes = 1.4%)\n",
            "- read 5 sequences so far (96528 of 5346377 bytes = 1.8%)\n",
            "- read 6 sequences so far (115829 of 5346377 bytes = 2.2%)\n",
            "- read 7 sequences so far (135130 of 5346377 bytes = 2.5%)\n",
            "- read 8 sequences so far (154431 of 5346377 bytes = 2.9%)\n",
            "- read 9 sequences so far (173732 of 5346377 bytes = 3.2%)\n",
            "- read 10 sequences so far (193033 of 5346377 bytes = 3.6%)\n",
            "- read 11 sequences so far (212334 of 5346377 bytes = 4.0%)\n",
            "- read 12 sequences so far (231635 of 5346377 bytes = 4.3%)\n",
            "- read 13 sequences so far (250936 of 5346377 bytes = 4.7%)\n",
            "- read 14 sequences so far (270237 of 5346377 bytes = 5.1%)\n",
            "- read 15 sequences so far (289538 of 5346377 bytes = 5.4%)\n",
            "- read 16 sequences so far (308839 of 5346377 bytes = 5.8%)\n",
            "- read 17 sequences so far (328140 of 5346377 bytes = 6.1%)\n",
            "- read 18 sequences so far (347441 of 5346377 bytes = 6.5%)\n",
            "- read 19 sequences so far (366742 of 5346377 bytes = 6.9%)\n",
            "- read 20 sequences so far (386043 of 5346377 bytes = 7.2%)\n",
            "- read 21 sequences so far (405344 of 5346377 bytes = 7.6%)\n",
            "- read 22 sequences so far (424645 of 5346377 bytes = 7.9%)\n",
            "- read 23 sequences so far (443946 of 5346377 bytes = 8.3%)\n",
            "- read 24 sequences so far (463247 of 5346377 bytes = 8.7%)\n",
            "- read 25 sequences so far (482548 of 5346377 bytes = 9.0%)\n",
            "- read 26 sequences so far (501849 of 5346377 bytes = 9.4%)\n",
            "- read 27 sequences so far (521150 of 5346377 bytes = 9.7%)\n",
            "- read 28 sequences so far (540451 of 5346377 bytes = 10.1%)\n",
            "- read 29 sequences so far (559752 of 5346377 bytes = 10.5%)\n",
            "- read 30 sequences so far (579053 of 5346377 bytes = 10.8%)\n",
            "- read 31 sequences so far (598354 of 5346377 bytes = 11.2%)\n",
            "- read 32 sequences so far (617655 of 5346377 bytes = 11.6%)\n",
            "- read 33 sequences so far (636956 of 5346377 bytes = 11.9%)\n",
            "- read 34 sequences so far (656257 of 5346377 bytes = 12.3%)\n",
            "- read 35 sequences so far (675558 of 5346377 bytes = 12.6%)\n",
            "- read 36 sequences so far (694859 of 5346377 bytes = 13.0%)\n",
            "- read 37 sequences so far (714160 of 5346377 bytes = 13.4%)\n",
            "- read 38 sequences so far (733461 of 5346377 bytes = 13.7%)\n",
            "- read 39 sequences so far (752762 of 5346377 bytes = 14.1%)\n",
            "- read 40 sequences so far (772063 of 5346377 bytes = 14.4%)\n",
            "- read 41 sequences so far (791364 of 5346377 bytes = 14.8%)\n",
            "- read 42 sequences so far (810665 of 5346377 bytes = 15.2%)\n",
            "- read 43 sequences so far (829966 of 5346377 bytes = 15.5%)\n",
            "- read 44 sequences so far (849267 of 5346377 bytes = 15.9%)\n",
            "- read 45 sequences so far (868568 of 5346377 bytes = 16.2%)\n",
            "- read 46 sequences so far (887869 of 5346377 bytes = 16.6%)\n",
            "- read 47 sequences so far (907170 of 5346377 bytes = 17.0%)\n",
            "- read 48 sequences so far (926471 of 5346377 bytes = 17.3%)\n",
            "- read 49 sequences so far (945772 of 5346377 bytes = 17.7%)\n",
            "- read 50 sequences so far (965073 of 5346377 bytes = 18.1%)\n",
            "- read 51 sequences so far (984374 of 5346377 bytes = 18.4%)\n",
            "- read 52 sequences so far (1003675 of 5346377 bytes = 18.8%)\n",
            "- read 53 sequences so far (1022976 of 5346377 bytes = 19.1%)\n",
            "- read 54 sequences so far (1042277 of 5346377 bytes = 19.5%)\n",
            "- read 55 sequences so far (1061578 of 5346377 bytes = 19.9%)\n",
            "- read 56 sequences so far (1080879 of 5346377 bytes = 20.2%)\n",
            "- read 57 sequences so far (1100180 of 5346377 bytes = 20.6%)\n",
            "- read 58 sequences so far (1119481 of 5346377 bytes = 20.9%)\n",
            "- read 59 sequences so far (1138782 of 5346377 bytes = 21.3%)\n",
            "- read 60 sequences so far (1158083 of 5346377 bytes = 21.7%)\n",
            "- read 61 sequences so far (1177384 of 5346377 bytes = 22.0%)\n",
            "- read 62 sequences so far (1196685 of 5346377 bytes = 22.4%)\n",
            "- read 63 sequences so far (1215986 of 5346377 bytes = 22.7%)\n",
            "- read 64 sequences so far (1235287 of 5346377 bytes = 23.1%)\n",
            "- read 65 sequences so far (1254588 of 5346377 bytes = 23.5%)\n",
            "- read 66 sequences so far (1273889 of 5346377 bytes = 23.8%)\n",
            "- read 67 sequences so far (1293190 of 5346377 bytes = 24.2%)\n",
            "- read 68 sequences so far (1312491 of 5346377 bytes = 24.5%)\n",
            "- read 69 sequences so far (1331792 of 5346377 bytes = 24.9%)\n",
            "- read 70 sequences so far (1351093 of 5346377 bytes = 25.3%)\n",
            "- read 71 sequences so far (1370394 of 5346377 bytes = 25.6%)\n",
            "- read 72 sequences so far (1389695 of 5346377 bytes = 26.0%)\n",
            "- read 73 sequences so far (1408996 of 5346377 bytes = 26.4%)\n",
            "- read 74 sequences so far (1428297 of 5346377 bytes = 26.7%)\n",
            "- read 75 sequences so far (1447598 of 5346377 bytes = 27.1%)\n",
            "- read 76 sequences so far (1466899 of 5346377 bytes = 27.4%)\n",
            "- read 77 sequences so far (1486200 of 5346377 bytes = 27.8%)\n",
            "- read 78 sequences so far (1505501 of 5346377 bytes = 28.2%)\n",
            "- read 79 sequences so far (1524802 of 5346377 bytes = 28.5%)\n",
            "- read 80 sequences so far (1544103 of 5346377 bytes = 28.9%)\n",
            "- read 81 sequences so far (1563404 of 5346377 bytes = 29.2%)\n",
            "- read 82 sequences so far (1582705 of 5346377 bytes = 29.6%)\n",
            "- read 83 sequences so far (1602006 of 5346377 bytes = 30.0%)\n",
            "- read 84 sequences so far (1621307 of 5346377 bytes = 30.3%)\n",
            "- read 85 sequences so far (1640608 of 5346377 bytes = 30.7%)\n",
            "- read 86 sequences so far (1659909 of 5346377 bytes = 31.0%)\n",
            "- read 87 sequences so far (1679210 of 5346377 bytes = 31.4%)\n",
            "- read 88 sequences so far (1698511 of 5346377 bytes = 31.8%)\n",
            "- read 89 sequences so far (1717812 of 5346377 bytes = 32.1%)\n",
            "- read 90 sequences so far (1737113 of 5346377 bytes = 32.5%)\n",
            "- read 91 sequences so far (1756414 of 5346377 bytes = 32.9%)\n",
            "- read 92 sequences so far (1775715 of 5346377 bytes = 33.2%)\n",
            "- read 93 sequences so far (1795016 of 5346377 bytes = 33.6%)\n",
            "- read 94 sequences so far (1814317 of 5346377 bytes = 33.9%)\n",
            "- read 95 sequences so far (1833618 of 5346377 bytes = 34.3%)\n",
            "- read 96 sequences so far (1852919 of 5346377 bytes = 34.7%)\n",
            "- read 97 sequences so far (1872220 of 5346377 bytes = 35.0%)\n",
            "- read 98 sequences so far (1891521 of 5346377 bytes = 35.4%)\n",
            "- read 99 sequences so far (1910822 of 5346377 bytes = 35.7%)\n",
            "- read 100 sequences so far (1930123 of 5346377 bytes = 36.1%)\n",
            "- read 101 sequences so far (1949424 of 5346377 bytes = 36.5%)\n",
            "- read 102 sequences so far (1968725 of 5346377 bytes = 36.8%)\n",
            "- read 103 sequences so far (1988026 of 5346377 bytes = 37.2%)\n",
            "- read 104 sequences so far (2007327 of 5346377 bytes = 37.5%)\n",
            "- read 105 sequences so far (2026628 of 5346377 bytes = 37.9%)\n",
            "- read 106 sequences so far (2045929 of 5346377 bytes = 38.3%)\n",
            "- read 107 sequences so far (2065230 of 5346377 bytes = 38.6%)\n",
            "- read 108 sequences so far (2084531 of 5346377 bytes = 39.0%)\n",
            "- read 109 sequences so far (2103832 of 5346377 bytes = 39.4%)\n",
            "- read 110 sequences so far (2123133 of 5346377 bytes = 39.7%)\n",
            "- read 111 sequences so far (2142434 of 5346377 bytes = 40.1%)\n",
            "- read 112 sequences so far (2161735 of 5346377 bytes = 40.4%)\n",
            "- read 113 sequences so far (2181036 of 5346377 bytes = 40.8%)\n",
            "- read 114 sequences so far (2200337 of 5346377 bytes = 41.2%)\n",
            "- read 115 sequences so far (2219638 of 5346377 bytes = 41.5%)\n",
            "- read 116 sequences so far (2238939 of 5346377 bytes = 41.9%)\n",
            "- read 117 sequences so far (2258240 of 5346377 bytes = 42.2%)\n",
            "- read 118 sequences so far (2277541 of 5346377 bytes = 42.6%)\n",
            "- read 119 sequences so far (2296842 of 5346377 bytes = 43.0%)\n",
            "- read 120 sequences so far (2316143 of 5346377 bytes = 43.3%)\n",
            "- read 121 sequences so far (2335444 of 5346377 bytes = 43.7%)\n",
            "- read 122 sequences so far (2354745 of 5346377 bytes = 44.0%)\n",
            "- read 123 sequences so far (2374046 of 5346377 bytes = 44.4%)\n",
            "- read 124 sequences so far (2393347 of 5346377 bytes = 44.8%)\n",
            "- read 125 sequences so far (2412648 of 5346377 bytes = 45.1%)\n",
            "- read 126 sequences so far (2431949 of 5346377 bytes = 45.5%)\n",
            "- read 127 sequences so far (2451250 of 5346377 bytes = 45.8%)\n",
            "- read 128 sequences so far (2470551 of 5346377 bytes = 46.2%)\n",
            "- read 129 sequences so far (2489852 of 5346377 bytes = 46.6%)\n",
            "- read 130 sequences so far (2509153 of 5346377 bytes = 46.9%)\n",
            "- read 131 sequences so far (2528454 of 5346377 bytes = 47.3%)\n",
            "- read 132 sequences so far (2547755 of 5346377 bytes = 47.7%)\n",
            "- read 133 sequences so far (2567056 of 5346377 bytes = 48.0%)\n",
            "- read 134 sequences so far (2586357 of 5346377 bytes = 48.4%)\n",
            "- read 135 sequences so far (2605658 of 5346377 bytes = 48.7%)\n",
            "- read 136 sequences so far (2624959 of 5346377 bytes = 49.1%)\n",
            "- read 137 sequences so far (2644260 of 5346377 bytes = 49.5%)\n",
            "- read 138 sequences so far (2663561 of 5346377 bytes = 49.8%)\n",
            "- read 139 sequences so far (2682862 of 5346377 bytes = 50.2%)\n",
            "- read 140 sequences so far (2702163 of 5346377 bytes = 50.5%)\n",
            "- read 141 sequences so far (2721464 of 5346377 bytes = 50.9%)\n",
            "- read 142 sequences so far (2740765 of 5346377 bytes = 51.3%)\n",
            "- read 143 sequences so far (2760066 of 5346377 bytes = 51.6%)\n",
            "- read 144 sequences so far (2779367 of 5346377 bytes = 52.0%)\n",
            "- read 145 sequences so far (2798668 of 5346377 bytes = 52.3%)\n",
            "- read 146 sequences so far (2817969 of 5346377 bytes = 52.7%)\n",
            "- read 147 sequences so far (2837270 of 5346377 bytes = 53.1%)\n",
            "- read 148 sequences so far (2856571 of 5346377 bytes = 53.4%)\n",
            "- read 149 sequences so far (2875872 of 5346377 bytes = 53.8%)\n",
            "- read 150 sequences so far (2895173 of 5346377 bytes = 54.2%)\n",
            "- read 151 sequences so far (2914474 of 5346377 bytes = 54.5%)\n",
            "- read 152 sequences so far (2933775 of 5346377 bytes = 54.9%)\n",
            "- read 153 sequences so far (2953076 of 5346377 bytes = 55.2%)\n",
            "- read 154 sequences so far (2972377 of 5346377 bytes = 55.6%)\n",
            "- read 155 sequences so far (2991678 of 5346377 bytes = 56.0%)\n",
            "- read 156 sequences so far (3010979 of 5346377 bytes = 56.3%)\n",
            "- read 157 sequences so far (3030280 of 5346377 bytes = 56.7%)\n",
            "- read 158 sequences so far (3049581 of 5346377 bytes = 57.0%)\n",
            "- read 159 sequences so far (3068882 of 5346377 bytes = 57.4%)\n",
            "- read 160 sequences so far (3088183 of 5346377 bytes = 57.8%)\n",
            "- read 161 sequences so far (3107484 of 5346377 bytes = 58.1%)\n",
            "- read 162 sequences so far (3126785 of 5346377 bytes = 58.5%)\n",
            "- read 163 sequences so far (3146086 of 5346377 bytes = 58.8%)\n",
            "- read 164 sequences so far (3165387 of 5346377 bytes = 59.2%)\n",
            "- read 165 sequences so far (3184688 of 5346377 bytes = 59.6%)\n",
            "- read 166 sequences so far (3203989 of 5346377 bytes = 59.9%)\n",
            "- read 167 sequences so far (3223290 of 5346377 bytes = 60.3%)\n",
            "- read 168 sequences so far (3242591 of 5346377 bytes = 60.7%)\n",
            "- read 169 sequences so far (3261892 of 5346377 bytes = 61.0%)\n",
            "- read 170 sequences so far (3281193 of 5346377 bytes = 61.4%)\n",
            "- read 171 sequences so far (3300494 of 5346377 bytes = 61.7%)\n",
            "- read 172 sequences so far (3319795 of 5346377 bytes = 62.1%)\n",
            "- read 173 sequences so far (3339096 of 5346377 bytes = 62.5%)\n",
            "- read 174 sequences so far (3358397 of 5346377 bytes = 62.8%)\n",
            "- read 175 sequences so far (3377698 of 5346377 bytes = 63.2%)\n",
            "- read 176 sequences so far (3396999 of 5346377 bytes = 63.5%)\n",
            "- read 177 sequences so far (3416300 of 5346377 bytes = 63.9%)\n",
            "- read 178 sequences so far (3435601 of 5346377 bytes = 64.3%)\n",
            "- read 179 sequences so far (3454902 of 5346377 bytes = 64.6%)\n",
            "- read 180 sequences so far (3474203 of 5346377 bytes = 65.0%)\n",
            "- read 181 sequences so far (3493504 of 5346377 bytes = 65.3%)\n",
            "- read 182 sequences so far (3512805 of 5346377 bytes = 65.7%)\n",
            "- read 183 sequences so far (3532106 of 5346377 bytes = 66.1%)\n",
            "- read 184 sequences so far (3551407 of 5346377 bytes = 66.4%)\n",
            "- read 185 sequences so far (3570708 of 5346377 bytes = 66.8%)\n",
            "- read 186 sequences so far (3590009 of 5346377 bytes = 67.1%)\n",
            "- read 187 sequences so far (3609310 of 5346377 bytes = 67.5%)\n",
            "- read 188 sequences so far (3628611 of 5346377 bytes = 67.9%)\n",
            "- read 189 sequences so far (3647912 of 5346377 bytes = 68.2%)\n",
            "- read 190 sequences so far (3667213 of 5346377 bytes = 68.6%)\n",
            "- read 191 sequences so far (3686514 of 5346377 bytes = 69.0%)\n",
            "- read 192 sequences so far (3705815 of 5346377 bytes = 69.3%)\n",
            "- read 193 sequences so far (3725116 of 5346377 bytes = 69.7%)\n",
            "- read 194 sequences so far (3744417 of 5346377 bytes = 70.0%)\n",
            "- read 195 sequences so far (3763718 of 5346377 bytes = 70.4%)\n",
            "- read 196 sequences so far (3783019 of 5346377 bytes = 70.8%)\n",
            "- read 197 sequences so far (3802320 of 5346377 bytes = 71.1%)\n",
            "- read 198 sequences so far (3821621 of 5346377 bytes = 71.5%)\n",
            "- read 199 sequences so far (3840922 of 5346377 bytes = 71.8%)\n",
            "- read 200 sequences so far (3860223 of 5346377 bytes = 72.2%)\n",
            "- read 201 sequences so far (3879524 of 5346377 bytes = 72.6%)\n",
            "- read 202 sequences so far (3898825 of 5346377 bytes = 72.9%)\n",
            "- read 203 sequences so far (3918126 of 5346377 bytes = 73.3%)\n",
            "- read 204 sequences so far (3937427 of 5346377 bytes = 73.6%)\n",
            "- read 205 sequences so far (3956728 of 5346377 bytes = 74.0%)\n",
            "- read 206 sequences so far (3976029 of 5346377 bytes = 74.4%)\n",
            "- read 207 sequences so far (3995330 of 5346377 bytes = 74.7%)\n",
            "- read 208 sequences so far (4014631 of 5346377 bytes = 75.1%)\n",
            "- read 209 sequences so far (4033932 of 5346377 bytes = 75.5%)\n",
            "- read 210 sequences so far (4053233 of 5346377 bytes = 75.8%)\n",
            "- read 211 sequences so far (4072534 of 5346377 bytes = 76.2%)\n",
            "- read 212 sequences so far (4091835 of 5346377 bytes = 76.5%)\n",
            "- read 213 sequences so far (4111136 of 5346377 bytes = 76.9%)\n",
            "- read 214 sequences so far (4130437 of 5346377 bytes = 77.3%)\n",
            "- read 215 sequences so far (4149738 of 5346377 bytes = 77.6%)\n",
            "- read 216 sequences so far (4169039 of 5346377 bytes = 78.0%)\n",
            "- read 217 sequences so far (4188340 of 5346377 bytes = 78.3%)\n",
            "- read 218 sequences so far (4207641 of 5346377 bytes = 78.7%)\n",
            "- read 219 sequences so far (4226942 of 5346377 bytes = 79.1%)\n",
            "- read 220 sequences so far (4246243 of 5346377 bytes = 79.4%)\n",
            "- read 221 sequences so far (4265544 of 5346377 bytes = 79.8%)\n",
            "- read 222 sequences so far (4284845 of 5346377 bytes = 80.1%)\n",
            "- read 223 sequences so far (4304146 of 5346377 bytes = 80.5%)\n",
            "- read 224 sequences so far (4323447 of 5346377 bytes = 80.9%)\n",
            "- read 225 sequences so far (4342748 of 5346377 bytes = 81.2%)\n",
            "- read 226 sequences so far (4362049 of 5346377 bytes = 81.6%)\n",
            "- read 227 sequences so far (4381350 of 5346377 bytes = 81.9%)\n",
            "- read 228 sequences so far (4400651 of 5346377 bytes = 82.3%)\n",
            "- read 229 sequences so far (4419952 of 5346377 bytes = 82.7%)\n",
            "- read 230 sequences so far (4439253 of 5346377 bytes = 83.0%)\n",
            "- read 231 sequences so far (4458554 of 5346377 bytes = 83.4%)\n",
            "- read 232 sequences so far (4477855 of 5346377 bytes = 83.8%)\n",
            "- read 233 sequences so far (4497156 of 5346377 bytes = 84.1%)\n",
            "- read 234 sequences so far (4516457 of 5346377 bytes = 84.5%)\n",
            "- read 235 sequences so far (4535758 of 5346377 bytes = 84.8%)\n",
            "- read 236 sequences so far (4555059 of 5346377 bytes = 85.2%)\n",
            "- read 237 sequences so far (4574360 of 5346377 bytes = 85.6%)\n",
            "- read 238 sequences so far (4593661 of 5346377 bytes = 85.9%)\n",
            "- read 239 sequences so far (4612962 of 5346377 bytes = 86.3%)\n",
            "- read 240 sequences so far (4632263 of 5346377 bytes = 86.6%)\n",
            "- read 241 sequences so far (4651564 of 5346377 bytes = 87.0%)\n",
            "- read 242 sequences so far (4670865 of 5346377 bytes = 87.4%)\n",
            "- read 243 sequences so far (4690166 of 5346377 bytes = 87.7%)\n",
            "- read 244 sequences so far (4709467 of 5346377 bytes = 88.1%)\n",
            "- read 245 sequences so far (4728768 of 5346377 bytes = 88.4%)\n",
            "- read 246 sequences so far (4748069 of 5346377 bytes = 88.8%)\n",
            "- read 247 sequences so far (4767370 of 5346377 bytes = 89.2%)\n",
            "- read 248 sequences so far (4786671 of 5346377 bytes = 89.5%)\n",
            "- read 249 sequences so far (4805972 of 5346377 bytes = 89.9%)\n",
            "- read 250 sequences so far (4825273 of 5346377 bytes = 90.3%)\n",
            "- read 251 sequences so far (4844574 of 5346377 bytes = 90.6%)\n",
            "- read 252 sequences so far (4863875 of 5346377 bytes = 91.0%)\n",
            "- read 253 sequences so far (4883176 of 5346377 bytes = 91.3%)\n",
            "- read 254 sequences so far (4902477 of 5346377 bytes = 91.7%)\n",
            "- read 255 sequences so far (4921778 of 5346377 bytes = 92.1%)\n",
            "- read 256 sequences so far (4941079 of 5346377 bytes = 92.4%)\n",
            "- read 257 sequences so far (4960380 of 5346377 bytes = 92.8%)\n",
            "- read 258 sequences so far (4979681 of 5346377 bytes = 93.1%)\n",
            "- read 259 sequences so far (4998982 of 5346377 bytes = 93.5%)\n",
            "- read 260 sequences so far (5018283 of 5346377 bytes = 93.9%)\n",
            "- read 261 sequences so far (5037584 of 5346377 bytes = 94.2%)\n",
            "- read 262 sequences so far (5056885 of 5346377 bytes = 94.6%)\n",
            "- read 263 sequences so far (5076186 of 5346377 bytes = 94.9%)\n",
            "- read 264 sequences so far (5095487 of 5346377 bytes = 95.3%)\n",
            "- read 265 sequences so far (5114788 of 5346377 bytes = 95.7%)\n",
            "- read 266 sequences so far (5134089 of 5346377 bytes = 96.0%)\n",
            "- read 267 sequences so far (5153390 of 5346377 bytes = 96.4%)\n",
            "- read 268 sequences so far (5172691 of 5346377 bytes = 96.8%)\n",
            "- read 269 sequences so far (5191992 of 5346377 bytes = 97.1%)\n",
            "- read 270 sequences so far (5211293 of 5346377 bytes = 97.5%)\n",
            "- read 271 sequences so far (5230594 of 5346377 bytes = 97.8%)\n",
            "- read 272 sequences so far (5249895 of 5346377 bytes = 98.2%)\n",
            "- read 273 sequences so far (5269196 of 5346377 bytes = 98.6%)\n",
            "- read 274 sequences so far (5288497 of 5346377 bytes = 98.9%)\n",
            "- read 275 sequences so far (5307798 of 5346377 bytes = 99.3%)\n",
            "- read 276 sequences so far (5327099 of 5346377 bytes = 99.6%)\n",
            "- read 277 sequences so far (5346377 of 5346377 bytes = 100.0%)\n",
            "Analysing fasta file results/Zaire_ebolavirus_aligned_headers_adjusted.afa\n",
            "- analysed 0 / 277 sequences\n",
            "- analysed 1 / 277 sequences\n",
            "- analysed 2 / 277 sequences\n",
            "- analysed 3 / 277 sequences\n",
            "- analysed 4 / 277 sequences\n",
            "- analysed 5 / 277 sequences\n",
            "- analysed 6 / 277 sequences\n",
            "- analysed 7 / 277 sequences\n",
            "- analysed 8 / 277 sequences\n",
            "- analysed 9 / 277 sequences\n",
            "- analysed 10 / 277 sequences\n",
            "- analysed 11 / 277 sequences\n",
            "- analysed 12 / 277 sequences\n",
            "- analysed 13 / 277 sequences\n",
            "- analysed 14 / 277 sequences\n",
            "- analysed 15 / 277 sequences\n",
            "- analysed 16 / 277 sequences\n",
            "- analysed 17 / 277 sequences\n",
            "WARNING (sequence 'KU156715.1|2017-12-13'): ambiguous state Y at site 7668 changed to N\n",
            "- analysed 18 / 277 sequences\n",
            "- analysed 19 / 277 sequences\n",
            "- analysed 20 / 277 sequences\n",
            "- analysed 21 / 277 sequences\n",
            "- analysed 22 / 277 sequences\n",
            "- analysed 23 / 277 sequences\n",
            "- analysed 24 / 277 sequences\n",
            "- analysed 25 / 277 sequences\n",
            "- analysed 26 / 277 sequences\n",
            "- analysed 27 / 277 sequences\n",
            "- analysed 28 / 277 sequences\n",
            "- analysed 29 / 277 sequences\n",
            "- analysed 30 / 277 sequences\n",
            "- analysed 31 / 277 sequences\n",
            "- analysed 32 / 277 sequences\n",
            "- analysed 33 / 277 sequences\n",
            "- analysed 34 / 277 sequences\n",
            "- analysed 35 / 277 sequences\n",
            "- analysed 36 / 277 sequences\n",
            "- analysed 37 / 277 sequences\n",
            "- analysed 38 / 277 sequences\n",
            "- analysed 39 / 277 sequences\n",
            "- analysed 40 / 277 sequences\n",
            "- analysed 41 / 277 sequences\n",
            "- analysed 42 / 277 sequences\n",
            "- analysed 43 / 277 sequences\n",
            "- analysed 44 / 277 sequences\n",
            "- analysed 45 / 277 sequences\n",
            "- analysed 46 / 277 sequences\n",
            "- analysed 47 / 277 sequences\n",
            "- analysed 48 / 277 sequences\n",
            "- analysed 49 / 277 sequences\n",
            "- analysed 50 / 277 sequences\n",
            "- analysed 51 / 277 sequences\n",
            "- analysed 52 / 277 sequences\n",
            "- analysed 53 / 277 sequences\n",
            "- analysed 54 / 277 sequences\n",
            "- analysed 55 / 277 sequences\n",
            "- analysed 56 / 277 sequences\n",
            "- analysed 57 / 277 sequences\n",
            "- analysed 58 / 277 sequences\n",
            "- analysed 59 / 277 sequences\n",
            "- analysed 60 / 277 sequences\n",
            "- analysed 61 / 277 sequences\n",
            "- analysed 62 / 277 sequences\n",
            "- analysed 63 / 277 sequences\n",
            "- analysed 64 / 277 sequences\n",
            "- analysed 65 / 277 sequences\n",
            "- analysed 66 / 277 sequences\n",
            "- analysed 67 / 277 sequences\n",
            "- analysed 68 / 277 sequences\n",
            "- analysed 69 / 277 sequences\n",
            "- analysed 70 / 277 sequences\n",
            "- analysed 71 / 277 sequences\n",
            "- analysed 72 / 277 sequences\n",
            "- analysed 73 / 277 sequences\n",
            "- analysed 74 / 277 sequences\n",
            "- analysed 75 / 277 sequences\n",
            "- analysed 76 / 277 sequences\n",
            "- analysed 77 / 277 sequences\n",
            "- analysed 78 / 277 sequences\n",
            "- analysed 79 / 277 sequences\n",
            "- analysed 80 / 277 sequences\n",
            "- analysed 81 / 277 sequences\n",
            "- analysed 82 / 277 sequences\n",
            "- analysed 83 / 277 sequences\n",
            "- analysed 84 / 277 sequences\n",
            "- analysed 85 / 277 sequences\n",
            "- analysed 86 / 277 sequences\n",
            "- analysed 87 / 277 sequences\n",
            "- analysed 88 / 277 sequences\n",
            "- analysed 89 / 277 sequences\n",
            "- analysed 90 / 277 sequences\n",
            "- analysed 91 / 277 sequences\n",
            "- analysed 92 / 277 sequences\n",
            "- analysed 93 / 277 sequences\n",
            "- analysed 94 / 277 sequences\n",
            "- analysed 95 / 277 sequences\n",
            "- analysed 96 / 277 sequences\n",
            "- analysed 97 / 277 sequences\n",
            "- analysed 98 / 277 sequences\n",
            "- analysed 99 / 277 sequences\n",
            "- analysed 100 / 277 sequences\n",
            "- analysed 101 / 277 sequences\n",
            "- analysed 102 / 277 sequences\n",
            "- analysed 103 / 277 sequences\n",
            "- analysed 104 / 277 sequences\n",
            "- analysed 105 / 277 sequences\n",
            "- analysed 106 / 277 sequences\n",
            "- analysed 107 / 277 sequences\n",
            "- analysed 108 / 277 sequences\n",
            "- analysed 109 / 277 sequences\n",
            "- analysed 110 / 277 sequences\n",
            "- analysed 111 / 277 sequences\n",
            "- analysed 112 / 277 sequences\n",
            "- analysed 113 / 277 sequences\n",
            "- analysed 114 / 277 sequences\n",
            "- analysed 115 / 277 sequences\n",
            "- analysed 116 / 277 sequences\n",
            "- analysed 117 / 277 sequences\n",
            "- analysed 118 / 277 sequences\n",
            "- analysed 119 / 277 sequences\n",
            "- analysed 120 / 277 sequences\n",
            "- analysed 121 / 277 sequences\n",
            "- analysed 122 / 277 sequences\n",
            "- analysed 123 / 277 sequences\n",
            "- analysed 124 / 277 sequences\n",
            "- analysed 125 / 277 sequences\n",
            "- analysed 126 / 277 sequences\n",
            "- analysed 127 / 277 sequences\n",
            "- analysed 128 / 277 sequences\n",
            "- analysed 129 / 277 sequences\n",
            "- analysed 130 / 277 sequences\n",
            "- analysed 131 / 277 sequences\n",
            "- analysed 132 / 277 sequences\n",
            "- analysed 133 / 277 sequences\n",
            "- analysed 134 / 277 sequences\n",
            "- analysed 135 / 277 sequences\n",
            "- analysed 136 / 277 sequences\n",
            "- analysed 137 / 277 sequences\n",
            "- analysed 138 / 277 sequences\n",
            "- analysed 139 / 277 sequences\n",
            "- analysed 140 / 277 sequences\n",
            "- analysed 141 / 277 sequences\n",
            "- analysed 142 / 277 sequences\n",
            "- analysed 143 / 277 sequences\n",
            "- analysed 144 / 277 sequences\n",
            "- analysed 145 / 277 sequences\n",
            "- analysed 146 / 277 sequences\n",
            "- analysed 147 / 277 sequences\n",
            "- analysed 148 / 277 sequences\n",
            "- analysed 149 / 277 sequences\n",
            "- analysed 150 / 277 sequences\n",
            "- analysed 151 / 277 sequences\n",
            "- analysed 152 / 277 sequences\n",
            "- analysed 153 / 277 sequences\n",
            "- analysed 154 / 277 sequences\n",
            "- analysed 155 / 277 sequences\n",
            "- analysed 156 / 277 sequences\n",
            "- analysed 157 / 277 sequences\n",
            "- analysed 158 / 277 sequences\n",
            "- analysed 159 / 277 sequences\n",
            "- analysed 160 / 277 sequences\n",
            "- analysed 161 / 277 sequences\n",
            "- analysed 162 / 277 sequences\n",
            "- analysed 163 / 277 sequences\n",
            "- analysed 164 / 277 sequences\n",
            "- analysed 165 / 277 sequences\n",
            "- analysed 166 / 277 sequences\n",
            "- analysed 167 / 277 sequences\n",
            "- analysed 168 / 277 sequences\n",
            "- analysed 169 / 277 sequences\n",
            "- analysed 170 / 277 sequences\n",
            "- analysed 171 / 277 sequences\n",
            "- analysed 172 / 277 sequences\n",
            "WARNING (sequence 'KP096421.1|2014-11-10'): ambiguous state Y at site 9267 changed to N\n",
            "- analysed 173 / 277 sequences\n",
            "- analysed 174 / 277 sequences\n",
            "- analysed 175 / 277 sequences\n",
            "- analysed 176 / 277 sequences\n",
            "- analysed 177 / 277 sequences\n",
            "- analysed 178 / 277 sequences\n",
            "- analysed 179 / 277 sequences\n",
            "- analysed 180 / 277 sequences\n",
            "- analysed 181 / 277 sequences\n",
            "- analysed 182 / 277 sequences\n",
            "- analysed 183 / 277 sequences\n",
            "- analysed 184 / 277 sequences\n",
            "- analysed 185 / 277 sequences\n",
            "- analysed 186 / 277 sequences\n",
            "- analysed 187 / 277 sequences\n",
            "- analysed 188 / 277 sequences\n",
            "- analysed 189 / 277 sequences\n",
            "- analysed 190 / 277 sequences\n",
            "- analysed 191 / 277 sequences\n",
            "- analysed 192 / 277 sequences\n",
            "- analysed 193 / 277 sequences\n",
            "- analysed 194 / 277 sequences\n",
            "- analysed 195 / 277 sequences\n",
            "- analysed 196 / 277 sequences\n",
            "- analysed 197 / 277 sequences\n",
            "- analysed 198 / 277 sequences\n",
            "- analysed 199 / 277 sequences\n",
            "- analysed 200 / 277 sequences\n",
            "- analysed 201 / 277 sequences\n",
            "- analysed 202 / 277 sequences\n",
            "- analysed 203 / 277 sequences\n",
            "- analysed 204 / 277 sequences\n",
            "- analysed 205 / 277 sequences\n",
            "- analysed 206 / 277 sequences\n",
            "- analysed 207 / 277 sequences\n",
            "- analysed 208 / 277 sequences\n",
            "- analysed 209 / 277 sequences\n",
            "- analysed 210 / 277 sequences\n",
            "- analysed 211 / 277 sequences\n",
            "- analysed 212 / 277 sequences\n",
            "- analysed 213 / 277 sequences\n",
            "- analysed 214 / 277 sequences\n",
            "- analysed 215 / 277 sequences\n",
            "- analysed 216 / 277 sequences\n",
            "- analysed 217 / 277 sequences\n",
            "- analysed 218 / 277 sequences\n",
            "- analysed 219 / 277 sequences\n",
            "- analysed 220 / 277 sequences\n",
            "- analysed 221 / 277 sequences\n",
            "- analysed 222 / 277 sequences\n",
            "- analysed 223 / 277 sequences\n",
            "- analysed 224 / 277 sequences\n",
            "- analysed 225 / 277 sequences\n",
            "- analysed 226 / 277 sequences\n",
            "- analysed 227 / 277 sequences\n",
            "- analysed 228 / 277 sequences\n",
            "- analysed 229 / 277 sequences\n",
            "- analysed 230 / 277 sequences\n",
            "- analysed 231 / 277 sequences\n",
            "- analysed 232 / 277 sequences\n",
            "- analysed 233 / 277 sequences\n",
            "- analysed 234 / 277 sequences\n",
            "- analysed 235 / 277 sequences\n",
            "- analysed 236 / 277 sequences\n",
            "- analysed 237 / 277 sequences\n",
            "- analysed 238 / 277 sequences\n",
            "- analysed 239 / 277 sequences\n",
            "- analysed 240 / 277 sequences\n",
            "- analysed 241 / 277 sequences\n",
            "- analysed 242 / 277 sequences\n",
            "- analysed 243 / 277 sequences\n",
            "- analysed 244 / 277 sequences\n",
            "- analysed 245 / 277 sequences\n",
            "- analysed 246 / 277 sequences\n",
            "- analysed 247 / 277 sequences\n",
            "- analysed 248 / 277 sequences\n",
            "- analysed 249 / 277 sequences\n",
            "- analysed 250 / 277 sequences\n",
            "- analysed 251 / 277 sequences\n",
            "- analysed 252 / 277 sequences\n",
            "- analysed 253 / 277 sequences\n",
            "- analysed 254 / 277 sequences\n",
            "- analysed 255 / 277 sequences\n",
            "- analysed 256 / 277 sequences\n",
            "- analysed 257 / 277 sequences\n",
            "- analysed 258 / 277 sequences\n",
            "- analysed 259 / 277 sequences\n",
            "- analysed 260 / 277 sequences\n",
            "- analysed 261 / 277 sequences\n",
            "- analysed 262 / 277 sequences\n",
            "- analysed 263 / 277 sequences\n",
            "- analysed 264 / 277 sequences\n",
            "- analysed 265 / 277 sequences\n",
            "- analysed 266 / 277 sequences\n",
            "- analysed 267 / 277 sequences\n",
            "- analysed 268 / 277 sequences\n",
            "- analysed 269 / 277 sequences\n",
            "- analysed 270 / 277 sequences\n",
            "- analysed 271 / 277 sequences\n",
            "- analysed 272 / 277 sequences\n",
            "- analysed 273 / 277 sequences\n",
            "- analysed 274 / 277 sequences\n",
            "- analysed 275 / 277 sequences\n",
            "- analysed 276 / 277 sequences\n",
            "- analysed 277 / 277 sequences\n",
            "Building rough initial tree...\n",
            "- added 0 / 277 tips\n",
            "- added 2 / 277 tips\n",
            "- added 3 / 277 tips\n",
            "- added 4 / 277 tips\n",
            "- added 5 / 277 tips\n",
            "- added 6 / 277 tips\n",
            "- added 7 / 277 tips\n",
            "- added 8 / 277 tips\n",
            "- added 9 / 277 tips\n",
            "- added 10 / 277 tips\n",
            "- added 11 / 277 tips\n",
            "- added 12 / 277 tips\n",
            "- added 13 / 277 tips\n",
            "- added 14 / 277 tips\n",
            "- added 15 / 277 tips\n",
            "- added 16 / 277 tips\n",
            "- added 17 / 277 tips\n",
            "- added 18 / 277 tips\n",
            "- added 19 / 277 tips\n",
            "- added 20 / 277 tips\n",
            "- added 21 / 277 tips\n",
            "- added 22 / 277 tips\n",
            "- added 23 / 277 tips\n",
            "- added 24 / 277 tips\n",
            "- added 25 / 277 tips\n",
            "- added 26 / 277 tips\n",
            "- added 27 / 277 tips\n",
            "- added 28 / 277 tips\n",
            "- added 29 / 277 tips\n",
            "- added 30 / 277 tips\n",
            "- added 31 / 277 tips\n",
            "- added 32 / 277 tips\n",
            "- added 33 / 277 tips\n",
            "- added 34 / 277 tips\n",
            "- added 35 / 277 tips\n",
            "- added 36 / 277 tips\n",
            "- added 37 / 277 tips\n",
            "- added 38 / 277 tips\n",
            "- added 39 / 277 tips\n",
            "- added 40 / 277 tips\n",
            "- added 41 / 277 tips\n",
            "- added 42 / 277 tips\n",
            "- added 43 / 277 tips\n",
            "- added 44 / 277 tips\n",
            "- added 45 / 277 tips\n",
            "- added 46 / 277 tips\n",
            "- added 47 / 277 tips\n",
            "- added 48 / 277 tips\n",
            "- added 49 / 277 tips\n",
            "- added 50 / 277 tips\n",
            "- added 51 / 277 tips\n",
            "- added 52 / 277 tips\n",
            "- added 53 / 277 tips\n",
            "- added 54 / 277 tips\n",
            "- added 55 / 277 tips\n",
            "- added 56 / 277 tips\n",
            "- added 57 / 277 tips\n",
            "- added 58 / 277 tips\n",
            "- added 59 / 277 tips\n",
            "- added 60 / 277 tips\n",
            "- added 61 / 277 tips\n",
            "- added 62 / 277 tips\n",
            "- added 63 / 277 tips\n",
            "- added 64 / 277 tips\n",
            "- added 65 / 277 tips\n",
            "- added 66 / 277 tips\n",
            "- added 67 / 277 tips\n",
            "- added 68 / 277 tips\n",
            "- added 69 / 277 tips\n",
            "- added 70 / 277 tips\n",
            "- added 71 / 277 tips\n",
            "- added 72 / 277 tips\n",
            "- added 73 / 277 tips\n",
            "- added 74 / 277 tips\n",
            "- added 75 / 277 tips\n",
            "- added 76 / 277 tips\n",
            "- added 77 / 277 tips\n",
            "- added 78 / 277 tips\n",
            "- added 79 / 277 tips\n",
            "- added 80 / 277 tips\n",
            "- added 81 / 277 tips\n",
            "- added 82 / 277 tips\n",
            "- added 83 / 277 tips\n",
            "- added 84 / 277 tips\n",
            "- added 85 / 277 tips\n",
            "- added 86 / 277 tips\n",
            "- added 87 / 277 tips\n",
            "- added 88 / 277 tips\n",
            "- added 89 / 277 tips\n",
            "- added 90 / 277 tips\n",
            "- added 91 / 277 tips\n",
            "- added 92 / 277 tips\n",
            "- added 93 / 277 tips\n",
            "- added 94 / 277 tips\n",
            "- added 95 / 277 tips\n",
            "- added 96 / 277 tips\n",
            "- added 97 / 277 tips\n",
            "- added 98 / 277 tips\n",
            "- added 99 / 277 tips\n",
            "- added 100 / 277 tips\n",
            "- added 101 / 277 tips\n",
            "- added 102 / 277 tips\n",
            "- added 103 / 277 tips\n",
            "- added 104 / 277 tips\n",
            "- added 105 / 277 tips\n",
            "- added 106 / 277 tips\n",
            "- added 107 / 277 tips\n",
            "- added 108 / 277 tips\n",
            "- added 109 / 277 tips\n",
            "- added 110 / 277 tips\n",
            "- added 111 / 277 tips\n",
            "- added 112 / 277 tips\n",
            "- added 113 / 277 tips\n",
            "- added 114 / 277 tips\n",
            "- added 115 / 277 tips\n",
            "- added 116 / 277 tips\n",
            "- added 117 / 277 tips\n",
            "- added 118 / 277 tips\n",
            "- added 119 / 277 tips\n",
            "- added 120 / 277 tips\n",
            "- added 121 / 277 tips\n",
            "- added 122 / 277 tips\n",
            "- added 123 / 277 tips\n",
            "- added 124 / 277 tips\n",
            "- added 125 / 277 tips\n",
            "- added 126 / 277 tips\n",
            "- added 127 / 277 tips\n",
            "- added 128 / 277 tips\n",
            "- added 129 / 277 tips\n",
            "- added 130 / 277 tips\n",
            "- added 131 / 277 tips\n",
            "- added 132 / 277 tips\n",
            "- added 133 / 277 tips\n",
            "- added 134 / 277 tips\n",
            "- added 135 / 277 tips\n",
            "- added 136 / 277 tips\n",
            "- added 137 / 277 tips\n",
            "- added 138 / 277 tips\n",
            "- added 139 / 277 tips\n",
            "- added 140 / 277 tips\n",
            "- added 141 / 277 tips\n",
            "- added 142 / 277 tips\n",
            "- added 143 / 277 tips\n",
            "- added 144 / 277 tips\n",
            "- added 145 / 277 tips\n",
            "- added 146 / 277 tips\n",
            "- added 147 / 277 tips\n",
            "- added 148 / 277 tips\n",
            "- added 149 / 277 tips\n",
            "- added 150 / 277 tips\n",
            "- added 151 / 277 tips\n",
            "- added 152 / 277 tips\n",
            "- added 153 / 277 tips\n",
            "- added 154 / 277 tips\n",
            "- added 155 / 277 tips\n",
            "- added 156 / 277 tips\n",
            "- added 157 / 277 tips\n",
            "- added 158 / 277 tips\n",
            "- added 159 / 277 tips\n",
            "- added 160 / 277 tips\n",
            "- added 161 / 277 tips\n",
            "- added 162 / 277 tips\n",
            "- added 163 / 277 tips\n",
            "- added 164 / 277 tips\n",
            "- added 165 / 277 tips\n",
            "- added 166 / 277 tips\n",
            "- added 167 / 277 tips\n",
            "- added 168 / 277 tips\n",
            "- added 169 / 277 tips\n",
            "- added 170 / 277 tips\n",
            "- added 171 / 277 tips\n",
            "- added 172 / 277 tips\n",
            "- added 173 / 277 tips\n",
            "- added 174 / 277 tips\n",
            "- added 175 / 277 tips\n",
            "- added 176 / 277 tips\n",
            "- added 177 / 277 tips\n",
            "- added 178 / 277 tips\n",
            "- added 179 / 277 tips\n",
            "- added 180 / 277 tips\n",
            "- added 181 / 277 tips\n",
            "- added 182 / 277 tips\n",
            "- added 183 / 277 tips\n",
            "- added 184 / 277 tips\n",
            "- added 185 / 277 tips\n",
            "- added 186 / 277 tips\n",
            "- added 187 / 277 tips\n",
            "- added 188 / 277 tips\n",
            "- added 189 / 277 tips\n",
            "- added 190 / 277 tips\n",
            "- added 191 / 277 tips\n",
            "- added 192 / 277 tips\n",
            "- added 193 / 277 tips\n",
            "- added 194 / 277 tips\n",
            "- added 195 / 277 tips\n",
            "- added 196 / 277 tips\n",
            "- added 197 / 277 tips\n",
            "- added 198 / 277 tips\n",
            "- added 199 / 277 tips\n",
            "- added 200 / 277 tips\n",
            "- added 201 / 277 tips\n",
            "- added 202 / 277 tips\n",
            "- added 203 / 277 tips\n",
            "- added 204 / 277 tips\n",
            "- added 205 / 277 tips\n",
            "- added 206 / 277 tips\n",
            "- added 207 / 277 tips\n",
            "- added 208 / 277 tips\n",
            "- added 209 / 277 tips\n",
            "- added 210 / 277 tips\n",
            "- added 211 / 277 tips\n",
            "- added 212 / 277 tips\n",
            "- added 213 / 277 tips\n",
            "- added 214 / 277 tips\n",
            "- added 215 / 277 tips\n",
            "- added 216 / 277 tips\n",
            "- added 217 / 277 tips\n",
            "- added 218 / 277 tips\n",
            "- added 219 / 277 tips\n",
            "- added 220 / 277 tips\n",
            "- added 221 / 277 tips\n",
            "- added 222 / 277 tips\n",
            "- added 223 / 277 tips\n",
            "- added 224 / 277 tips\n",
            "- added 225 / 277 tips\n",
            "- added 226 / 277 tips\n",
            "- added 227 / 277 tips\n",
            "- added 228 / 277 tips\n",
            "- added 229 / 277 tips\n",
            "- added 230 / 277 tips\n",
            "- added 231 / 277 tips\n",
            "- added 232 / 277 tips\n",
            "- added 233 / 277 tips\n",
            "- added 234 / 277 tips\n",
            "- added 235 / 277 tips\n",
            "- added 236 / 277 tips\n",
            "- added 237 / 277 tips\n",
            "- added 238 / 277 tips\n",
            "- added 239 / 277 tips\n",
            "- added 240 / 277 tips\n",
            "- added 241 / 277 tips\n",
            "- added 242 / 277 tips\n",
            "- added 243 / 277 tips\n",
            "- added 244 / 277 tips\n",
            "- added 245 / 277 tips\n",
            "- added 246 / 277 tips\n",
            "- added 247 / 277 tips\n",
            "- added 248 / 277 tips\n",
            "- added 249 / 277 tips\n",
            "- added 250 / 277 tips\n",
            "- added 251 / 277 tips\n",
            "- added 252 / 277 tips\n",
            "- added 253 / 277 tips\n",
            "- added 254 / 277 tips\n",
            "- added 255 / 277 tips\n",
            "- added 256 / 277 tips\n",
            "- added 257 / 277 tips\n",
            "- added 258 / 277 tips\n",
            "- added 259 / 277 tips\n",
            "- added 260 / 277 tips\n",
            "- added 261 / 277 tips\n",
            "- added 262 / 277 tips\n",
            "- added 263 / 277 tips\n",
            "- added 264 / 277 tips\n",
            "- added 265 / 277 tips\n",
            "- added 266 / 277 tips\n",
            "- added 267 / 277 tips\n",
            "- added 268 / 277 tips\n",
            "- added 269 / 277 tips\n",
            "- added 270 / 277 tips\n",
            "- added 271 / 277 tips\n",
            "- added 272 / 277 tips\n",
            "- added 273 / 277 tips\n",
            "- added 274 / 277 tips\n",
            "- added 275 / 277 tips\n",
            "- added 276 / 277 tips\n",
            "- added 277 / 277 tips\n",
            "Built initial tree with 277 tips\n",
            "# Parallelism: 2\n",
            "0.0 M steps/s, Step 0, log_posterior = -67423.37, log_G = -64877.11, log_coal = -6777.30, log_other_priors = 4231.03, num_muts = 1652, T = 307473.99, t_MRCA = 1992-11-28, n0 = 1095.0, g = 0.0000, mu = 1 * 10^-3 subst / site / year, M_ts = 1263, M_tv = 389, kappa = 1, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4053, 3754, 5099], num_muts_ab counts: [73, 309, 48, 49, 27, 285, 309, 58, 54, 38, 360, 42]\n",
            "0.0 M steps/s, Step 692500, log_posterior = -46696.77, log_G = -48314.54, log_coal = -2609.98, log_other_priors = 4227.74, num_muts = 1409, T = 104543.68, t_MRCA = 1948-06-18, n0 = 4841.3, g = -0.0000, mu = 0.28 * 10^-3 subst / site / year, M_ts = 1069, M_tv = 340, kappa = 6.52, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4058, 3744, 5101], num_muts_ab counts: [63, 287, 43, 42, 19, 253, 239, 51, 44, 34, 290, 44]\n",
            "0.5 M steps/s, Step 1385000, log_posterior = -46397.81, log_G = -47980.76, log_coal = -2644.72, log_other_priors = 4227.66, num_muts = 1385, T = 106957.03, t_MRCA = 1952-02-21, n0 = 5789.3, g = 0.0000, mu = 0.26 * 10^-3 subst / site / year, M_ts = 1055, M_tv = 330, kappa = 6.11, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6044, 4062, 3753, 5101], num_muts_ab counts: [57, 270, 45, 45, 17, 253, 244, 51, 39, 32, 288, 44]\n",
            "0.5 M steps/s, Step 2077500, log_posterior = -46230.55, log_G = -47812.89, log_coal = -2645.00, log_other_priors = 4227.34, num_muts = 1374, T = 106441.50, t_MRCA = 1952-03-23, n0 = 6102.9, g = 0.0000, mu = 0.25 * 10^-3 subst / site / year, M_ts = 1045, M_tv = 329, kappa = 7.27, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6070, 4066, 3725, 5099], num_muts_ab counts: [62, 298, 36, 42, 17, 260, 211, 49, 42, 38, 276, 43]\n",
            "0.5 M steps/s, Step 2770000, log_posterior = -46003.12, log_G = -47580.03, log_coal = -2650.71, log_other_priors = 4227.62, num_muts = 1357, T = 108690.00, t_MRCA = 1950-11-12, n0 = 5614.7, g = 0.0000, mu = 0.25 * 10^-3 subst / site / year, M_ts = 1035, M_tv = 322, kappa = 6.44, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6048, 4054, 3753, 5105], num_muts_ab counts: [64, 272, 36, 40, 15, 249, 235, 47, 41, 37, 279, 42]\n",
            "0.5 M steps/s, Step 3462500, log_posterior = -46011.65, log_G = -47581.67, log_coal = -2657.40, log_other_priors = 4227.42, num_muts = 1356, T = 109485.84, t_MRCA = 1950-07-12, n0 = 6491.5, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1032, M_tv = 324, kappa = 6.66, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6060, 4068, 3750, 5082], num_muts_ab counts: [64, 276, 44, 39, 16, 263, 228, 46, 44, 31, 265, 40]\n",
            "0.5 M steps/s, Step 4155000, log_posterior = -45888.92, log_G = -47451.09, log_coal = -2665.38, log_other_priors = 4227.55, num_muts = 1347, T = 109883.85, t_MRCA = 1951-09-07, n0 = 6217.6, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1026, M_tv = 321, kappa = 6.28, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4043, 3748, 5108], num_muts_ab counts: [70, 272, 41, 31, 18, 240, 230, 43, 42, 32, 284, 44]\n",
            "0.5 M steps/s, Step 4847500, log_posterior = -45896.52, log_G = -47451.89, log_coal = -2672.28, log_other_priors = 4227.64, num_muts = 1347, T = 109990.49, t_MRCA = 1952-03-22, n0 = 5757.0, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1026, M_tv = 321, kappa = 6.24, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4043, 3748, 5108], num_muts_ab counts: [70, 272, 41, 31, 17, 240, 230, 43, 42, 32, 284, 45]\n",
            "0.5 M steps/s, Step 5540000, log_posterior = -45818.48, log_G = -47389.61, log_coal = -2656.27, log_other_priors = 4227.40, num_muts = 1343, T = 110032.82, t_MRCA = 1951-03-12, n0 = 6782.0, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1024, M_tv = 319, kappa = 6.55, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4062, 3740, 5095], num_muts_ab counts: [65, 279, 42, 36, 19, 251, 226, 43, 39, 30, 268, 45]\n",
            "0.5 M steps/s, Step 6232500, log_posterior = -45847.63, log_G = -47393.97, log_coal = -2680.97, log_other_priors = 4227.31, num_muts = 1340, T = 113765.20, t_MRCA = 1950-01-27, n0 = 7383.8, g = 0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 320, kappa = 6.57, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4062, 3740, 5095], num_muts_ab counts: [64, 279, 42, 36, 19, 248, 226, 44, 40, 30, 267, 45]\n",
            "0.5 M steps/s, Step 6925000, log_posterior = -45810.83, log_G = -47374.89, log_coal = -2663.54, log_other_priors = 4227.59, num_muts = 1341, T = 111183.38, t_MRCA = 1951-03-08, n0 = 5814.2, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1022, M_tv = 319, kappa = 6.38, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4062, 3740, 5095], num_muts_ab counts: [65, 279, 42, 36, 20, 249, 226, 43, 39, 30, 268, 44]\n",
            "0.5 M steps/s, Step 7617500, log_posterior = -45728.84, log_G = -47323.00, log_coal = -2633.54, log_other_priors = 4227.70, num_muts = 1342, T = 104920.19, t_MRCA = 1951-04-11, n0 = 5825.0, g = 0.0000, mu = 0.26 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 322, kappa = 5.96, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6062, 4063, 3740, 5095], num_muts_ab counts: [66, 278, 42, 37, 19, 249, 226, 44, 39, 30, 267, 45]\n",
            "0.5 M steps/s, Step 8310000, log_posterior = -45658.28, log_G = -47248.30, log_coal = -2637.56, log_other_priors = 4227.57, num_muts = 1338, T = 104380.66, t_MRCA = 1951-01-09, n0 = 5006.5, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 320, kappa = 7.11, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6063, 4062, 3740, 5095], num_muts_ab counts: [66, 278, 43, 37, 19, 248, 227, 44, 37, 30, 265, 44]\n",
            "0.5 M steps/s, Step 9002500, log_posterior = -45684.82, log_G = -47264.57, log_coal = -2647.66, log_other_priors = 4227.41, num_muts = 1336, T = 107919.49, t_MRCA = 1950-09-10, n0 = 4885.9, g = -0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 7.98, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6041, 4063, 3767, 5089], num_muts_ab counts: [63, 255, 46, 41, 16, 247, 249, 44, 41, 26, 266, 42]\n",
            "0.5 M steps/s, Step 9695000, log_posterior = -45757.66, log_G = -47317.41, log_coal = -2667.96, log_other_priors = 4227.71, num_muts = 1337, T = 111080.55, t_MRCA = 1950-03-27, n0 = 6126.5, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 5.72, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6041, 4063, 3767, 5089], num_muts_ab counts: [63, 256, 46, 41, 17, 247, 249, 44, 41, 26, 266, 41]\n",
            "0.5 M steps/s, Step 10387500, log_posterior = -45721.43, log_G = -47292.73, log_coal = -2656.17, log_other_priors = 4227.47, num_muts = 1337, T = 109012.00, t_MRCA = 1950-05-22, n0 = 5699.6, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 320, kappa = 6.99, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6041, 4063, 3767, 5089], num_muts_ab counts: [63, 255, 46, 42, 17, 246, 250, 44, 41, 26, 266, 41]\n",
            "0.4 M steps/s, Step 11080000, log_posterior = -45708.98, log_G = -47282.40, log_coal = -2654.05, log_other_priors = 4227.48, num_muts = 1336, T = 109334.20, t_MRCA = 1950-02-27, n0 = 5731.0, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 321, kappa = 6.93, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6041, 4063, 3767, 5089], num_muts_ab counts: [63, 254, 46, 42, 18, 246, 249, 44, 41, 26, 266, 41]\n",
            "0.5 M steps/s, Step 11772500, log_posterior = -45830.30, log_G = -47425.22, log_coal = -2632.66, log_other_priors = 4227.57, num_muts = 1347, T = 107423.89, t_MRCA = 1950-08-10, n0 = 5582.0, g = 0.0000, mu = 0.25 * 10^-3 subst / site / year, M_ts = 1022, M_tv = 325, kappa = 6.63, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4043, 3744, 5118], num_muts_ab counts: [61, 279, 40, 43, 18, 228, 226, 46, 40, 36, 289, 41]\n",
            "0.5 M steps/s, Step 12465000, log_posterior = -45893.69, log_G = -47448.26, log_coal = -2672.81, log_other_priors = 4227.38, num_muts = 1345, T = 111730.85, t_MRCA = 1950-08-15, n0 = 6398.1, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1021, M_tv = 324, kappa = 6.89, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4043, 3744, 5118], num_muts_ab counts: [61, 278, 39, 43, 17, 228, 227, 47, 40, 35, 288, 42]\n",
            "0.5 M steps/s, Step 13157500, log_posterior = -45848.90, log_G = -47421.84, log_coal = -2654.48, log_other_priors = 4227.42, num_muts = 1345, T = 109553.03, t_MRCA = 1951-03-18, n0 = 6657.6, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 325, kappa = 6.56, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4043, 3744, 5118], num_muts_ab counts: [61, 277, 39, 43, 18, 228, 227, 47, 40, 35, 288, 42]\n",
            "0.5 M steps/s, Step 13850000, log_posterior = -45871.71, log_G = -47438.24, log_coal = -2660.97, log_other_priors = 4227.49, num_muts = 1345, T = 110832.06, t_MRCA = 1949-08-31, n0 = 6853.3, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 325, kappa = 6.13, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4043, 3744, 5118], num_muts_ab counts: [61, 278, 39, 43, 18, 227, 227, 47, 40, 35, 288, 42]\n",
            "0.5 M steps/s, Step 14542500, log_posterior = -45818.37, log_G = -47410.70, log_coal = -2635.26, log_other_priors = 4227.59, num_muts = 1346, T = 107310.16, t_MRCA = 1949-02-26, n0 = 5606.2, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1021, M_tv = 325, kappa = 6.55, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4043, 3744, 5118], num_muts_ab counts: [61, 278, 40, 43, 18, 227, 227, 47, 39, 35, 289, 42]\n",
            "0.5 M steps/s, Step 15235000, log_posterior = -45833.87, log_G = -47413.83, log_coal = -2647.54, log_other_priors = 4227.50, num_muts = 1345, T = 109033.11, t_MRCA = 1948-01-15, n0 = 6301.6, g = 0.0001, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1021, M_tv = 324, kappa = 6.44, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6055, 4043, 3744, 5118], num_muts_ab counts: [61, 278, 39, 43, 16, 228, 227, 47, 40, 35, 288, 43]\n",
            "0.5 M steps/s, Step 15927500, log_posterior = -45640.49, log_G = -47232.05, log_coal = -2636.02, log_other_priors = 4227.57, num_muts = 1335, T = 106656.42, t_MRCA = 1948-12-11, n0 = 5297.7, g = -0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.86, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4044, 3757, 5112], num_muts_ab counts: [64, 265, 39, 41, 17, 229, 237, 45, 42, 33, 284, 39]\n",
            "0.5 M steps/s, Step 16620000, log_posterior = -45813.43, log_G = -47393.08, log_coal = -2647.84, log_other_priors = 4227.50, num_muts = 1341, T = 111777.72, t_MRCA = 1945-06-29, n0 = 6626.0, g = 0.0000, mu = 0.25 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 324, kappa = 6.24, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6071, 4078, 3731, 5080], num_muts_ab counts: [64, 284, 42, 41, 21, 262, 215, 42, 41, 31, 256, 42]\n",
            "0.5 M steps/s, Step 17312500, log_posterior = -45893.07, log_G = -47448.74, log_coal = -2671.70, log_other_priors = 4227.37, num_muts = 1342, T = 115429.81, t_MRCA = 1945-10-13, n0 = 6547.7, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 323, kappa = 6.82, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4059, 3746, 5094], num_muts_ab counts: [68, 273, 39, 37, 21, 247, 227, 42, 45, 33, 272, 38]\n",
            "0.5 M steps/s, Step 18005000, log_posterior = -45757.44, log_G = -47324.47, log_coal = -2660.56, log_other_priors = 4227.60, num_muts = 1335, T = 114393.83, t_MRCA = 1944-11-14, n0 = 5697.6, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.46, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6035, 4049, 3762, 5114], num_muts_ab counts: [60, 254, 40, 45, 18, 231, 246, 44, 38, 31, 285, 43]\n",
            "0.5 M steps/s, Step 18697500, log_posterior = -45798.66, log_G = -47345.34, log_coal = -2680.88, log_other_priors = 4227.56, num_muts = 1334, T = 117630.66, t_MRCA = 1943-06-13, n0 = 6059.4, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 319, kappa = 6.34, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6046, 4047, 3763, 5104], num_muts_ab counts: [68, 256, 41, 35, 17, 238, 244, 44, 41, 32, 277, 41]\n",
            "0.5 M steps/s, Step 19390000, log_posterior = -45813.80, log_G = -47387.78, log_coal = -2653.67, log_other_priors = 4227.65, num_muts = 1339, T = 114630.68, t_MRCA = 1943-09-29, n0 = 5432.9, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 321, kappa = 6.42, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6058, 4055, 3748, 5099], num_muts_ab counts: [67, 274, 39, 40, 18, 241, 231, 43, 43, 32, 272, 39]\n",
            "0.4 M steps/s, Step 20082500, log_posterior = -45893.81, log_G = -47441.39, log_coal = -2679.87, log_other_priors = 4227.45, num_muts = 1340, T = 117877.18, t_MRCA = 1943-02-18, n0 = 6393.6, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 322, kappa = 6.57, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6058, 4055, 3748, 5099], num_muts_ab counts: [67, 273, 40, 40, 19, 241, 231, 43, 42, 32, 273, 39]\n",
            "0.4 M steps/s, Step 20775000, log_posterior = -45776.48, log_G = -47352.88, log_coal = -2651.15, log_other_priors = 4227.54, num_muts = 1338, T = 112609.01, t_MRCA = 1945-06-14, n0 = 5487.5, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 321, kappa = 6.85, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6049, 4075, 3755, 5081], num_muts_ab counts: [63, 265, 42, 44, 19, 256, 238, 42, 43, 29, 258, 39]\n",
            "0.4 M steps/s, Step 21467500, log_posterior = -45851.15, log_G = -47408.03, log_coal = -2670.72, log_other_priors = 4227.60, num_muts = 1339, T = 116222.00, t_MRCA = 1943-05-24, n0 = 5486.8, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 322, kappa = 6.6, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6035, 4076, 3765, 5084], num_muts_ab counts: [60, 254, 42, 44, 19, 255, 251, 44, 40, 31, 257, 42]\n",
            "0.4 M steps/s, Step 22160000, log_posterior = -45887.89, log_G = -47447.17, log_coal = -2668.21, log_other_priors = 4227.49, num_muts = 1340, T = 118065.94, t_MRCA = 1941-10-23, n0 = 6074.7, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 322, kappa = 6.65, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6035, 4076, 3765, 5084], num_muts_ab counts: [60, 254, 41, 44, 19, 255, 251, 44, 41, 31, 258, 42]\n",
            "0.4 M steps/s, Step 22852500, log_posterior = -45847.66, log_G = -47387.17, log_coal = -2687.85, log_other_priors = 4227.36, num_muts = 1336, T = 118428.31, t_MRCA = 1943-10-24, n0 = 7374.0, g = 0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 321, kappa = 6.38, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6035, 4076, 3765, 5084], num_muts_ab counts: [59, 254, 41, 44, 18, 255, 250, 45, 41, 31, 256, 42]\n",
            "0.4 M steps/s, Step 23545000, log_posterior = -45792.72, log_G = -47355.03, log_coal = -2665.10, log_other_priors = 4227.41, num_muts = 1337, T = 114511.82, t_MRCA = 1943-07-17, n0 = 6091.3, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 6.94, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4068, 3740, 5091], num_muts_ab counts: [60, 282, 40, 45, 15, 250, 224, 47, 39, 30, 262, 43]\n",
            "0.4 M steps/s, Step 24237500, log_posterior = -45821.96, log_G = -47368.57, log_coal = -2680.63, log_other_priors = 4227.24, num_muts = 1336, T = 117190.41, t_MRCA = 1944-07-21, n0 = 7023.9, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 7.08, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4068, 3740, 5091], num_muts_ab counts: [59, 282, 41, 45, 14, 250, 224, 48, 38, 30, 262, 43]\n",
            "0.4 M steps/s, Step 24930000, log_posterior = -45806.01, log_G = -47354.50, log_coal = -2679.00, log_other_priors = 4227.49, num_muts = 1336, T = 115652.91, t_MRCA = 1946-02-26, n0 = 5728.8, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.87, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4068, 3740, 5091], num_muts_ab counts: [60, 281, 40, 45, 15, 250, 224, 47, 39, 30, 262, 43]\n",
            "0.4 M steps/s, Step 25622500, log_posterior = -45759.85, log_G = -47338.64, log_coal = -2648.81, log_other_priors = 4227.60, num_muts = 1338, T = 111913.27, t_MRCA = 1946-05-05, n0 = 5610.6, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 319, kappa = 6.5, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4068, 3740, 5091], num_muts_ab counts: [60, 282, 40, 45, 15, 251, 224, 47, 39, 30, 262, 43]\n",
            "0.5 M steps/s, Step 26315000, log_posterior = -45746.95, log_G = -47316.62, log_coal = -2657.92, log_other_priors = 4227.59, num_muts = 1336, T = 112474.72, t_MRCA = 1945-10-28, n0 = 5483.9, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.65, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4068, 3740, 5091], num_muts_ab counts: [60, 280, 41, 44, 16, 249, 225, 47, 39, 30, 262, 43]\n",
            "0.5 M steps/s, Step 27007500, log_posterior = -45775.71, log_G = -47339.01, log_coal = -2664.20, log_other_priors = 4227.50, num_muts = 1337, T = 113213.68, t_MRCA = 1945-12-15, n0 = 6575.0, g = 0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 6.28, pi = [0.32, 0.21, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4068, 3740, 5091], num_muts_ab counts: [59, 280, 42, 44, 16, 250, 225, 46, 38, 30, 264, 43]\n",
            "0.5 M steps/s, Step 27700000, log_posterior = -45787.31, log_G = -47339.08, log_coal = -2675.68, log_other_priors = 4227.46, num_muts = 1336, T = 114747.17, t_MRCA = 1944-08-16, n0 = 6151.0, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 317, kappa = 6.71, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4068, 3740, 5091], num_muts_ab counts: [60, 280, 40, 43, 16, 250, 226, 46, 39, 30, 263, 43]\n",
            "0.5 M steps/s, Step 28392500, log_posterior = -45778.24, log_G = -47347.16, log_coal = -2658.47, log_other_priors = 4227.39, num_muts = 1337, T = 114069.77, t_MRCA = 1944-11-21, n0 = 5826.1, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 7.26, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4068, 3740, 5091], num_muts_ab counts: [60, 280, 41, 44, 16, 250, 227, 46, 38, 30, 262, 43]\n",
            "0.5 M steps/s, Step 29085000, log_posterior = -45723.71, log_G = -47309.23, log_coal = -2642.10, log_other_priors = 4227.62, num_muts = 1337, T = 110412.25, t_MRCA = 1945-12-02, n0 = 5611.6, g = -0.0000, mu = 0.24 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 320, kappa = 6.43, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4036, 3747, 5120], num_muts_ab counts: [64, 275, 41, 40, 18, 219, 233, 45, 39, 33, 290, 40]\n",
            "0.5 M steps/s, Step 29777500, log_posterior = -45873.73, log_G = -47416.83, log_coal = -2684.21, log_other_priors = 4227.31, num_muts = 1338, T = 118218.28, t_MRCA = 1943-01-25, n0 = 5765.8, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 321, kappa = 7.67, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4036, 3747, 5120], num_muts_ab counts: [64, 274, 40, 39, 19, 220, 232, 45, 40, 34, 291, 40]\n",
            "0.5 M steps/s, Step 30470000, log_posterior = -45869.36, log_G = -47414.06, log_coal = -2682.98, log_other_priors = 4227.69, num_muts = 1337, T = 119669.93, t_MRCA = 1941-05-11, n0 = 6029.8, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 5.85, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4036, 3747, 5120], num_muts_ab counts: [64, 275, 40, 39, 18, 219, 233, 45, 40, 33, 291, 40]\n",
            "0.5 M steps/s, Step 31162500, log_posterior = -45836.54, log_G = -47395.28, log_coal = -2668.75, log_other_priors = 4227.49, num_muts = 1337, T = 118071.99, t_MRCA = 1940-06-08, n0 = 6203.7, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 6.56, pi = [0.32, 0.21, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4036, 3747, 5120], num_muts_ab counts: [64, 275, 40, 39, 18, 219, 233, 45, 40, 33, 291, 40]\n",
            "0.5 M steps/s, Step 31855000, log_posterior = -45930.16, log_G = -47455.41, log_coal = -2702.25, log_other_priors = 4227.51, num_muts = 1337, T = 122776.24, t_MRCA = 1941-02-03, n0 = 6365.6, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 6.37, pi = [0.31, 0.21, 0.21, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4036, 3747, 5120], num_muts_ab counts: [64, 275, 40, 39, 18, 219, 233, 45, 40, 33, 291, 40]\n",
            "0.5 M steps/s, Step 32547500, log_posterior = -45867.72, log_G = -47414.87, log_coal = -2680.67, log_other_priors = 4227.82, num_muts = 1337, T = 119851.95, t_MRCA = 1940-04-15, n0 = 5480.4, g = -0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 5.72, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4036, 3747, 5120], num_muts_ab counts: [63, 276, 40, 40, 17, 219, 232, 45, 40, 33, 292, 40]\n",
            "0.4 M steps/s, Step 33240000, log_posterior = -45839.78, log_G = -47400.85, log_coal = -2666.59, log_other_priors = 4227.66, num_muts = 1337, T = 118485.31, t_MRCA = 1939-10-17, n0 = 5287.2, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 6.49, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6060, 4049, 3747, 5104], num_muts_ab counts: [62, 273, 46, 42, 17, 231, 231, 43, 41, 27, 283, 41]\n",
            "0.4 M steps/s, Step 33932500, log_posterior = -45911.03, log_G = -47451.76, log_coal = -2686.75, log_other_priors = 4227.48, num_muts = 1337, T = 122813.25, t_MRCA = 1938-08-05, n0 = 6038.2, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 320, kappa = 6.7, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6060, 4049, 3747, 5104], num_muts_ab counts: [62, 272, 46, 42, 18, 231, 231, 43, 41, 27, 283, 41]\n",
            "0.4 M steps/s, Step 34625000, log_posterior = -45872.82, log_G = -47428.02, log_coal = -2672.46, log_other_priors = 4227.66, num_muts = 1337, T = 120609.47, t_MRCA = 1938-05-07, n0 = 5745.8, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 321, kappa = 6.15, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6060, 4049, 3747, 5104], num_muts_ab counts: [62, 271, 46, 41, 19, 231, 231, 43, 41, 28, 283, 41]\n",
            "0.4 M steps/s, Step 35317500, log_posterior = -45916.23, log_G = -47463.75, log_coal = -2680.07, log_other_priors = 4227.59, num_muts = 1339, T = 121472.28, t_MRCA = 1938-07-13, n0 = 5969.1, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 319, kappa = 6.31, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4050, 3759, 5104], num_muts_ab counts: [68, 262, 38, 35, 17, 239, 245, 44, 41, 34, 274, 42]\n",
            "0.5 M steps/s, Step 36010000, log_posterior = -45902.53, log_G = -47450.76, log_coal = -2678.91, log_other_priors = 4227.14, num_muts = 1339, T = 120034.78, t_MRCA = 1939-12-30, n0 = 7221.6, g = 0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 320, kappa = 7.39, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4050, 3759, 5104], num_muts_ab counts: [68, 261, 38, 35, 18, 239, 245, 44, 41, 34, 274, 42]\n",
            "0.5 M steps/s, Step 36702500, log_posterior = -45917.21, log_G = -47462.25, log_coal = -2682.60, log_other_priors = 4227.64, num_muts = 1339, T = 121434.09, t_MRCA = 1939-01-18, n0 = 6109.9, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 319, kappa = 5.99, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4050, 3759, 5104], num_muts_ab counts: [68, 261, 38, 34, 18, 239, 246, 44, 41, 34, 274, 42]\n",
            "0.5 M steps/s, Step 37395000, log_posterior = -45955.38, log_G = -47494.78, log_coal = -2688.07, log_other_priors = 4227.47, num_muts = 1340, T = 123006.96, t_MRCA = 1938-12-11, n0 = 6168.6, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 320, kappa = 6.65, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4050, 3759, 5104], num_muts_ab counts: [67, 261, 39, 35, 18, 239, 245, 44, 41, 34, 275, 42]\n",
            "0.4 M steps/s, Step 38087500, log_posterior = -45971.55, log_G = -47502.16, log_coal = -2696.97, log_other_priors = 4227.57, num_muts = 1339, T = 124919.02, t_MRCA = 1937-03-03, n0 = 6573.1, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 319, kappa = 5.97, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4050, 3759, 5104], num_muts_ab counts: [68, 261, 38, 34, 18, 239, 246, 44, 41, 34, 274, 42]\n",
            "0.4 M steps/s, Step 38780000, log_posterior = -45987.88, log_G = -47494.57, log_coal = -2720.49, log_other_priors = 4227.19, num_muts = 1339, T = 124361.06, t_MRCA = 1940-01-27, n0 = 8120.3, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 320, kappa = 6.69, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6047, 4050, 3759, 5104], num_muts_ab counts: [68, 260, 38, 34, 19, 239, 246, 44, 41, 34, 274, 42]\n",
            "0.4 M steps/s, Step 39472500, log_posterior = -45973.82, log_G = -47493.07, log_coal = -2708.24, log_other_priors = 4227.49, num_muts = 1339, T = 124163.66, t_MRCA = 1940-10-30, n0 = 6477.0, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 320, kappa = 6.37, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4055, 3751, 5097], num_muts_ab counts: [67, 269, 40, 36, 19, 242, 236, 43, 43, 31, 272, 41]\n",
            "0.4 M steps/s, Step 40165000, log_posterior = -45864.47, log_G = -47406.35, log_coal = -2685.53, log_other_priors = 4227.41, num_muts = 1337, T = 118680.15, t_MRCA = 1942-07-14, n0 = 7531.4, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 321, kappa = 6.06, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4055, 3751, 5097], num_muts_ab counts: [67, 269, 41, 37, 20, 240, 235, 43, 42, 30, 272, 41]\n",
            "0.4 M steps/s, Step 40857500, log_posterior = -45913.48, log_G = -47443.20, log_coal = -2697.63, log_other_priors = 4227.35, num_muts = 1337, T = 122137.41, t_MRCA = 1940-04-03, n0 = 6872.0, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 320, kappa = 6.72, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6057, 4055, 3751, 5097], num_muts_ab counts: [67, 269, 40, 36, 20, 240, 235, 43, 43, 30, 273, 41]\n",
            "0.4 M steps/s, Step 41550000, log_posterior = -45849.29, log_G = -47403.33, log_coal = -2673.49, log_other_priors = 4227.54, num_muts = 1337, T = 118578.07, t_MRCA = 1939-10-18, n0 = 5538.2, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 6.83, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6037, 4065, 3759, 5099], num_muts_ab counts: [60, 261, 37, 40, 20, 243, 244, 42, 42, 39, 270, 39]\n",
            "0.4 M steps/s, Step 42242500, log_posterior = -45839.41, log_G = -47396.64, log_coal = -2670.31, log_other_priors = 4227.54, num_muts = 1336, T = 119017.89, t_MRCA = 1939-04-07, n0 = 5674.1, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1014, M_tv = 322, kappa = 6.73, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6037, 4065, 3759, 5099], num_muts_ab counts: [60, 260, 36, 41, 22, 243, 242, 42, 43, 39, 269, 39]\n",
            "0.4 M steps/s, Step 42935000, log_posterior = -45965.87, log_G = -47477.24, log_coal = -2715.96, log_other_priors = 4227.33, num_muts = 1336, T = 126236.90, t_MRCA = 1939-08-14, n0 = 7000.5, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1013, M_tv = 323, kappa = 6.72, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6037, 4065, 3759, 5099], num_muts_ab counts: [60, 259, 36, 41, 23, 243, 242, 42, 43, 39, 269, 39]\n",
            "0.5 M steps/s, Step 43627500, log_posterior = -45846.97, log_G = -47408.60, log_coal = -2666.12, log_other_priors = 4227.74, num_muts = 1337, T = 119093.12, t_MRCA = 1938-09-16, n0 = 5511.8, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 320, kappa = 6.01, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6037, 4065, 3759, 5099], num_muts_ab counts: [60, 261, 36, 41, 20, 243, 243, 42, 43, 39, 270, 39]\n",
            "0.5 M steps/s, Step 44320000, log_posterior = -45909.98, log_G = -47443.26, log_coal = -2693.96, log_other_priors = 4227.25, num_muts = 1337, T = 122343.06, t_MRCA = 1940-06-04, n0 = 6929.0, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 7.12, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6037, 4065, 3759, 5099], num_muts_ab counts: [60, 262, 36, 41, 19, 243, 243, 42, 43, 39, 270, 39]\n",
            "0.5 M steps/s, Step 45012500, log_posterior = -45855.33, log_G = -47411.97, log_coal = -2670.80, log_other_priors = 4227.43, num_muts = 1337, T = 119181.38, t_MRCA = 1940-03-31, n0 = 5431.5, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 320, kappa = 7.36, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6037, 4065, 3759, 5099], num_muts_ab counts: [60, 261, 36, 41, 20, 243, 243, 42, 43, 39, 270, 39]\n",
            "0.4 M steps/s, Step 45705000, log_posterior = -45910.61, log_G = -47452.76, log_coal = -2685.39, log_other_priors = 4227.54, num_muts = 1340, T = 119033.75, t_MRCA = 1940-09-18, n0 = 6473.7, g = 0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 321, kappa = 6.17, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6066, 4052, 3745, 5097], num_muts_ab counts: [65, 280, 43, 35, 18, 236, 228, 45, 44, 33, 275, 38]\n",
            "0.4 M steps/s, Step 46397500, log_posterior = -45672.67, log_G = -47234.07, log_coal = -2666.24, log_other_priors = 4227.64, num_muts = 1326, T = 118365.75, t_MRCA = 1939-09-08, n0 = 5820.5, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1011, M_tv = 315, kappa = 6.2, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6049, 4052, 3751, 5108], num_muts_ab counts: [63, 267, 36, 41, 18, 232, 232, 42, 41, 34, 280, 40]\n",
            "0.4 M steps/s, Step 47090000, log_posterior = -45688.64, log_G = -47244.49, log_coal = -2671.79, log_other_priors = 4227.64, num_muts = 1326, T = 119410.88, t_MRCA = 1939-12-04, n0 = 6112.8, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1010, M_tv = 316, kappa = 5.98, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6049, 4052, 3751, 5108], num_muts_ab counts: [63, 266, 36, 41, 19, 232, 232, 42, 41, 34, 280, 40]\n",
            "0.4 M steps/s, Step 47782500, log_posterior = -45703.49, log_G = -47263.71, log_coal = -2667.31, log_other_priors = 4227.53, num_muts = 1327, T = 119705.89, t_MRCA = 1939-06-25, n0 = 5867.2, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1011, M_tv = 316, kappa = 6.62, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6049, 4052, 3751, 5108], num_muts_ab counts: [63, 265, 37, 40, 20, 232, 234, 42, 40, 34, 280, 40]\n",
            "0.4 M steps/s, Step 48475000, log_posterior = -45872.80, log_G = -47414.11, log_coal = -2686.41, log_other_priors = 4227.71, num_muts = 1331, T = 127626.88, t_MRCA = 1933-11-10, n0 = 5904.5, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1012, M_tv = 319, kappa = 5.84, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6061, 4045, 3753, 5101], num_muts_ab counts: [69, 271, 38, 36, 20, 229, 231, 42, 48, 31, 281, 35]\n",
            "0.4 M steps/s, Step 49167500, log_posterior = -45937.66, log_G = -47453.98, log_coal = -2710.92, log_other_priors = 4227.24, num_muts = 1332, T = 130607.04, t_MRCA = 1934-04-14, n0 = 7477.5, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1014, M_tv = 318, kappa = 6.83, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6051, 4056, 3746, 5107], num_muts_ab counts: [57, 273, 41, 46, 20, 228, 232, 42, 40, 31, 281, 41]\n",
            "0.4 M steps/s, Step 49860000, log_posterior = -45832.34, log_G = -47407.98, log_coal = -2651.86, log_other_priors = 4227.50, num_muts = 1337, T = 119298.17, t_MRCA = 1935-07-13, n0 = 6190.4, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 6.49, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [66, 269, 37, 35, 18, 258, 235, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 50552500, log_posterior = -45952.26, log_G = -47484.06, log_coal = -2695.58, log_other_priors = 4227.39, num_muts = 1338, T = 125130.35, t_MRCA = 1936-01-31, n0 = 7027.8, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1021, M_tv = 317, kappa = 6.46, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [65, 269, 38, 34, 18, 259, 236, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 51245000, log_posterior = -45918.96, log_G = -47457.48, log_coal = -2688.95, log_other_priors = 4227.47, num_muts = 1337, T = 123897.06, t_MRCA = 1934-09-19, n0 = 5828.7, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 6.89, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [66, 268, 37, 34, 19, 259, 237, 46, 39, 34, 255, 43]\n",
            "0.4 M steps/s, Step 51937500, log_posterior = -45907.93, log_G = -47465.32, log_coal = -2670.09, log_other_priors = 4227.48, num_muts = 1338, T = 122621.76, t_MRCA = 1935-08-06, n0 = 5772.6, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 320, kappa = 6.91, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [66, 268, 38, 35, 18, 258, 236, 46, 39, 34, 256, 44]\n",
            "0.4 M steps/s, Step 52630000, log_posterior = -46080.00, log_G = -47565.86, log_coal = -2741.30, log_other_priors = 4227.16, num_muts = 1337, T = 134478.58, t_MRCA = 1934-10-08, n0 = 7528.5, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 317, kappa = 7.15, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [66, 269, 37, 34, 18, 258, 237, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 53322500, log_posterior = -45973.46, log_G = -47491.13, log_coal = -2709.59, log_other_priors = 4227.26, num_muts = 1337, T = 126831.84, t_MRCA = 1935-05-26, n0 = 6112.9, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 317, kappa = 7.61, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [66, 269, 37, 33, 18, 258, 237, 46, 39, 35, 256, 43]\n",
            "0.4 M steps/s, Step 54015000, log_posterior = -45948.80, log_G = -47478.31, log_coal = -2698.02, log_other_priors = 4227.53, num_muts = 1337, T = 125871.43, t_MRCA = 1935-02-08, n0 = 5578.3, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 317, kappa = 6.83, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [66, 269, 37, 34, 18, 258, 237, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 54707500, log_posterior = -45927.18, log_G = -47467.32, log_coal = -2687.30, log_other_priors = 4227.44, num_muts = 1337, T = 124607.08, t_MRCA = 1934-11-30, n0 = 6814.2, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 6.37, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [66, 269, 38, 35, 18, 258, 236, 46, 38, 34, 256, 43]\n",
            "0.4 M steps/s, Step 55400000, log_posterior = -46004.91, log_G = -47524.58, log_coal = -2707.59, log_other_priors = 4227.27, num_muts = 1337, T = 130347.07, t_MRCA = 1933-10-18, n0 = 7695.6, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 317, kappa = 6.57, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [66, 269, 37, 34, 18, 258, 237, 46, 39, 34, 256, 43]\n",
            "0.5 M steps/s, Step 56092500, log_posterior = -45939.04, log_G = -47482.16, log_coal = -2684.56, log_other_priors = 4227.67, num_muts = 1337, T = 126280.95, t_MRCA = 1932-08-25, n0 = 5377.6, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 317, kappa = 6.37, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [65, 269, 37, 35, 18, 258, 237, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 56785000, log_posterior = -45968.94, log_G = -47498.04, log_coal = -2698.39, log_other_priors = 4227.48, num_muts = 1337, T = 127688.37, t_MRCA = 1933-01-21, n0 = 6229.7, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 6.56, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [65, 269, 37, 36, 18, 258, 235, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 57477500, log_posterior = -45979.13, log_G = -47512.29, log_coal = -2694.21, log_other_priors = 4227.37, num_muts = 1338, T = 127471.72, t_MRCA = 1933-06-09, n0 = 6392.9, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1020, M_tv = 318, kappa = 6.95, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [65, 268, 37, 36, 19, 259, 235, 46, 38, 34, 258, 43]\n",
            "0.4 M steps/s, Step 58170000, log_posterior = -45954.00, log_G = -47484.21, log_coal = -2697.38, log_other_priors = 4227.60, num_muts = 1337, T = 125637.17, t_MRCA = 1936-03-20, n0 = 5436.8, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 6.65, pi = [0.31, 0.21, 0.2, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 37, 34, 20, 259, 236, 46, 38, 35, 257, 44]\n",
            "0.4 M steps/s, Step 58862500, log_posterior = -45945.38, log_G = -47466.01, log_coal = -2706.83, log_other_priors = 4227.46, num_muts = 1336, T = 125645.30, t_MRCA = 1936-12-08, n0 = 6786.1, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.29, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 22, 259, 234, 46, 38, 35, 257, 43]\n",
            "0.4 M steps/s, Step 59555000, log_posterior = -45893.49, log_G = -47439.94, log_coal = -2681.06, log_other_priors = 4227.52, num_muts = 1336, T = 123279.15, t_MRCA = 1935-10-23, n0 = 6264.9, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.39, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 34, 22, 259, 235, 46, 38, 35, 257, 43]\n",
            "0.4 M steps/s, Step 60247500, log_posterior = -45819.34, log_G = -47391.02, log_coal = -2655.68, log_other_priors = 4227.37, num_muts = 1335, T = 119933.91, t_MRCA = 1937-04-07, n0 = 5550.1, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 7.56, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 34, 21, 258, 235, 46, 39, 35, 257, 43]\n",
            "0.4 M steps/s, Step 60940000, log_posterior = -45878.14, log_G = -47416.20, log_coal = -2689.42, log_other_priors = 4227.49, num_muts = 1335, T = 122312.43, t_MRCA = 1936-11-17, n0 = 6110.1, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1014, M_tv = 321, kappa = 6.6, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 265, 37, 36, 20, 258, 234, 46, 39, 35, 257, 44]\n",
            "0.4 M steps/s, Step 61632500, log_posterior = -45924.92, log_G = -47449.13, log_coal = -2703.36, log_other_priors = 4227.57, num_muts = 1335, T = 125273.39, t_MRCA = 1937-11-18, n0 = 6151.2, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.26, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 36, 20, 258, 234, 46, 39, 35, 257, 43]\n",
            "0.4 M steps/s, Step 62325000, log_posterior = -45834.21, log_G = -47393.28, log_coal = -2668.50, log_other_priors = 4227.57, num_muts = 1335, T = 120373.70, t_MRCA = 1935-11-30, n0 = 5929.5, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.38, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 36, 20, 258, 234, 46, 39, 35, 257, 43]\n",
            "0.4 M steps/s, Step 63017500, log_posterior = -45881.54, log_G = -47430.89, log_coal = -2678.12, log_other_priors = 4227.47, num_muts = 1335, T = 123985.09, t_MRCA = 1935-06-14, n0 = 6647.7, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.35, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 19, 258, 235, 46, 39, 35, 257, 44]\n",
            "0.4 M steps/s, Step 63710000, log_posterior = -45962.27, log_G = -47479.35, log_coal = -2710.23, log_other_priors = 4227.31, num_muts = 1335, T = 128358.98, t_MRCA = 1934-02-14, n0 = 6958.5, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.83, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 265, 37, 35, 20, 258, 235, 46, 39, 35, 257, 44]\n",
            "0.4 M steps/s, Step 64402500, log_posterior = -45912.23, log_G = -47450.24, log_coal = -2689.33, log_other_priors = 4227.35, num_muts = 1335, T = 125810.45, t_MRCA = 1932-11-09, n0 = 7125.3, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.56, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 36, 19, 258, 235, 46, 39, 34, 257, 44]\n",
            "0.4 M steps/s, Step 65095000, log_posterior = -45910.89, log_G = -47450.03, log_coal = -2688.20, log_other_priors = 4227.34, num_muts = 1335, T = 125574.51, t_MRCA = 1932-07-25, n0 = 5942.8, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 7.4, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 35, 19, 258, 235, 46, 38, 35, 257, 44]\n",
            "0.4 M steps/s, Step 65787500, log_posterior = -46059.93, log_G = -47553.53, log_coal = -2733.54, log_other_priors = 4227.14, num_muts = 1335, T = 135532.39, t_MRCA = 1930-08-14, n0 = 7204.7, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 7.42, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 265, 37, 35, 20, 258, 235, 46, 39, 35, 257, 44]\n",
            "0.5 M steps/s, Step 66480000, log_posterior = -45940.34, log_G = -47482.92, log_coal = -2684.94, log_other_priors = 4227.52, num_muts = 1335, T = 128452.60, t_MRCA = 1931-05-28, n0 = 6020.3, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1014, M_tv = 321, kappa = 6.57, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [65, 266, 37, 35, 20, 258, 234, 46, 39, 36, 256, 43]\n",
            "0.5 M steps/s, Step 67172500, log_posterior = -46025.32, log_G = -47531.54, log_coal = -2721.30, log_other_priors = 4227.52, num_muts = 1335, T = 133322.76, t_MRCA = 1931-10-23, n0 = 6782.5, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.07, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 258, 235, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 67865000, log_posterior = -45991.50, log_G = -47521.91, log_coal = -2697.08, log_other_priors = 4227.50, num_muts = 1337, T = 129593.41, t_MRCA = 1931-10-13, n0 = 6849.2, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 320, kappa = 6.11, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 259, 235, 46, 39, 35, 257, 43]\n",
            "0.5 M steps/s, Step 68557500, log_posterior = -46062.53, log_G = -47557.16, log_coal = -2732.55, log_other_priors = 4227.17, num_muts = 1336, T = 134242.33, t_MRCA = 1931-01-19, n0 = 7575.4, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 7.06, pi = [0.32, 0.22, 0.19, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 35, 21, 258, 236, 46, 38, 35, 256, 43]\n",
            "0.5 M steps/s, Step 69250000, log_posterior = -45940.72, log_G = -47484.61, log_coal = -2683.78, log_other_priors = 4227.67, num_muts = 1336, T = 127656.98, t_MRCA = 1932-01-28, n0 = 5560.7, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.24, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 34, 21, 258, 237, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 69942500, log_posterior = -45877.64, log_G = -47433.78, log_coal = -2671.64, log_other_priors = 4227.78, num_muts = 1335, T = 124256.99, t_MRCA = 1932-06-12, n0 = 4803.9, g = -0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.4, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 34, 21, 258, 236, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 70635000, log_posterior = -45867.61, log_G = -47435.46, log_coal = -2659.60, log_other_priors = 4227.44, num_muts = 1335, T = 124065.82, t_MRCA = 1932-01-27, n0 = 5411.8, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 7.33, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 258, 235, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 71327500, log_posterior = -45941.60, log_G = -47477.83, log_coal = -2691.19, log_other_priors = 4227.42, num_muts = 1335, T = 127634.47, t_MRCA = 1932-07-28, n0 = 5581.8, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 7.29, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 258, 235, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 72020000, log_posterior = -45877.06, log_G = -47443.34, log_coal = -2661.16, log_other_priors = 4227.43, num_muts = 1335, T = 125091.38, t_MRCA = 1931-06-05, n0 = 6007.9, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.93, pi = [0.32, 0.21, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 37, 34, 20, 258, 236, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 72712500, log_posterior = -45920.36, log_G = -47460.11, log_coal = -2687.61, log_other_priors = 4227.36, num_muts = 1335, T = 126230.39, t_MRCA = 1933-02-16, n0 = 6553.2, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.87, pi = [0.33, 0.21, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 34, 21, 258, 236, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 73405000, log_posterior = -45859.92, log_G = -47423.47, log_coal = -2664.13, log_other_priors = 4227.68, num_muts = 1335, T = 123062.03, t_MRCA = 1931-06-14, n0 = 4835.0, g = -0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.78, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 258, 235, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 74097500, log_posterior = -45899.49, log_G = -47453.02, log_coal = -2674.23, log_other_priors = 4227.77, num_muts = 1335, T = 125938.29, t_MRCA = 1930-12-24, n0 = 5418.7, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 5.97, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 34, 21, 258, 236, 46, 38, 35, 256, 43]\n",
            "0.5 M steps/s, Step 74790000, log_posterior = -45963.12, log_G = -47493.23, log_coal = -2697.03, log_other_priors = 4227.15, num_muts = 1335, T = 129451.56, t_MRCA = 1931-08-10, n0 = 7050.5, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 7.49, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 258, 235, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 75482500, log_posterior = -45888.46, log_G = -47448.28, log_coal = -2667.64, log_other_priors = 4227.46, num_muts = 1336, T = 124038.10, t_MRCA = 1931-11-19, n0 = 6139.8, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 6.73, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 35, 19, 258, 236, 46, 38, 35, 256, 43]\n",
            "0.5 M steps/s, Step 76175000, log_posterior = -46034.01, log_G = -47541.33, log_coal = -2719.99, log_other_priors = 4227.32, num_muts = 1337, T = 131536.64, t_MRCA = 1932-10-08, n0 = 7637.9, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 320, kappa = 6.41, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 259, 236, 46, 39, 35, 256, 43]\n",
            "0.5 M steps/s, Step 76867500, log_posterior = -45943.11, log_G = -47485.97, log_coal = -2684.83, log_other_priors = 4227.69, num_muts = 1336, T = 127558.66, t_MRCA = 1932-08-14, n0 = 5071.9, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.56, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 36, 21, 259, 235, 46, 39, 34, 256, 43]\n",
            "0.5 M steps/s, Step 77560000, log_posterior = -45779.45, log_G = -47390.30, log_coal = -2616.77, log_other_priors = 4227.63, num_muts = 1336, T = 118788.72, t_MRCA = 1932-09-21, n0 = 5537.1, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.45, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 259, 236, 46, 39, 34, 256, 43]\n",
            "0.5 M steps/s, Step 78252500, log_posterior = -45855.10, log_G = -47426.64, log_coal = -2656.16, log_other_priors = 4227.71, num_muts = 1336, T = 122232.25, t_MRCA = 1934-01-26, n0 = 4875.6, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.65, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 35, 21, 259, 236, 46, 39, 34, 256, 43]\n",
            "0.5 M steps/s, Step 78945000, log_posterior = -45799.82, log_G = -47389.61, log_coal = -2637.93, log_other_priors = 4227.72, num_muts = 1336, T = 118748.12, t_MRCA = 1934-08-20, n0 = 4444.0, g = -0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.98, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 37, 37, 20, 259, 235, 46, 39, 33, 256, 43]\n",
            "0.4 M steps/s, Step 79637500, log_posterior = -45937.98, log_G = -47481.29, log_coal = -2684.51, log_other_priors = 4227.82, num_muts = 1337, T = 125422.07, t_MRCA = 1934-03-15, n0 = 5671.7, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 5.57, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 37, 35, 20, 259, 236, 46, 40, 34, 256, 43]\n",
            "0.5 M steps/s, Step 80330000, log_posterior = -45907.12, log_G = -47453.86, log_coal = -2680.56, log_other_priors = 4227.30, num_muts = 1336, T = 124400.98, t_MRCA = 1934-11-20, n0 = 6899.4, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.9, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 36, 21, 259, 235, 46, 39, 34, 256, 43]\n",
            "0.5 M steps/s, Step 81022500, log_posterior = -45896.91, log_G = -47452.55, log_coal = -2672.10, log_other_priors = 4227.74, num_muts = 1336, T = 124322.68, t_MRCA = 1934-11-01, n0 = 5635.0, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 5.9, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 37, 36, 20, 259, 235, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 81715000, log_posterior = -45923.64, log_G = -47460.84, log_coal = -2690.15, log_other_priors = 4227.35, num_muts = 1336, T = 125331.00, t_MRCA = 1936-02-27, n0 = 6427.8, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 7.01, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 37, 37, 20, 259, 235, 46, 39, 33, 256, 43]\n",
            "0.4 M steps/s, Step 82407500, log_posterior = -45821.75, log_G = -47397.11, log_coal = -2652.18, log_other_priors = 4227.53, num_muts = 1337, T = 117949.85, t_MRCA = 1938-10-22, n0 = 5893.7, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 318, kappa = 6.58, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 37, 35, 20, 259, 236, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 83100000, log_posterior = -45865.38, log_G = -47408.99, log_coal = -2683.89, log_other_priors = 4227.50, num_muts = 1335, T = 121681.21, t_MRCA = 1939-03-16, n0 = 7022.6, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 37, 21, 259, 234, 46, 39, 33, 256, 43]\n",
            "0.4 M steps/s, Step 83792500, log_posterior = -45893.39, log_G = -47425.71, log_coal = -2695.12, log_other_priors = 4227.44, num_muts = 1335, T = 123503.63, t_MRCA = 1939-03-11, n0 = 6074.7, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.83, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 37, 37, 19, 259, 234, 46, 39, 33, 256, 43]\n",
            "0.4 M steps/s, Step 84485000, log_posterior = -45814.16, log_G = -47377.56, log_coal = -2664.60, log_other_priors = 4228.00, num_muts = 1335, T = 118989.35, t_MRCA = 1938-06-16, n0 = 4513.8, g = -0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 5.76, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 37, 36, 19, 259, 234, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 85177500, log_posterior = -45845.70, log_G = -47398.57, log_coal = -2674.83, log_other_priors = 4227.69, num_muts = 1335, T = 120900.23, t_MRCA = 1938-07-17, n0 = 5187.5, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.44, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 37, 36, 20, 259, 234, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 85870000, log_posterior = -45782.91, log_G = -47358.27, log_coal = -2652.04, log_other_priors = 4227.40, num_muts = 1336, T = 116268.91, t_MRCA = 1939-04-18, n0 = 5835.6, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 317, kappa = 7.21, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 37, 34, 20, 260, 236, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 86562500, log_posterior = -45879.20, log_G = -47428.48, log_coal = -2678.37, log_other_priors = 4227.65, num_muts = 1336, T = 122075.16, t_MRCA = 1938-10-12, n0 = 5712.7, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.21, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 37, 36, 21, 259, 235, 46, 39, 34, 256, 43]\n",
            "0.4 M steps/s, Step 87255000, log_posterior = -45800.73, log_G = -47380.49, log_coal = -2647.71, log_other_priors = 4227.47, num_muts = 1336, T = 117805.36, t_MRCA = 1937-12-27, n0 = 5401.8, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 7.22, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 20, 259, 233, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 87947500, log_posterior = -45838.03, log_G = -47397.81, log_coal = -2667.69, log_other_priors = 4227.46, num_muts = 1335, T = 120895.12, t_MRCA = 1939-01-18, n0 = 6452.9, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.49, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 20, 258, 234, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 88640000, log_posterior = -45812.62, log_G = -47376.50, log_coal = -2663.35, log_other_priors = 4227.23, num_muts = 1335, T = 119183.33, t_MRCA = 1940-04-02, n0 = 6497.1, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 317, kappa = 7.47, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 34, 19, 258, 236, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 89332500, log_posterior = -45846.92, log_G = -47403.53, log_coal = -2671.09, log_other_priors = 4227.71, num_muts = 1336, T = 119884.92, t_MRCA = 1939-02-22, n0 = 5684.7, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.02, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 39, 36, 19, 257, 235, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 90025000, log_posterior = -45868.57, log_G = -47429.58, log_coal = -2666.45, log_other_priors = 4227.47, num_muts = 1336, T = 122438.37, t_MRCA = 1936-12-26, n0 = 5924.3, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.83, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 265, 38, 36, 20, 260, 234, 45, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 90717500, log_posterior = -45914.10, log_G = -47449.09, log_coal = -2692.68, log_other_priors = 4227.67, num_muts = 1337, T = 122696.68, t_MRCA = 1937-03-31, n0 = 6274.5, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 5.77, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 265, 38, 37, 20, 260, 235, 45, 39, 33, 258, 43]\n",
            "0.4 M steps/s, Step 91410000, log_posterior = -45859.07, log_G = -47414.97, log_coal = -2671.46, log_other_priors = 4227.36, num_muts = 1336, T = 121358.76, t_MRCA = 1936-10-19, n0 = 6825.4, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 6.67, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 36, 19, 260, 234, 45, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 92102500, log_posterior = -45873.34, log_G = -47432.80, log_coal = -2668.02, log_other_priors = 4227.48, num_muts = 1336, T = 122641.09, t_MRCA = 1934-07-12, n0 = 5369.3, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 7.22, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 265, 38, 36, 20, 260, 234, 45, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 92795000, log_posterior = -45827.20, log_G = -47396.04, log_coal = -2658.86, log_other_priors = 4227.70, num_muts = 1335, T = 120912.11, t_MRCA = 1934-07-18, n0 = 5454.1, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.19, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 36, 18, 258, 234, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 93487500, log_posterior = -45856.85, log_G = -47413.26, log_coal = -2671.23, log_other_priors = 4227.63, num_muts = 1335, T = 121936.90, t_MRCA = 1935-04-26, n0 = 5942.1, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.15, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 36, 20, 258, 234, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 94180000, log_posterior = -45837.68, log_G = -47401.70, log_coal = -2663.46, log_other_priors = 4227.48, num_muts = 1335, T = 121205.05, t_MRCA = 1936-09-25, n0 = 5700.1, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.95, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 35, 20, 257, 236, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 94872500, log_posterior = -45835.27, log_G = -47403.89, log_coal = -2659.21, log_other_priors = 4227.83, num_muts = 1336, T = 119685.52, t_MRCA = 1936-04-03, n0 = 5341.3, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 5.76, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 258, 236, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 95565000, log_posterior = -45839.82, log_G = -47411.15, log_coal = -2656.34, log_other_priors = 4227.67, num_muts = 1335, T = 121562.95, t_MRCA = 1934-03-15, n0 = 5328.6, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.44, pi = [0.31, 0.21, 0.2, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 36, 20, 257, 235, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 96257500, log_posterior = -45952.39, log_G = -47474.74, log_coal = -2704.97, log_other_priors = 4227.32, num_muts = 1335, T = 127946.96, t_MRCA = 1935-01-24, n0 = 6233.5, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 7.25, pi = [0.31, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 96950000, log_posterior = -45884.09, log_G = -47436.16, log_coal = -2675.33, log_other_priors = 4227.40, num_muts = 1335, T = 124662.02, t_MRCA = 1935-12-04, n0 = 6181.3, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 317, kappa = 6.94, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [63, 268, 38, 36, 18, 257, 234, 46, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 97642500, log_posterior = -45848.07, log_G = -47408.21, log_coal = -2667.17, log_other_priors = 4227.31, num_muts = 1335, T = 121951.53, t_MRCA = 1937-01-02, n0 = 5814.1, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 317, kappa = 7.64, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [63, 268, 38, 36, 18, 257, 234, 46, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 98335000, log_posterior = -45818.00, log_G = -47390.80, log_coal = -2654.75, log_other_priors = 4227.55, num_muts = 1335, T = 120251.80, t_MRCA = 1937-06-27, n0 = 5439.6, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.84, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [63, 266, 38, 36, 20, 257, 234, 46, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 99027500, log_posterior = -45840.24, log_G = -47405.12, log_coal = -2663.10, log_other_priors = 4227.98, num_muts = 1336, T = 120174.01, t_MRCA = 1937-02-13, n0 = 4538.1, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 5.84, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [63, 267, 38, 36, 19, 257, 235, 46, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 99720000, log_posterior = -45858.91, log_G = -47409.44, log_coal = -2677.16, log_other_priors = 4227.69, num_muts = 1335, T = 122301.28, t_MRCA = 1937-01-31, n0 = 5256.0, g = -0.0001, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 317, kappa = 6.39, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [63, 268, 38, 36, 18, 257, 234, 46, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 100412500, log_posterior = -45781.22, log_G = -47362.35, log_coal = -2646.45, log_other_priors = 4227.58, num_muts = 1335, T = 117482.35, t_MRCA = 1937-03-16, n0 = 5333.4, g = -0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.79, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [63, 266, 38, 36, 20, 257, 234, 46, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 101105000, log_posterior = -45843.11, log_G = -47407.19, log_coal = -2663.68, log_other_priors = 4227.75, num_muts = 1335, T = 121406.56, t_MRCA = 1936-09-30, n0 = 5461.2, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [63, 266, 38, 36, 20, 257, 234, 46, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 101797500, log_posterior = -45853.88, log_G = -47416.98, log_coal = -2664.77, log_other_priors = 4227.87, num_muts = 1336, T = 121316.39, t_MRCA = 1936-06-14, n0 = 4657.8, g = -0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.16, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 258, 235, 46, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 102490000, log_posterior = -45869.40, log_G = -47428.37, log_coal = -2668.56, log_other_priors = 4227.53, num_muts = 1335, T = 123722.67, t_MRCA = 1935-05-30, n0 = 5558.9, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.84, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 103182500, log_posterior = -45927.06, log_G = -47455.39, log_coal = -2699.03, log_other_priors = 4227.37, num_muts = 1335, T = 126133.79, t_MRCA = 1935-05-18, n0 = 6451.3, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.9, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 103875000, log_posterior = -45952.54, log_G = -47480.65, log_coal = -2699.27, log_other_priors = 4227.38, num_muts = 1335, T = 128573.22, t_MRCA = 1933-05-11, n0 = 7420.1, g = 0.0001, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.25, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 104567500, log_posterior = -45929.70, log_G = -47477.57, log_coal = -2679.52, log_other_priors = 4227.39, num_muts = 1335, T = 127939.35, t_MRCA = 1932-01-18, n0 = 6078.0, g = -0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 7.05, pi = [0.32, 0.21, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 105260000, log_posterior = -45872.73, log_G = -47443.39, log_coal = -2657.04, log_other_priors = 4227.70, num_muts = 1335, T = 124978.09, t_MRCA = 1930-07-22, n0 = 4682.1, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.84, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 36, 20, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 105952500, log_posterior = -46009.63, log_G = -47527.58, log_coal = -2709.16, log_other_priors = 4227.11, num_muts = 1335, T = 133276.97, t_MRCA = 1929-05-11, n0 = 7782.0, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 7.22, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 36, 18, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 106645000, log_posterior = -45963.78, log_G = -47496.41, log_coal = -2694.73, log_other_priors = 4227.36, num_muts = 1336, T = 128627.06, t_MRCA = 1931-06-02, n0 = 6618.1, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.82, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 39, 36, 18, 256, 234, 46, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 107337500, log_posterior = -46009.49, log_G = -47519.28, log_coal = -2717.55, log_other_priors = 4227.35, num_muts = 1335, T = 132307.16, t_MRCA = 1931-06-04, n0 = 7848.0, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.17, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 108030000, log_posterior = -45952.03, log_G = -47483.74, log_coal = -2695.76, log_other_priors = 4227.48, num_muts = 1335, T = 128668.24, t_MRCA = 1931-09-27, n0 = 6364.1, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.49, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 108722500, log_posterior = -45947.97, log_G = -47479.28, log_coal = -2696.16, log_other_priors = 4227.47, num_muts = 1335, T = 128527.67, t_MRCA = 1932-08-14, n0 = 6590.2, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.37, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 235, 47, 39, 34, 257, 43]\n",
            "0.5 M steps/s, Step 109415000, log_posterior = -45866.05, log_G = -47425.97, log_coal = -2667.75, log_other_priors = 4227.68, num_muts = 1335, T = 123472.59, t_MRCA = 1933-02-01, n0 = 5791.8, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.07, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 34, 19, 257, 236, 47, 39, 34, 257, 43]\n",
            "0.5 M steps/s, Step 110107500, log_posterior = -45950.23, log_G = -47485.75, log_coal = -2692.13, log_other_priors = 4227.64, num_muts = 1336, T = 127304.33, t_MRCA = 1933-03-01, n0 = 5295.9, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.56, pi = [0.32, 0.21, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 47, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 110800000, log_posterior = -45845.31, log_G = -47421.32, log_coal = -2651.60, log_other_priors = 4227.62, num_muts = 1336, T = 120931.31, t_MRCA = 1935-07-01, n0 = 5261.2, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 321, kappa = 6.71, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 39, 37, 19, 258, 233, 47, 39, 33, 257, 43]\n",
            "0.4 M steps/s, Step 111492500, log_posterior = -45950.00, log_G = -47484.41, log_coal = -2693.02, log_other_priors = 4227.43, num_muts = 1336, T = 126696.72, t_MRCA = 1936-05-02, n0 = 7215.2, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1014, M_tv = 322, kappa = 6.19, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 39, 37, 20, 257, 234, 47, 39, 33, 257, 43]\n",
            "0.4 M steps/s, Step 112185000, log_posterior = -45899.52, log_G = -47441.86, log_coal = -2685.16, log_other_priors = 4227.50, num_muts = 1335, T = 123847.71, t_MRCA = 1936-06-23, n0 = 5387.9, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1013, M_tv = 322, kappa = 7.12, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 39, 37, 20, 257, 233, 47, 39, 33, 257, 43]\n",
            "0.4 M steps/s, Step 112877500, log_posterior = -45863.33, log_G = -47418.37, log_coal = -2672.44, log_other_priors = 4227.49, num_muts = 1335, T = 122204.91, t_MRCA = 1937-06-04, n0 = 6295.3, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.49, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 39, 37, 19, 257, 233, 46, 39, 33, 258, 43]\n",
            "0.4 M steps/s, Step 113570000, log_posterior = -45848.20, log_G = -47404.48, log_coal = -2671.40, log_other_priors = 4227.69, num_muts = 1336, T = 119918.74, t_MRCA = 1938-05-07, n0 = 5861.9, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 5.97, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 40, 36, 19, 257, 234, 45, 38, 33, 259, 44]\n",
            "0.4 M steps/s, Step 114262500, log_posterior = -45906.66, log_G = -47440.55, log_coal = -2693.73, log_other_priors = 4227.62, num_muts = 1336, T = 122781.33, t_MRCA = 1939-11-18, n0 = 5940.1, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 321, kappa = 6.2, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 39, 37, 20, 257, 233, 45, 39, 33, 259, 44]\n",
            "0.4 M steps/s, Step 114955000, log_posterior = -45754.97, log_G = -47347.15, log_coal = -2635.56, log_other_priors = 4227.75, num_muts = 1337, T = 113778.92, t_MRCA = 1941-12-21, n0 = 4867.2, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 6.48, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6053, 4070, 3752, 5085], num_muts_ab counts: [64, 268, 40, 37, 18, 257, 234, 45, 38, 33, 259, 44]\n",
            "0.4 M steps/s, Step 115647500, log_posterior = -45761.79, log_G = -47343.66, log_coal = -2645.60, log_other_priors = 4227.46, num_muts = 1336, T = 114397.64, t_MRCA = 1942-08-07, n0 = 5542.9, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 321, kappa = 7.14, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 39, 37, 20, 257, 233, 45, 39, 33, 259, 44]\n",
            "0.4 M steps/s, Step 116340000, log_posterior = -45866.15, log_G = -47422.95, log_coal = -2670.82, log_other_priors = 4227.62, num_muts = 1338, T = 118458.94, t_MRCA = 1941-09-08, n0 = 5872.2, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 321, kappa = 6.22, pi = [0.31, 0.22, 0.2, 0.28], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 39, 37, 20, 259, 233, 45, 39, 33, 259, 44]\n",
            "0.4 M steps/s, Step 117032500, log_posterior = -45884.21, log_G = -47421.97, log_coal = -2689.57, log_other_priors = 4227.33, num_muts = 1336, T = 121518.92, t_MRCA = 1940-04-07, n0 = 6788.4, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 6.85, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 39, 37, 18, 257, 233, 45, 39, 33, 260, 43]\n",
            "0.4 M steps/s, Step 117725000, log_posterior = -45767.75, log_G = -47346.71, log_coal = -2648.58, log_other_priors = 4227.54, num_muts = 1335, T = 116189.59, t_MRCA = 1940-01-11, n0 = 5813.9, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.61, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 39, 36, 18, 257, 233, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 118417500, log_posterior = -45811.27, log_G = -47385.83, log_coal = -2652.98, log_other_priors = 4227.54, num_muts = 1337, T = 116711.49, t_MRCA = 1940-12-20, n0 = 5715.4, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 322, kappa = 6.67, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 39, 36, 20, 257, 233, 47, 39, 34, 259, 43]\n",
            "0.4 M steps/s, Step 119110000, log_posterior = -45860.54, log_G = -47406.98, log_coal = -2681.05, log_other_priors = 4227.49, num_muts = 1335, T = 121336.76, t_MRCA = 1939-06-03, n0 = 6367.7, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.42, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 39, 37, 19, 257, 233, 46, 39, 33, 258, 43]\n",
            "0.4 M steps/s, Step 119802500, log_posterior = -45833.85, log_G = -47396.86, log_coal = -2664.56, log_other_priors = 4227.57, num_muts = 1335, T = 120554.88, t_MRCA = 1938-06-10, n0 = 5641.3, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.61, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 39, 37, 19, 257, 233, 46, 39, 33, 258, 43]\n",
            "0.4 M steps/s, Step 120495000, log_posterior = -45869.68, log_G = -47418.22, log_coal = -2679.02, log_other_priors = 4227.56, num_muts = 1335, T = 122285.65, t_MRCA = 1937-02-08, n0 = 6306.4, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1014, M_tv = 321, kappa = 6.18, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 39, 37, 20, 257, 233, 46, 39, 33, 258, 43]\n",
            "0.5 M steps/s, Step 121187500, log_posterior = -45914.57, log_G = -47455.88, log_coal = -2686.01, log_other_priors = 4227.33, num_muts = 1335, T = 126096.57, t_MRCA = 1936-10-30, n0 = 6533.1, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 7.01, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 121880000, log_posterior = -45844.82, log_G = -47408.72, log_coal = -2663.77, log_other_priors = 4227.68, num_muts = 1335, T = 122060.53, t_MRCA = 1935-12-29, n0 = 5415.6, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.34, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 36, 18, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 122572500, log_posterior = -45967.36, log_G = -47487.24, log_coal = -2707.43, log_other_priors = 4227.32, num_muts = 1335, T = 129049.40, t_MRCA = 1934-05-01, n0 = 6821.1, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.88, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 123265000, log_posterior = -45894.63, log_G = -47445.40, log_coal = -2676.62, log_other_priors = 4227.39, num_muts = 1335, T = 125181.02, t_MRCA = 1934-12-08, n0 = 6326.0, g = 0.0000, mu = 0.19 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 317, kappa = 6.87, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 35, 18, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 123957500, log_posterior = -45974.48, log_G = -47492.46, log_coal = -2709.41, log_other_priors = 4227.38, num_muts = 1335, T = 129758.89, t_MRCA = 1935-02-28, n0 = 6468.8, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.81, pi = [0.32, 0.21, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 124650000, log_posterior = -45968.37, log_G = -47484.09, log_coal = -2711.88, log_other_priors = 4227.60, num_muts = 1336, T = 127704.24, t_MRCA = 1935-06-20, n0 = 5874.2, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1019, M_tv = 317, kappa = 6.31, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [65, 267, 38, 34, 19, 257, 237, 45, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 125342500, log_posterior = -45819.96, log_G = -47390.72, log_coal = -2656.83, log_other_priors = 4227.58, num_muts = 1335, T = 120437.45, t_MRCA = 1937-04-09, n0 = 5411.2, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.74, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 126035000, log_posterior = -45816.12, log_G = -47390.98, log_coal = -2652.86, log_other_priors = 4227.72, num_muts = 1335, T = 120077.40, t_MRCA = 1936-02-12, n0 = 5345.7, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.2, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 126727500, log_posterior = -45897.15, log_G = -47430.69, log_coal = -2693.78, log_other_priors = 4227.32, num_muts = 1335, T = 123719.54, t_MRCA = 1937-01-14, n0 = 6690.2, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.95, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 127420000, log_posterior = -45866.64, log_G = -47419.99, log_coal = -2674.30, log_other_priors = 4227.65, num_muts = 1336, T = 121009.78, t_MRCA = 1938-01-01, n0 = 5456.1, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.41, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 36, 20, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 128112500, log_posterior = -45926.28, log_G = -47457.64, log_coal = -2696.29, log_other_priors = 4227.65, num_muts = 1336, T = 124927.32, t_MRCA = 1936-06-26, n0 = 6252.2, g = 0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 5.88, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 236, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 128805000, log_posterior = -45823.73, log_G = -47404.78, log_coal = -2646.55, log_other_priors = 4227.60, num_muts = 1336, T = 120176.06, t_MRCA = 1937-02-07, n0 = 5129.1, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 6.88, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 236, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 129497500, log_posterior = -45799.76, log_G = -47378.85, log_coal = -2648.78, log_other_priors = 4227.88, num_muts = 1336, T = 117674.84, t_MRCA = 1937-07-27, n0 = 4567.1, g = -0.0001, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 6.2, pi = [0.31, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 130190000, log_posterior = -45848.93, log_G = -47401.40, log_coal = -2675.13, log_other_priors = 4227.60, num_muts = 1336, T = 120024.75, t_MRCA = 1938-07-10, n0 = 6304.6, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 6.01, pi = [0.31, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 36, 18, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 130882500, log_posterior = -45910.34, log_G = -47448.05, log_coal = -2689.83, log_other_priors = 4227.54, num_muts = 1337, T = 122613.31, t_MRCA = 1938-12-08, n0 = 5671.0, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 319, kappa = 6.7, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 235, 46, 39, 34, 259, 43]\n",
            "0.5 M steps/s, Step 131575000, log_posterior = -45847.14, log_G = -47410.93, log_coal = -2663.78, log_other_priors = 4227.58, num_muts = 1336, T = 120645.58, t_MRCA = 1937-11-10, n0 = 5909.0, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 320, kappa = 6.38, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 266, 38, 36, 20, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 132267500, log_posterior = -45831.08, log_G = -47397.47, log_coal = -2661.05, log_other_priors = 4227.44, num_muts = 1336, T = 119475.40, t_MRCA = 1936-12-08, n0 = 6187.8, g = 0.0000, mu = 0.23 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 6.75, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 236, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 132960000, log_posterior = -45813.41, log_G = -47385.26, log_coal = -2655.41, log_other_priors = 4227.26, num_muts = 1336, T = 118380.91, t_MRCA = 1937-03-23, n0 = 5942.4, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 7.76, pi = [0.32, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 236, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 133652500, log_posterior = -45861.81, log_G = -47418.30, log_coal = -2671.07, log_other_priors = 4227.55, num_muts = 1336, T = 121138.24, t_MRCA = 1938-03-16, n0 = 7010.1, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 319, kappa = 5.81, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 36, 19, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 134345000, log_posterior = -45827.11, log_G = -47396.13, log_coal = -2658.56, log_other_priors = 4227.58, num_muts = 1336, T = 119442.07, t_MRCA = 1938-01-20, n0 = 5877.5, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1018, M_tv = 318, kappa = 6.4, pi = [0.31, 0.22, 0.19, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 37, 36, 19, 257, 235, 46, 39, 34, 258, 43]\n",
            "0.5 M steps/s, Step 135037500, log_posterior = -45881.81, log_G = -47418.04, log_coal = -2690.99, log_other_priors = 4227.22, num_muts = 1335, T = 122659.39, t_MRCA = 1938-02-10, n0 = 7409.4, g = 0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.95, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 37, 36, 19, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 135730000, log_posterior = -45839.28, log_G = -47399.79, log_coal = -2667.05, log_other_priors = 4227.55, num_muts = 1335, T = 121273.60, t_MRCA = 1937-09-18, n0 = 5402.7, g = -0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.88, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 36, 18, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 136422500, log_posterior = -45930.02, log_G = -47452.56, log_coal = -2705.16, log_other_priors = 4227.70, num_muts = 1335, T = 125991.34, t_MRCA = 1936-11-19, n0 = 5386.7, g = -0.0001, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1017, M_tv = 318, kappa = 6.27, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 36, 18, 257, 234, 46, 39, 34, 258, 43]\n",
            "0.4 M steps/s, Step 137115000, log_posterior = -45824.20, log_G = -47386.25, log_coal = -2665.35, log_other_priors = 4227.40, num_muts = 1334, T = 120874.57, t_MRCA = 1937-05-16, n0 = 5899.8, g = -0.0000, mu = 0.2 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 319, kappa = 7.15, pi = [0.32, 0.22, 0.2, 0.26], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 268, 38, 36, 18, 257, 233, 47, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 137807500, log_posterior = -45857.85, log_G = -47415.32, log_coal = -2670.14, log_other_priors = 4227.61, num_muts = 1335, T = 122467.13, t_MRCA = 1936-03-27, n0 = 5433.8, g = 0.0000, mu = 0.21 * 10^-3 subst / site / year, M_ts = 1016, M_tv = 319, kappa = 6.6, pi = [0.32, 0.21, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [64, 267, 38, 35, 19, 257, 235, 47, 39, 34, 257, 43]\n",
            "0.4 M steps/s, Step 138500000, log_posterior = -45765.75, log_G = -47361.11, log_coal = -2632.38, log_other_priors = 4227.74, num_muts = 1335, T = 117419.44, t_MRCA = 1935-04-09, n0 = 5007.1, g = -0.0000, mu = 0.22 * 10^-3 subst / site / year, M_ts = 1015, M_tv = 320, kappa = 6.41, pi = [0.32, 0.22, 0.2, 0.27], alpha OFF, nu = 1.000, Root Seq counts: [6054, 4070, 3751, 5085], num_muts_ab counts: [63, 267, 39, 36, 19, 257, 233, 47, 39, 34, 258, 43]\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "    <h2 style='color:green'>All done! üéâ</h2>\n",
              "    <p>Files generated in this notebook are in the left-hand file browser and in the <code>results/</code> folder.</p>\n",
              "    <ul>\n",
              "      <li>Aligned FASTA: <code>Zaire_ebolavirus_aligned_headers_adjusted.afa</code></li>\n",
              "      <li>Delphy output (.dphy): <code>Zaire_ebolavirus_delphy_out.dphy</code></li>\n",
              "      <li>Delphy log (.log): <code>Zaire_ebolavirus_delphy_beast.log</code></li>\n",
              "      <li>Delphy trees (.trees): <code>Zaire_ebolavirus_delphy_beast.trees</code></li>\n",
              "    </ul>\n",
              "    <p><strong>We recommend downloading the <code>.afa</code>, <code>.dphy</code>, and <code>metadata.csv</code> files. \n",
              "    To visualize your Delphy output, upload the <code>.dphy</code> (and optionally <code>metadata.csv</code>) file(s) to \n",
              "    <a href='https://delphy.fathom.info/' target='_blank'>https://delphy.fathom.info/</a>.</strong></p>\n",
              "    "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "CPU times: user 4.13 s, sys: 624 ms, total: 4.76 s\n",
            "Wall time: 6min 33s\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "D16JRTgMkkbU"
      },
      "execution_count": 5,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
